// @node-red/editor-client/public/vendor/ace/worker-jsonata.js
var runtime=function(e){"use strict";var t,r=Object.prototype,n=r.hasOwnProperty,a=Object.defineProperty||function(e,t,r){e[t]=r.value},o="function"==typeof Symbol?Symbol:{},i=o.iterator||"@@iterator",s=o.asyncIterator||"@@asyncIterator",u=o.toStringTag||"@@toStringTag";function c(e,t,r){return Object.defineProperty(e,t,{value:r,enumerable:!0,configurable:!0,writable:!0}),e[t]}try{c({},"")}catch(e){c=function(e,t,r){return e[t]=r}}function f(e,r,n,o){var i=r&&r.prototype instanceof b?r:b,s=Object.create(i.prototype),u=new j(o||[]);return a(s,"_invoke",{value:function(e,r,n){var a=p;return function(o,i){if(a===d)throw new Error("Generator is already running");if(a===v){if("throw"===o)throw i;return{value:t,done:!0}}for(n.method=o,n.arg=i;;){var s=n.delegate;if(s){var u=T(s,n);if(u){if(u===g)continue;return u}}if("next"===n.method)n.sent=n._sent=n.arg;else if("throw"===n.method){if(a===p)throw a=v,n.arg;n.dispatchException(n.arg)}else"return"===n.method&&n.abrupt("return",n.arg);a=d;var c=l(e,r,n);if("normal"===c.type){if(a=n.done?v:h,c.arg===g)continue;return{value:c.arg,done:n.done}}"throw"===c.type&&(a=v,n.method="throw",n.arg=c.arg)}}}(e,n,u)}),s}function l(e,t,r){try{return{type:"normal",arg:e.call(t,r)}}catch(e){return{type:"throw",arg:e}}}e.wrap=f;var p="suspendedStart",h="suspendedYield",d="executing",v="completed",g={};function b(){}function y(){}function m(){}var k={};c(k,i,function(){return this});var x=Object.getPrototypeOf,w=x&&x(x(P([])));w&&w!==r&&n.call(w,i)&&(k=w);var E=m.prototype=b.prototype=Object.create(k);function A(e){["next","throw","return"].forEach(function(t){c(e,t,function(e){return this._invoke(t,e)})})}function S(e,t){var r;a(this,"_invoke",{value:function(a,o){function i(){return new t(function(r,i){!function r(a,o,i,s){var u=l(e[a],e,o);if("throw"!==u.type){var c=u.arg,f=c.value;return f&&"object"==typeof f&&n.call(f,"__await")?t.resolve(f.__await).then(function(e){r("next",e,i,s)},function(e){r("throw",e,i,s)}):t.resolve(f).then(function(e){c.value=e,i(c)},function(e){return r("throw",e,i,s)})}s(u.arg)}(a,o,r,i)})}return r=r?r.then(i,i):i()}})}function T(e,r){var n=r.method,a=e.iterator[n];if(a===t)return r.delegate=null,"throw"===n&&e.iterator.return&&(r.method="return",r.arg=t,T(e,r),"throw"===r.method)?g:("return"!==n&&(r.method="throw",r.arg=new TypeError("The iterator does not provide a '"+n+"' method")),g);var o=l(a,e.iterator,r.arg);if("throw"===o.type)return r.method="throw",r.arg=o.arg,r.delegate=null,g;var i=o.arg;return i?i.done?(r[e.resultName]=i.value,r.next=e.nextLoc,"return"!==r.method&&(r.method="next",r.arg=t),r.delegate=null,g):i:(r.method="throw",r.arg=new TypeError("iterator result is not an object"),r.delegate=null,g)}function O(e){var t={tryLoc:e[0]};1 in e&&(t.catchLoc=e[1]),2 in e&&(t.finallyLoc=e[2],t.afterLoc=e[3]),this.tryEntries.push(t)}function D(e){var t=e.completion||{};t.type="normal",delete t.arg,e.completion=t}function j(e){this.tryEntries=[{tryLoc:"root"}],e.forEach(O,this),this.reset(!0)}function P(e){if(null!=e){var r=e[i];if(r)return r.call(e);if("function"==typeof e.next)return e;if(!isNaN(e.length)){var a=-1,o=function r(){for(;++a<e.length;)if(n.call(e,a))return r.value=e[a],r.done=!1,r;return r.value=t,r.done=!0,r};return o.next=o}}throw new TypeError(typeof e+" is not iterable")}return y.prototype=m,a(E,"constructor",{value:m,configurable:!0}),a(m,"constructor",{value:y,configurable:!0}),y.displayName=c(m,u,"GeneratorFunction"),e.isGeneratorFunction=function(e){var t="function"==typeof e&&e.constructor;return!!t&&(t===y||"GeneratorFunction"===(t.displayName||t.name))},e.mark=function(e){return Object.setPrototypeOf?Object.setPrototypeOf(e,m):(e.__proto__=m,c(e,u,"GeneratorFunction")),e.prototype=Object.create(E),e},e.awrap=function(e){return{__await:e}},A(S.prototype),c(S.prototype,s,function(){return this}),e.AsyncIterator=S,e.async=function(t,r,n,a,o){void 0===o&&(o=Promise);var i=new S(f(t,r,n,a),o);return e.isGeneratorFunction(r)?i:i.next().then(function(e){return e.done?e.value:i.next()})},A(E),c(E,u,"Generator"),c(E,i,function(){return this}),c(E,"toString",function(){return"[object Generator]"}),e.keys=function(e){var t=Object(e),r=[];for(var n in t)r.push(n);return r.reverse(),function e(){for(;r.length;){var n=r.pop();if(n in t)return e.value=n,e.done=!1,e}return e.done=!0,e}},e.values=P,j.prototype={constructor:j,reset:function(e){if(this.prev=0,this.next=0,this.sent=this._sent=t,this.done=!1,this.delegate=null,this.method="next",this.arg=t,this.tryEntries.forEach(D),!e)for(var r in this)"t"===r.charAt(0)&&n.call(this,r)&&!isNaN(+r.slice(1))&&(this[r]=t)},stop:function(){this.done=!0;var e=this.tryEntries[0].completion;if("throw"===e.type)throw e.arg;return this.rval},dispatchException:function(e){if(this.done)throw e;var r=this;function a(n,a){return s.type="throw",s.arg=e,r.next=n,a&&(r.method="next",r.arg=t),!!a}for(var o=this.tryEntries.length-1;o>=0;--o){var i=this.tryEntries[o],s=i.completion;if("root"===i.tryLoc)return a("end");if(i.tryLoc<=this.prev){var u=n.call(i,"catchLoc"),c=n.call(i,"finallyLoc");if(u&&c){if(this.prev<i.catchLoc)return a(i.catchLoc,!0);if(this.prev<i.finallyLoc)return a(i.finallyLoc)}else if(u){if(this.prev<i.catchLoc)return a(i.catchLoc,!0)}else{if(!c)throw new Error("try statement without catch or finally");if(this.prev<i.finallyLoc)return a(i.finallyLoc)}}}},abrupt:function(e,t){for(var r=this.tryEntries.length-1;r>=0;--r){var a=this.tryEntries[r];if(a.tryLoc<=this.prev&&n.call(a,"finallyLoc")&&this.prev<a.finallyLoc){var o=a;break}}o&&("break"===e||"continue"===e)&&o.tryLoc<=t&&t<=o.finallyLoc&&(o=null);var i=o?o.completion:{};return i.type=e,i.arg=t,o?(this.method="next",this.next=o.finallyLoc,g):this.complete(i)},complete:function(e,t){if("throw"===e.type)throw e.arg;return"break"===e.type||"continue"===e.type?this.next=e.arg:"return"===e.type?(this.rval=this.arg=e.arg,this.method="return",this.next="end"):"normal"===e.type&&t&&(this.next=t),g},finish:function(e){for(var t=this.tryEntries.length-1;t>=0;--t){var r=this.tryEntries[t];if(r.finallyLoc===e)return this.complete(r.completion,r.afterLoc),D(r),g}},catch:function(e){for(var t=this.tryEntries.length-1;t>=0;--t){var r=this.tryEntries[t];if(r.tryLoc===e){var n=r.completion;if("throw"===n.type){var a=n.arg;D(r)}return a}}throw new Error("illegal catch attempt")},delegateYield:function(e,r,n){return this.delegate={iterator:P(e),resultName:r,nextLoc:n},"next"===this.method&&(this.arg=t),g}},e}("object"==typeof module?module.exports:{});try{regeneratorRuntime=runtime}catch(e){"object"==typeof globalThis?globalThis.regeneratorRuntime=runtime:Function("r","regeneratorRuntime = r")(runtime)}Number.isInteger=Number.isInteger||function(e){return"number"==typeof e&&isFinite(e)&&Math.floor(e)===e},String.fromCodePoint||function(e){var t=function(t){for(var r=[],n=0,a="",o=0,i=arguments.length;o!==i;++o){var s=+arguments[o];if(!(s<1114111&&s>>>0===s))throw RangeError("Invalid code point: "+s);s<=65535?n=r.push(s):(s-=65536,n=r.push(55296+(s>>10),s%1024+56320)),n>=16383&&(a+=e.apply(null,r),r.length=0)}return a+e.apply(null,r)};try{Object.defineProperty(String,"fromCodePoint",{value:t,configurable:!0,writable:!0})}catch(e){String.fromCodePoint=t}}(String.fromCharCode),Object.is||(Object.is=function(e,t){return e===t?0!==e||1/e==1/t:e!=e&&t!=t}),String.prototype.codePointAt||function(){"use strict";var e=function(){try{var e={},t=Object.defineProperty,r=t(e,e,e)&&t}catch(e){}return r}(),t=function(e){if(null==this)throw TypeError();var t=String(this),r=t.length,n=e?Number(e):0;if(n!=n&&(n=0),!(n<0||n>=r)){var a,o=t.charCodeAt(n);return o>=55296&&o<=56319&&r>n+1&&(a=t.charCodeAt(n+1))>=56320&&a<=57343?1024*(o-55296)+a-56320+65536:o}};e?e(String.prototype,"codePointAt",{value:t,configurable:!0,writable:!0}):String.prototype.codePointAt=t}(),Math.log10=Math.log10||function(e){return Math.log(e)*Math.LOG10E},function(e){if("object"==typeof exports&&"undefined"!=typeof module)module.exports=e();else if("function"==typeof define&&define.amd)define([],e);else{("undefined"!=typeof window?window:"undefined"!=typeof global?global:"undefined"!=typeof self?self:this).jsonata=e()}}(function(){return function(){return function e(t,r,n){function a(i,s){if(!r[i]){if(!t[i]){var u="function"==typeof require&&require;if(!s&&u)return u(i,!0);if(o)return o(i,!0);var c=new Error("Cannot find module '"+i+"'");throw c.code="MODULE_NOT_FOUND",c}var f=r[i]={exports:{}};t[i][0].call(f.exports,function(e){return a(t[i][1][e]||e)},f,f.exports,e,t,r,n)}return r[i].exports}for(var o="function"==typeof require&&require,i=0;i<n.length;i++)a(n[i]);return a}}()({1:[function(e,t,r){"use strict";var n=e("./utils"),a=function(){var e=n.stringToArray,t=["Zero","One","Two","Three","Four","Five","Six","Seven","Eight","Nine","Ten","Eleven","Twelve","Thirteen","Fourteen","Fifteen","Sixteen","Seventeen","Eighteen","Nineteen"],r=["Zeroth","First","Second","Third","Fourth","Fifth","Sixth","Seventh","Eighth","Ninth","Tenth","Eleventh","Twelfth","Thirteenth","Fourteenth","Fifteenth","Sixteenth","Seventeenth","Eighteenth","Nineteenth"],a=["Twenty","Thirty","Forty","Fifty","Sixty","Seventy","Eighty","Ninety","Hundred"],o=["Thousand","Million","Billion","Trillion"];var i={};t.forEach(function(e,t){i[e.toLowerCase()]=t}),r.forEach(function(e,t){i[e.toLowerCase()]=t}),a.forEach(function(e,t){var r=e.toLowerCase();i[r]=10*(t+2),i[r.substring(0,e.length-1)+"ieth"]=i[r]}),i.hundredth=100,o.forEach(function(e,t){var r=e.toLowerCase(),n=Math.pow(10,3*(t+1));i[r]=n,i[r+"th"]=n});var s=[[1e3,"m"],[900,"cm"],[500,"d"],[400,"cd"],[100,"c"],[90,"xc"],[50,"l"],[40,"xl"],[10,"x"],[9,"ix"],[5,"v"],[4,"iv"],[1,"i"]],u={M:1e3,D:500,C:100,L:50,X:10,V:5,I:1};function c(e,t){if(void 0!==e)return p(e=Math.floor(e),d(t))}var f={DECIMAL:"decimal",LETTERS:"letters",ROMAN:"roman",WORDS:"words",SEQUENCE:"sequence"},l={UPPER:"upper",LOWER:"lower",TITLE:"title"};function p(n,i){var u,c=n<0;switch(n=Math.abs(n),i.primary){case f.LETTERS:u=function(e,t){for(var r=[],n=t.charCodeAt(0);e>0;)r.unshift(String.fromCharCode((e-1)%26+n)),e=Math.floor((e-1)/26);return r.join("")}(n,i.case===l.UPPER?"A":"a");break;case f.ROMAN:u=function e(t){for(var r=0;r<s.length;r++){var n=s[r];if(t>=n[0])return n[1]+e(t-n[0])}return""}(n),i.case===l.UPPER&&(u=u.toUpperCase());break;case f.WORDS:u=function(e,n){return function e(n,i,s){var u="";if(n<=19)u=(i?" and ":"")+(s?r[n]:t[n]);else if(n<100){var c=Math.floor(n/10),f=n%10;u=(i?" and ":"")+a[c-2],f>0?u+="-"+e(f,!1,s):s&&(u=u.substring(0,u.length-1)+"ieth")}else if(n<1e3){var l=Math.floor(n/100),p=n%100;u=(i?", ":"")+t[l]+" Hundred",p>0?u+=e(p,!0,s):s&&(u+="th")}else{var h=Math.floor(Math.log10(n)/3);h>o.length&&(h=o.length);var d=Math.pow(10,3*h),v=Math.floor(n/d),g=n-v*d;u=(i?", ":"")+e(v,!1,!1)+" "+o[h-1],g>0?u+=e(g,!0,s):s&&(u+="th")}return u}(e,!1,n)}(n,i.ordinal),i.case===l.UPPER?u=u.toUpperCase():i.case===l.LOWER&&(u=u.toLowerCase());break;case f.DECIMAL:u=""+n;var p=i.mandatoryDigits-u.length;if(p>0){var h=new Array(p+1).join("0");u=h+u}if(48!==i.zeroCode&&(u=e(u).map(function(e){return String.fromCodePoint(e.codePointAt(0)+i.zeroCode-48)}).join("")),i.regular)for(var d=Math.floor((u.length-1)/i.groupingSeparators.position);d>0;d--){var v=u.length-d*i.groupingSeparators.position;u=u.substr(0,v)+i.groupingSeparators.character+u.substr(v)}else i.groupingSeparators.reverse().forEach(function(e){var t=u.length-e.position;u=u.substr(0,t)+e.character+u.substr(t)});if(i.ordinal){var g={1:"st",2:"nd",3:"rd"}[u[u.length-1]];(!g||u.length>1&&"1"===u[u.length-2])&&(g="th"),u+=g}break;case f.SEQUENCE:throw{code:"D3130",value:i.token}}return c&&(u="-"+u),u}var h=[48,1632,1776,1984,2406,2534,2662,2790,2918,3046,3174,3302,3430,3558,3664,3792,3872,4160,4240,6112,6160,6470,6608,6784,6800,6992,7088,7232,7248,42528,43216,43264,43472,43504,43600,44016,65296];function d(t){var r,n={type:"integer",primary:f.DECIMAL,case:l.LOWER,ordinal:!1},a=t.lastIndexOf(";");switch(-1===a?r=t:(r=t.substring(0,a),"o"===t.substring(a+1)[0]&&(n.ordinal=!0)),r){case"A":n.case=l.UPPER;case"a":n.primary=f.LETTERS;break;case"I":n.case=l.UPPER;case"i":n.primary=f.ROMAN;break;case"W":n.case=l.UPPER,n.primary=f.WORDS;break;case"Ww":n.case=l.TITLE,n.primary=f.WORDS;break;case"w":n.primary=f.WORDS;break;default:var o=null,i=0,s=0,u=[],c=0;if(e(r).map(function(e){return e.codePointAt(0)}).reverse().forEach(function(e){for(var t=!1,r=0;r<h.length;r++){var n=h[r];if(e>=n&&e<=n+9){if(t=!0,i++,c++,null===o)o=n;else if(n!==o)throw{code:"D3131"};break}}t||(35===e?(c++,s++):u.push({position:c,character:String.fromCodePoint(e)}))}),i>0){n.primary=f.DECIMAL,n.zeroCode=o,n.mandatoryDigits=i,n.optionalDigits=s;var p=function(e){if(0===e.length)return 0;for(var t=e[0].character,r=1;r<e.length;r++)if(e[r].character!==t)return 0;for(var n=e.map(function(e){return e.position}),a=n.reduce(function e(t,r){return 0===r?t:e(r,t%r)}),o=1;o<=n.length;o++)if(-1===n.indexOf(o*a))return 0;return a}(u);p>0?(n.regular=!0,n.groupingSeparators={position:p,character:u[0].character}):(n.regular=!1,n.groupingSeparators=u)}else n.primary=f.SEQUENCE,n.token=r}return n}var v={Y:"1",M:"1",D:"1",d:"1",F:"n",W:"1",w:"1",X:"1",x:"1",H:"1",h:"1",P:"n",m:"01",s:"01",f:"1",Z:"01:01",z:"01:01",C:"n",E:"n"};function g(e){for(var t=[],r={type:"datetime",parts:t},n=function(r,n){if(n>r){var a=e.substring(r,n);a=a.split("]]").join("]"),t.push({type:"literal",value:a})}},a=0,o=0;o<e.length;){if("["===e.charAt(o)){if("["===e.charAt(o+1)){n(a,o),t.push({type:"literal",value:"["}),a=o+=2;continue}if(n(a,o),a=o,-1===(o=e.indexOf("]",a)))throw{code:"D3135"};var i,s=e.substring(a+1,o),u={type:"marker",component:(s=s.split(/\s+/).join("")).charAt(0)},c=s.lastIndexOf(",");if(-1!==c){var f=s.substring(c+1),p=f.indexOf("-"),h=void 0,g=void 0,b=function(e){return void 0===e||"*"===e?void 0:parseInt(e)};-1===p?h=f:(h=f.substring(0,p),g=f.substring(p+1));var y={min:b(h),max:b(g)};u.width=y,i=s.substring(1,c)}else i=s.substring(1);if(1===i.length)u.presentation1=i;else if(i.length>1){var m=i.charAt(i.length-1);-1!=="atco".indexOf(m)?(u.presentation2=m,"o"===m&&(u.ordinal=!0),u.presentation1=i.substring(0,i.length-1)):u.presentation1=i}else u.presentation1=v[u.component];if(void 0===u.presentation1)throw{code:"D3132",value:u.component};if("n"===u.presentation1[0])u.names=l.LOWER;else if("N"===u.presentation1[0])"n"===u.presentation1[1]?u.names=l.TITLE:u.names=l.UPPER;else if(-1!=="YMDdFWwXxHhmsf".indexOf(u.component)){var k=u.presentation1;if(u.presentation2&&(k+=";"+u.presentation2),u.integerFormat=d(k),u.width&&void 0!==u.width.min&&u.integerFormat.mandatoryDigits<u.width.min&&(u.integerFormat.mandatoryDigits=u.width.min),-1!=="YMD".indexOf(u.component))if(u.n=-1,u.width&&void 0!==u.width.max)u.n=u.width.max,u.integerFormat.mandatoryDigits=u.n;else{var x=u.integerFormat.mandatoryDigits+u.integerFormat.optionalDigits;x>=2&&(u.n=x)}}"Z"!==u.component&&"z"!==u.component||(u.integerFormat=d(u.presentation1)),t.push(u),a=o+1}o++}return n(a,o),r}var b=["","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday","Sunday"],y=["January","February","March","April","May","June","July","August","September","October","November","December"],m=function(e){var t=Date.UTC(e.year,e.month),r=new Date(t).getUTCDay();return 0===r&&(r=7),r>4?t+864e5*(8-r):t-864e5*(r-1)},k=function e(t,r){return{year:t,month:r,nextMonth:function(){return 11===r?e(t+1,0):e(t,r+1)},previousMonth:function(){return 0===r?e(t-1,11):e(t,r-1)},nextYear:function(){return e(t+1,r)},previousYear:function(){return e(t-1,r)}}},x=function(e,t){return(t-e)/6048e5+1},w=function(e,t){var r;switch(t){case"Y":r=e.getUTCFullYear();break;case"M":r=e.getUTCMonth()+1;break;case"D":r=e.getUTCDate();break;case"d":r=(Date.UTC(e.getUTCFullYear(),e.getUTCMonth(),e.getUTCDate())-Date.UTC(e.getUTCFullYear(),0))/864e5+1;break;case"F":0===(r=e.getUTCDay())&&(r=7);break;case"W":var n=k(e.getUTCFullYear(),0),a=m(n),o=Date.UTC(n.year,e.getUTCMonth(),e.getUTCDate()),i=x(a,o);if(i>52)o>=m(n.nextYear())&&(i=1);else if(i<1){var s=m(n.previousYear());i=x(s,o)}r=Math.floor(i);break;case"w":var u=k(e.getUTCFullYear(),e.getUTCMonth()),c=m(u),f=Date.UTC(u.year,u.month,e.getUTCDate()),l=x(c,f);if(l>4)f>=m(u.nextMonth())&&(l=1);else if(l<1){var p=m(u.previousMonth());l=x(p,f)}r=Math.floor(l);break;case"X":var h=k(e.getUTCFullYear(),0),d=m(h),v=m(h.nextYear()),g=e.getTime();r=g<d?h.year-1:g>=v?h.year+1:h.year;break;case"x":var b=k(e.getUTCFullYear(),e.getUTCMonth()),y=m(b),w=b.nextMonth(),E=m(w),A=e.getTime();r=A<y?b.previousMonth().month+1:A>=E?w.month+1:b.month+1;break;case"H":r=e.getUTCHours();break;case"h":r=e.getUTCHours(),0===(r%=12)&&(r=12);break;case"P":r=e.getUTCHours()>=12?"pm":"am";break;case"m":r=e.getUTCMinutes();break;case"s":r=e.getUTCSeconds();break;case"f":r=e.getUTCMilliseconds();break;case"Z":case"z":break;case"C":case"E":r="ISO"}return r},E=null;function A(e,t,r){var n=0,a=0;if(void 0!==r){var o=parseInt(r);n=Math.floor(o/100),a=o%100}var i;void 0===t?(null===E&&(E=g("[Y0001]-[M01]-[D01]T[H01]:[m01]:[s01].[f001][Z01:01t]")),i=E):i=g(t);var s=new Date(e+60*(60*n+a)*1e3),u="";return i.parts.forEach(function(e){"literal"===e.type?u+=e.value:u+=function(e,t){var r=w(e,t.component);if(-1!=="YMDdFWwXxHhms".indexOf(t.component))if("Y"===t.component&&-1!==t.n&&(r%=Math.pow(10,t.n)),t.names){if("M"===t.component||"x"===t.component)r=y[r-1];else{if("F"!==t.component)throw{code:"D3133",value:t.component};r=b[r]}t.names===l.UPPER?r=r.toUpperCase():t.names===l.LOWER&&(r=r.toLowerCase()),t.width&&r.length>t.width.max&&(r=r.substring(0,t.width.max))}else r=p(r,t.integerFormat);else if("f"===t.component)r=p(r,t.integerFormat);else if("Z"===t.component||"z"===t.component){var o=100*n+a;if(t.integerFormat.regular)r=p(o,t.integerFormat);else{var i=t.integerFormat.mandatoryDigits;if(1===i||2===i)r=p(n,t.integerFormat),0!==a&&(r+=":"+c(a,"00"));else{if(3!==i&&4!==i)throw{code:"D3134",value:i};r=p(o,t.integerFormat)}}o>=0&&(r="+"+r),"z"===t.component&&(r="GMT"+r),0===o&&"t"===t.presentation2&&(r="Z")}return r}(s,e)}),u}function S(e){var t={};if("datetime"===e.type)t.type="datetime",t.parts=e.parts.map(function(e){var t={};if("literal"===e.type)t.regex=e.value.replace(/[.*+?^${}()|[\]\\]/g,"\\$&");else if("Z"===e.component||"z"===e.component){var r;Array.isArray(e.integerFormat.groupingSeparators)||(r=e.integerFormat.groupingSeparators),t.regex="","z"===e.component&&(t.regex="GMT"),t.regex+="[-+][0-9]+",r&&(t.regex+=r.character+"[0-9]+"),t.parse=function(t){"z"===e.component&&(t=t.substring(3));var n=0,a=0;r?(n=Number.parseInt(t.substring(0,t.indexOf(r.character))),a=Number.parseInt(t.substring(t.indexOf(r.character)+1))):t.length-1<=2?n=Number.parseInt(t):(n=Number.parseInt(t.substring(0,3)),a=Number.parseInt(t.substring(3)));return 60*n+a}}else if(e.integerFormat)e.integerFormat.n=e.n,t=S(e.integerFormat);else{t.regex="[a-zA-Z]+";var n={};if("M"===e.component||"x"===e.component)y.forEach(function(t,r){e.width&&e.width.max?n[t.substring(0,e.width.max)]=r+1:n[t]=r+1});else if("F"===e.component)b.forEach(function(t,r){r>0&&(e.width&&e.width.max?n[t.substring(0,e.width.max)]=r:n[t]=r)});else{if("P"!==e.component)throw{code:"D3133",value:e.component};n={am:0,AM:0,pm:1,PM:1}}t.parse=function(e){return n[e]}}return t.component=e.component,t});else{t.type="integer";var r,n=e.case===l.UPPER;switch(r=e.n&&e.n>0?0===e.optionalDigits?"{".concat(e.n,"}"):"{".concat(e.n-e.optionalDigits,",").concat(e.n,"}"):"+",e.primary){case f.LETTERS:t.regex=n?"[A-Z]+":"[a-z]+",t.parse=function(e){return function(e,t){for(var r=t.charCodeAt(0),n=0,a=0;a<e.length;a++)n+=(e.charCodeAt(e.length-a-1)-r+1)*Math.pow(26,a);return n}(e,n?"A":"a")};break;case f.ROMAN:t.regex=n?"[MDCLXVI]+":"[mdclxvi]+",t.parse=function(e){return function(e){for(var t=0,r=1,n=e.length-1;n>=0;n--){var a=e[n],o=u[a];o<r?t-=o:(r=o,t+=o)}return t}(n?e:e.toUpperCase())};break;case f.WORDS:t.regex="(?:"+Object.keys(i).concat("and","[\\-, ]").join("|")+")+",t.parse=function(e){return t=e.toLowerCase(),r=t.split(/,\s|\sand\s|[\s\\-]/).map(function(e){return i[e]}),n=[0],r.forEach(function(e){if(e<100){var t=n.pop();t>=1e3&&(n.push(t),t=0),n.push(t+e)}else n.push(n.pop()*e)}),n.reduce(function(e,t){return e+t},0);var t,r,n};break;case f.DECIMAL:t.regex="[0-9]".concat(r),e.ordinal&&(t.regex+="(?:th|st|nd|rd)"),t.parse=function(t){var r=t;return e.ordinal&&(r=t.substring(0,t.length-2)),e.regular?r=r.split(",").join(""):e.groupingSeparators.forEach(function(e){r=r.split(e.character).join("")}),48!==e.zeroCode&&(r=r.split("").map(function(t){return String.fromCodePoint(t.codePointAt(0)-e.zeroCode+48)}).join("")),parseInt(r)};break;case f.SEQUENCE:throw{code:"D3130",value:e.token}}}return t}var T=new RegExp("^\\d{4}(-[01]\\d)*(-[0-3]\\d)*(T[0-2]\\d:[0-5]\\d:[0-5]\\d)*(\\.\\d+)?([+-][0-2]\\d:?[0-5]\\d|Z)?$");return{formatInteger:c,parseInteger:function(e,t){if(void 0!==e)return S(d(t)).parse(e)},fromMillis:function(e,t,r){if(void 0!==e)return A.call(this,e,t,r)},toMillis:function(e,t){if(void 0!==e){if(void 0===t){if(!T.test(e))throw{stack:(new Error).stack,code:"D3110",value:e};return Date.parse(e)}return function(e,t){var r=S(g(t)),n="^"+r.parts.map(function(e){return"("+e.regex+")"}).join("")+"$",a=new RegExp(n,"i").exec(e);if(null!==a){for(var o={},i=1;i<a.length;i++){var s=r.parts[i-1];s.parse&&(o[s.component]=s.parse(a[i]))}if(0===Object.getOwnPropertyNames(o).length)return;var u=0,c=function(e){u<<=1,u+=e?1:0},f=function(e){return!(~e&u||!(e&u))};"YXMxWwdD".split("").forEach(function(e){return c(o[e])});var l=!f(161)&&f(130),p=f(84),h=!p&&f(72);u=0,"PHhmsf".split("").forEach(function(e){return c(o[e])});var d=!f(23)&&f(47),v=(l?"YD":p?"XxwF":h?"XWF":"YMD")+(d?"Phmsf":"Hmsf"),b=this.environment.timestamp,y=!1,m=!1;if(v.split("").forEach(function(e){if(void 0===o[e])y?(o[e]=-1!=="MDd".indexOf(e)?1:0,m=!0):o[e]=w(b,e);else if(y=!0,m)throw{code:"D3136"}}),o.M>0?o.M-=1:o.M=0,l){var k=Date.UTC(o.Y,0),x=1e3*(o.d-1)*60*60*24,E=new Date(k+x);o.M=E.getUTCMonth(),o.D=E.getUTCDate()}if(p)throw{code:"D3136"};if(h)throw{code:"D3136"};d&&(o.H=12===o.h?0:o.h,1===o.P&&(o.H+=12));var A=Date.UTC(o.Y,o.M,o.D,o.H,o.m,o.s,o.f);return(o.Z||o.z)&&(A-=60*(o.Z||o.z)*1e3),A}}.call(this,e,t)}}}}();t.exports=a},{"./utils":6}],2:[function(e,t,r){(function(r){(function(){"use strict";function n(e){"@babel/helpers - typeof";return(n="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":typeof e})(e)}function a(){a=function(){return t};var e,t={},r=Object.prototype,o=r.hasOwnProperty,i=Object.defineProperty||function(e,t,r){e[t]=r.value},s="function"==typeof Symbol?Symbol:{},u=s.iterator||"@@iterator",c=s.asyncIterator||"@@asyncIterator",f=s.toStringTag||"@@toStringTag";function l(e,t,r){return Object.defineProperty(e,t,{value:r,enumerable:!0,configurable:!0,writable:!0}),e[t]}try{l({},"")}catch(e){l=function(e,t,r){return e[t]=r}}function p(e,t,r,n){var a=t&&t.prototype instanceof m?t:m,o=Object.create(a.prototype),s=new L(n||[]);return i(o,"_invoke",{value:D(e,r,s)}),o}function h(e,t,r){try{return{type:"normal",arg:e.call(t,r)}}catch(e){return{type:"throw",arg:e}}}t.wrap=p;var d="suspendedStart",v="suspendedYield",g="executing",b="completed",y={};function m(){}function k(){}function x(){}var w={};l(w,u,function(){return this});var E=Object.getPrototypeOf,A=E&&E(E(M([])));A&&A!==r&&o.call(A,u)&&(w=A);var S=x.prototype=m.prototype=Object.create(w);function T(e){["next","throw","return"].forEach(function(t){l(e,t,function(e){return this._invoke(t,e)})})}function O(e,t){function r(a,i,s,u){var c=h(e[a],e,i);if("throw"!==c.type){var f=c.arg,l=f.value;return l&&"object"==n(l)&&o.call(l,"__await")?t.resolve(l.__await).then(function(e){r("next",e,s,u)},function(e){r("throw",e,s,u)}):t.resolve(l).then(function(e){f.value=e,s(f)},function(e){return r("throw",e,s,u)})}u(c.arg)}var a;i(this,"_invoke",{value:function(e,n){function o(){return new t(function(t,a){r(e,n,t,a)})}return a=a?a.then(o,o):o()}})}function D(t,r,n){var a=d;return function(o,i){if(a===g)throw new Error("Generator is already running");if(a===b){if("throw"===o)throw i;return{value:e,done:!0}}for(n.method=o,n.arg=i;;){var s=n.delegate;if(s){var u=j(s,n);if(u){if(u===y)continue;return u}}if("next"===n.method)n.sent=n._sent=n.arg;else if("throw"===n.method){if(a===d)throw a=b,n.arg;n.dispatchException(n.arg)}else"return"===n.method&&n.abrupt("return",n.arg);a=g;var c=h(t,r,n);if("normal"===c.type){if(a=n.done?b:v,c.arg===y)continue;return{value:c.arg,done:n.done}}"throw"===c.type&&(a=b,n.method="throw",n.arg=c.arg)}}}function j(t,r){var n=r.method,a=t.iterator[n];if(a===e)return r.delegate=null,"throw"===n&&t.iterator.return&&(r.method="return",r.arg=e,j(t,r),"throw"===r.method)||"return"!==n&&(r.method="throw",r.arg=new TypeError("The iterator does not provide a '"+n+"' method")),y;var o=h(a,t.iterator,r.arg);if("throw"===o.type)return r.method="throw",r.arg=o.arg,r.delegate=null,y;var i=o.arg;return i?i.done?(r[t.resultName]=i.value,r.next=t.nextLoc,"return"!==r.method&&(r.method="next",r.arg=e),r.delegate=null,y):i:(r.method="throw",r.arg=new TypeError("iterator result is not an object"),r.delegate=null,y)}function P(e){var t={tryLoc:e[0]};1 in e&&(t.catchLoc=e[1]),2 in e&&(t.finallyLoc=e[2],t.afterLoc=e[3]),this.tryEntries.push(t)}function Y(e){var t=e.completion||{};t.type="normal",delete t.arg,e.completion=t}function L(e){this.tryEntries=[{tryLoc:"root"}],e.forEach(P,this),this.reset(!0)}function M(t){if(t||""===t){var r=t[u];if(r)return r.call(t);if("function"==typeof t.next)return t;if(!isNaN(t.length)){var a=-1,i=function r(){for(;++a<t.length;)if(o.call(t,a))return r.value=t[a],r.done=!1,r;return r.value=e,r.done=!0,r};return i.next=i}}throw new TypeError(n(t)+" is not iterable")}return k.prototype=x,i(S,"constructor",{value:x,configurable:!0}),i(x,"constructor",{value:k,configurable:!0}),k.displayName=l(x,f,"GeneratorFunction"),t.isGeneratorFunction=function(e){var t="function"==typeof e&&e.constructor;return!!t&&(t===k||"GeneratorFunction"===(t.displayName||t.name))},t.mark=function(e){return Object.setPrototypeOf?Object.setPrototypeOf(e,x):(e.__proto__=x,l(e,f,"GeneratorFunction")),e.prototype=Object.create(S),e},t.awrap=function(e){return{__await:e}},T(O.prototype),l(O.prototype,c,function(){return this}),t.AsyncIterator=O,t.async=function(e,r,n,a,o){void 0===o&&(o=Promise);var i=new O(p(e,r,n,a),o);return t.isGeneratorFunction(r)?i:i.next().then(function(e){return e.done?e.value:i.next()})},T(S),l(S,f,"Generator"),l(S,u,function(){return this}),l(S,"toString",function(){return"[object Generator]"}),t.keys=function(e){var t=Object(e),r=[];for(var n in t)r.push(n);return r.reverse(),function e(){for(;r.length;){var n=r.pop();if(n in t)return e.value=n,e.done=!1,e}return e.done=!0,e}},t.values=M,L.prototype={constructor:L,reset:function(t){if(this.prev=0,this.next=0,this.sent=this._sent=e,this.done=!1,this.delegate=null,this.method="next",this.arg=e,this.tryEntries.forEach(Y),!t)for(var r in this)"t"===r.charAt(0)&&o.call(this,r)&&!isNaN(+r.slice(1))&&(this[r]=e)},stop:function(){this.done=!0;var e=this.tryEntries[0].completion;if("throw"===e.type)throw e.arg;return this.rval},dispatchException:function(t){if(this.done)throw t;var r=this;function n(n,a){return s.type="throw",s.arg=t,r.next=n,a&&(r.method="next",r.arg=e),!!a}for(var a=this.tryEntries.length-1;a>=0;--a){var i=this.tryEntries[a],s=i.completion;if("root"===i.tryLoc)return n("end");if(i.tryLoc<=this.prev){var u=o.call(i,"catchLoc"),c=o.call(i,"finallyLoc");if(u&&c){if(this.prev<i.catchLoc)return n(i.catchLoc,!0);if(this.prev<i.finallyLoc)return n(i.finallyLoc)}else if(u){if(this.prev<i.catchLoc)return n(i.catchLoc,!0)}else{if(!c)throw new Error("try statement without catch or finally");if(this.prev<i.finallyLoc)return n(i.finallyLoc)}}}},abrupt:function(e,t){for(var r=this.tryEntries.length-1;r>=0;--r){var n=this.tryEntries[r];if(n.tryLoc<=this.prev&&o.call(n,"finallyLoc")&&this.prev<n.finallyLoc){var a=n;break}}a&&("break"===e||"continue"===e)&&a.tryLoc<=t&&t<=a.finallyLoc&&(a=null);var i=a?a.completion:{};return i.type=e,i.arg=t,a?(this.method="next",this.next=a.finallyLoc,y):this.complete(i)},complete:function(e,t){if("throw"===e.type)throw e.arg;return"break"===e.type||"continue"===e.type?this.next=e.arg:"return"===e.type?(this.rval=this.arg=e.arg,this.method="return",this.next="end"):"normal"===e.type&&t&&(this.next=t),y},finish:function(e){for(var t=this.tryEntries.length-1;t>=0;--t){var r=this.tryEntries[t];if(r.finallyLoc===e)return this.complete(r.completion,r.afterLoc),Y(r),y}},catch:function(e){for(var t=this.tryEntries.length-1;t>=0;--t){var r=this.tryEntries[t];if(r.tryLoc===e){var n=r.completion;if("throw"===n.type){var a=n.arg;Y(r)}return a}}throw new Error("illegal catch attempt")},delegateYield:function(t,r,n){return this.delegate={iterator:M(t),resultName:r,nextLoc:n},"next"===this.method&&(this.arg=e),y}},t}var o=e("./utils"),i=function(){var e=a().mark(j),t=a().mark(P),i=a().mark(Y),s=a().mark(L),u=a().mark(M),c=a().mark(U),f=a().mark(I),l=a().mark(_),p=a().mark(R),h=a().mark(z),d=a().mark($),v=a().mark(q),g=o.isNumeric,b=o.isArrayOfStrings,y=o.isArrayOfNumbers,m=o.createSequence,k=o.isSequence,x=o.isFunction,w=o.isLambda,E=o.isIterable,A=o.getFunctionArity,S=o.isDeepEqual,T=o.stringToArray;function O(e,t,r){if(void 0!==e){var n=T(e),a=n.length;if(a+t<0&&(t=0),void 0!==r){if(r<=0)return"";var o=t>=0?t+r:a+t+r;return n.slice(t,o).join("")}return n.slice(t).join("")}}function D(e){if(void 0!==e)return T(e).length}function j(t,r){var n;return a().wrap(function(e){for(;;)switch(e.prev=e.next){case 0:if(n=t.apply(this,[r]),!E(n)){e.next=4;break}return e.delegateYield(n,"t0",3);case 3:n=e.t0;case 4:if(!n||"number"==typeof n.start||"number"===n.end||Array.isArray(n.groups)||x(n.next)){e.next=6;break}throw{code:"T1010",stack:(new Error).stack};case 6:return e.abrupt("return",n);case 7:case"end":return e.stop()}},e,this)}function P(e,r){var n,o;return a().wrap(function(t){for(;;)switch(t.prev=t.next){case 0:if(void 0!==e){t.next=2;break}return t.abrupt("return",void 0);case 2:if("string"!=typeof r){t.next=6;break}n=-1!==e.indexOf(r),t.next=9;break;case 6:return t.delegateYield(j(r,e),"t0",7);case 7:o=t.t0,n=void 0!==o;case 9:return t.abrupt("return",n);case 10:case"end":return t.stop()}},t)}function Y(e,t,r){var n,o,s;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(void 0!==e){a.next=2;break}return a.abrupt("return",void 0);case 2:if(!(r<0)){a.next=4;break}throw{stack:(new Error).stack,value:r,code:"D3040",index:3};case 4:if(n=m(),!(void 0===r||r>0)){a.next=17;break}return o=0,a.delegateYield(j(t,e),"t0",8);case 8:if(void 0===(s=a.t0)){a.next=17;break}case 10:if(void 0===s||!(void 0===r||o<r)){a.next=17;break}return n.push({match:s.match,index:s.start,groups:s.groups}),a.delegateYield(j(s.next),"t1",13);case 13:s=a.t1,o++,a.next=10;break;case 17:return a.abrupt("return",n);case 18:case"end":return a.stop()}},i)}function L(e,t,r,n){var o,i,u,c,f,l,p,h;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(void 0!==e){a.next=2;break}return a.abrupt("return",void 0);case 2:if(o=this,""!==t){a.next=5;break}throw{code:"D3010",stack:(new Error).stack,value:t,index:2};case 5:if(!(n<0)){a.next=7;break}throw{code:"D3011",stack:(new Error).stack,value:n,index:4};case 7:if(i="string"==typeof r?function(e){for(var t="",n=0,a=r.indexOf("$",n);-1!==a&&n<r.length;){t+=r.substring(n,a),n=a+1;var o=r.charAt(n);if("$"===o)t+="$",n++;else if("0"===o)t+=e.match,n++;else{var i;if(i=0===e.groups.length?1:Math.floor(Math.log(e.groups.length)*Math.LOG10E)+1,a=parseInt(r.substring(n,n+i),10),i>1&&a>e.groups.length&&(a=parseInt(r.substring(n,n+i-1),10)),isNaN(a))t+="$";else{if(e.groups.length>0){var s=e.groups[a-1];void 0!==s&&(t+=s)}n+=a.toString().length}}a=r.indexOf("$",n)}return t+=r.substring(n)}:r,u="",c=0,!(void 0===n||n>0)){a.next=44;break}if(f=0,"string"!=typeof t){a.next=18;break}for(l=e.indexOf(t,c);-1!==l&&(void 0===n||f<n);)u+=e.substring(c,l),u+=r,c=l+t.length,f++,l=e.indexOf(t,c);u+=e.substring(c),a.next=42;break;case 18:return a.delegateYield(j(t,e),"t0",19);case 19:if(void 0===(p=a.t0)){a.next=41;break}case 21:if(void 0===p||!(void 0===n||f<n)){a.next=38;break}if(u+=e.substring(c,p.start),h=i.apply(o,[p]),!E(h)){a.next=27;break}return a.delegateYield(h,"t1",26);case 26:h=a.t1;case 27:if("string"!=typeof h){a.next=31;break}u+=h,a.next=32;break;case 31:throw{code:"D3012",stack:(new Error).stack,value:h};case 32:return c=p.start+p.match.length,f++,a.delegateYield(j(p.next),"t2",35);case 35:p=a.t2,a.next=21;break;case 38:u+=e.substring(c),a.next=42;break;case 41:u=e;case 42:a.next=45;break;case 44:u=e;case 45:return a.abrupt("return",u);case 46:case"end":return a.stop()}},s,this)}function M(e,t,r){var n,o,i,s;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(void 0!==e){a.next=2;break}return a.abrupt("return",void 0);case 2:if(!(r<0)){a.next=4;break}throw{code:"D3020",stack:(new Error).stack,value:r,index:3};case 4:if(n=[],!(void 0===r||r>0)){a.next=27;break}if("string"!=typeof t){a.next=10;break}n=e.split(t,r),a.next=27;break;case 10:return o=0,a.delegateYield(j(t,e),"t0",12);case 12:if(void 0===(i=a.t0)){a.next=26;break}s=0;case 15:if(void 0===i||!(void 0===r||o<r)){a.next=23;break}return n.push(e.substring(s,i.start)),s=i.end,a.delegateYield(j(i.next),"t1",19);case 19:i=a.t1,o++,a.next=15;break;case 23:(void 0===r||o<r)&&n.push(e.substring(s)),a.next=27;break;case 26:n.push(e);case 27:return a.abrupt("return",n);case 28:case"end":return a.stop()}},u)}function C(e,t){var r;if(void 0!==e){if(t){var n=e.toString().split("e");e=+(n[0]+"e"+(n[1]?+n[1]+t:t))}var a=(r=Math.round(e))-e;return.5===Math.abs(a)&&1===Math.abs(r%2)&&(r-=1),t&&(r=+((n=r.toString().split("e"))[0]+"e"+(n[1]?+n[1]-t:-t))),Object.is(r,-0)&&(r=0),r}}function F(e){if(void 0!==e){var t=!1;if(Array.isArray(e)){if(1===e.length)t=F(e[0]);else if(e.length>1){t=e.filter(function(e){return F(e)}).length>0}}else"string"==typeof e?e.length>0&&(t=!0):g(e)?0!==e&&(t=!0):null!==e&&"object"===n(e)?Object.keys(e).length>0&&(t=!0):"boolean"==typeof e&&!0===e&&(t=!0);return t}}function N(e,t,r,n){var a=[t],o=A(e);return o>=2&&a.push(r),o>=3&&a.push(n),a}function U(e,t){var r,n,o,i;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(void 0!==e){a.next=2;break}return a.abrupt("return",void 0);case 2:r=m(),n=0;case 4:if(!(n<e.length)){a.next=12;break}return o=N(t,e[n],n,e),a.delegateYield(t.apply(this,o),"t0",7);case 7:void 0!==(i=a.t0)&&r.push(i);case 9:n++,a.next=4;break;case 12:return a.abrupt("return",r);case 13:case"end":return a.stop()}},c,this)}function I(e,t){var r,n,o,i;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(void 0!==e){a.next=2;break}return a.abrupt("return",void 0);case 2:r=m(),n=0;case 4:if(!(n<e.length)){a.next=13;break}return o=e[n],i=N(t,o,n,e),a.delegateYield(t.apply(this,i),"t0",8);case 8:F(a.t0)&&r.push(o);case 10:n++,a.next=4;break;case 13:return a.abrupt("return",r);case 14:case"end":return a.stop()}},f,this)}function _(e,t){var r,n,o,i,s,u,c;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(void 0!==e){a.next=2;break}return a.abrupt("return",void 0);case 2:r=!1,o=0;case 4:if(!(o<e.length)){a.next=22;break}if(i=e[o],s=!0,void 0===t){a.next=12;break}return u=N(t,i,o,e),a.delegateYield(t.apply(this,u),"t0",10);case 10:c=a.t0,s=F(c);case 12:if(!s){a.next=19;break}if(r){a.next=18;break}n=i,r=!0,a.next=19;break;case 18:throw{stack:(new Error).stack,code:"D3138",index:o};case 19:o++,a.next=4;break;case 22:if(r){a.next=24;break}throw{stack:(new Error).stack,code:"D3139"};case 24:return a.abrupt("return",n);case 25:case"end":return a.stop()}},l,this)}function R(e,t,r){var n,o,i,s;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(void 0!==e){a.next=2;break}return a.abrupt("return",void 0);case 2:if(!((o=A(t))<2)){a.next=5;break}throw{stack:(new Error).stack,code:"D3050",index:1};case 5:void 0===r&&e.length>0?(n=e[0],i=1):(n=r,i=0);case 6:if(!(i<e.length)){a.next=15;break}return s=[n,e[i]],o>=3&&s.push(i),o>=4&&s.push(e),a.delegateYield(t.apply(this,s),"t0",11);case 11:n=a.t0,i++,a.next=6;break;case 15:return a.abrupt("return",n);case 16:case"end":return a.stop()}},p,this)}function G(e,t){return void 0===e?t:void 0===t?e:(Array.isArray(e)||(e=m(e)),Array.isArray(t)||(t=[t]),e.concat(t))}function z(e,t){var r,n,o,i;return a().wrap(function(s){for(;;)switch(s.prev=s.next){case 0:r=m(),s.t0=a().keys(e);case 2:if((s.t1=s.t0()).done){s.next=10;break}return n=s.t1.value,o=N(t,e[n],n,e),s.delegateYield(t.apply(this,o),"t2",6);case 6:void 0!==(i=s.t2)&&r.push(i),s.next=2;break;case 10:return s.abrupt("return",r);case 11:case"end":return s.stop()}},h,this)}function $(e,t){var r,n,o,i;return a().wrap(function(s){for(;;)switch(s.prev=s.next){case 0:if(void 0!==e){s.next=2;break}return s.abrupt("return",void 0);case 2:if(!(e.length<=1)){s.next=4;break}return s.abrupt("return",e);case 4:if(void 0!==t){s.next=10;break}if(y(e)||b(e)){s.next=7;break}throw{stack:(new Error).stack,code:"D3070",index:1};case 7:r=a().mark(function e(t,r){return a().wrap(function(e){for(;;)switch(e.prev=e.next){case 0:return e.abrupt("return",t>r);case 1:case"end":return e.stop()}},e)}),s.next=11;break;case 10:r=t;case 11:return n=a().mark(function e(t,n){var o,i;return a().wrap(function(e){for(;;)switch(e.prev=e.next){case 0:return o=a().mark(function e(t,n,o){return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(0!==n.length){a.next=4;break}Array.prototype.push.apply(t,o),a.next=16;break;case 4:if(0!==o.length){a.next=8;break}Array.prototype.push.apply(t,n),a.next=16;break;case 8:return a.delegateYield(r(n[0],o[0]),"t0",9);case 9:if(!a.t0){a.next=14;break}return t.push(o[0]),a.delegateYield(e(t,n,o.slice(1)),"t1",12);case 12:a.next=16;break;case 14:return t.push(n[0]),a.delegateYield(e(t,n.slice(1),o),"t2",16);case 16:case"end":return a.stop()}},e)}),i=[],e.delegateYield(o(i,t,n),"t0",3);case 3:return e.abrupt("return",i);case 4:case"end":return e.stop()}},e)}),o=a().mark(function e(t){var r,o,i;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(Array.isArray(t)&&!(t.length<=1)){a.next=4;break}return a.abrupt("return",t);case 4:return r=Math.floor(t.length/2),o=t.slice(0,r),i=t.slice(r),a.delegateYield(e(o),"t0",8);case 8:return o=a.t0,a.delegateYield(e(i),"t1",10);case 10:return i=a.t1,a.delegateYield(n(o,i),"t2",12);case 12:return a.abrupt("return",a.t2);case 13:case"end":return a.stop()}},e)}),s.delegateYield(o(e),"t0",14);case 14:return i=s.t0,s.abrupt("return",i);case 16:case"end":return s.stop()}},d)}function q(e,t){var r,n,o,i;return a().wrap(function(s){for(;;)switch(s.prev=s.next){case 0:r={},s.t0=a().keys(e);case 2:if((s.t1=s.t0()).done){s.next=11;break}return n=s.t1.value,o=e[n],i=N(t,o,n,e),s.delegateYield(t.apply(this,i),"t2",7);case 7:F(s.t2)&&(r[n]=o),s.next=2;break;case 11:return 0===Object.keys(r).length&&(r=void 0),s.abrupt("return",r);case 13:case"end":return s.stop()}},v,this)}return{sum:function(e){if(void 0!==e){var t=0;return e.forEach(function(e){t+=e}),t}},count:function(e){return void 0===e?0:e.length},max:function(e){if(void 0!==e&&0!==e.length)return Math.max.apply(Math,e)},min:function(e){if(void 0!==e&&0!==e.length)return Math.min.apply(Math,e)},average:function(e){if(void 0!==e&&0!==e.length){var t=0;return e.forEach(function(e){t+=e}),t/e.length}},string:function(e){var t=arguments.length>1&&void 0!==arguments[1]&&arguments[1];if(void 0!==e){var r;if("string"==typeof e)r=e;else if(x(e))r="";else{if("number"==typeof e&&!isFinite(e))throw{code:"D3001",value:e,stack:(new Error).stack};var n=t?2:0;Array.isArray(e)&&e.outerWrapper&&(e=e[0]),r=JSON.stringify(e,function(e,t){return null!=t&&t.toPrecision&&g(t)?Number(t.toPrecision(15)):t&&x(t)?"":t},n)}return r}},substring:O,substringBefore:function(e,t){if(void 0!==e){var r=e.indexOf(t);return r>-1?e.substr(0,r):e}},substringAfter:function(e,t){if(void 0!==e){var r=e.indexOf(t);return r>-1?e.substr(r+t.length):e}},lowercase:function(e){if(void 0!==e)return e.toLowerCase()},uppercase:function(e){if(void 0!==e)return e.toUpperCase()},length:D,trim:function(e){if(void 0!==e){var t=e.replace(/[ \t\n\r]+/gm," ");return" "===t.charAt(0)&&(t=t.substring(1))," "===t.charAt(t.length-1)&&(t=t.substring(0,t.length-1)),t}},pad:function(e,t,r){if(void 0!==e){var n;void 0!==r&&0!==r.length||(r=" ");var a=Math.abs(t)-D(e);if(a>0){var o=new Array(a+1).join(r);r.length>1&&(o=O(o,0,a)),n=t>0?e+o:o+e}else n=e;return n}},match:Y,contains:P,replace:L,split:M,join:function(e,t){if(void 0!==e)return void 0===t&&(t=""),e.join(t)},formatNumber:function(e,t,r){if(void 0!==e){var n={"decimal-separator":".","grouping-separator":",","exponent-separator":"e",infinity:"Infinity","minus-sign":"-",NaN:"NaN",percent:"%","per-mille":"‰","zero-digit":"0",digit:"#","pattern-separator":";"};void 0!==r&&Object.keys(r).forEach(function(e){n[e]=r[e]});for(var a=[],o=n["zero-digit"].charCodeAt(0),i=o;i<o+10;i++)a.push(String.fromCharCode(i));var s=a.concat([n["decimal-separator"],n["exponent-separator"],n["grouping-separator"],n.digit,n["pattern-separator"]]),u=t.split(n["pattern-separator"]);if(u.length>2)throw{code:"D3080",stack:(new Error).stack};var c=u.map(function(e){var t,r,a,o,i=function(){for(var t,r=0;r<e.length;r++)if(t=e.charAt(r),-1!==s.indexOf(t)&&t!==n["exponent-separator"])return e.substring(0,r)}(),u=function(){for(var t,r=e.length-1;r>=0;r--)if(t=e.charAt(r),-1!==s.indexOf(t)&&t!==n["exponent-separator"])return e.substring(r+1)}(),c=e.substring(i.length,e.length-u.length),f=e.indexOf(n["exponent-separator"],i.length);-1===f||f>e.length-u.length?(t=c,r=void 0):(t=c.substring(0,f),r=c.substring(f+1));var l=t.indexOf(n["decimal-separator"]);return-1===l?(a=t,o=u):(a=t.substring(0,l),o=t.substring(l+1)),{prefix:i,suffix:u,activePart:c,mantissaPart:t,exponentPart:r,integerPart:a,fractionalPart:o,subpicture:e}});c.forEach(function(e){var t,r,o=e.subpicture,i=o.indexOf(n["decimal-separator"]);i!==o.lastIndexOf(n["decimal-separator"])&&(t="D3081"),o.indexOf(n.percent)!==o.lastIndexOf(n.percent)&&(t="D3082"),o.indexOf(n["per-mille"])!==o.lastIndexOf(n["per-mille"])&&(t="D3083"),-1!==o.indexOf(n.percent)&&-1!==o.indexOf(n["per-mille"])&&(t="D3084");var u=!1;for(r=0;r<e.mantissaPart.length;r++){var c=e.mantissaPart.charAt(r);if(-1!==a.indexOf(c)||c===n.digit){u=!0;break}}u||(t="D3085"),-1!==e.activePart.split("").map(function(e){return-1===s.indexOf(e)?"p":"a"}).join("").indexOf("p")&&(t="D3086"),-1!==i?o.charAt(i-1)!==n["grouping-separator"]&&o.charAt(i+1)!==n["grouping-separator"]||(t="D3087"):e.integerPart.charAt(e.integerPart.length-1)===n["grouping-separator"]&&(t="D3088"),-1!==o.indexOf(n["grouping-separator"]+n["grouping-separator"])&&(t="D3089");var f=e.integerPart.indexOf(n.digit);-1!==f&&e.integerPart.substring(0,f).split("").filter(function(e){return a.indexOf(e)>-1}).length>0&&(t="D3090"),-1!==(f=e.fractionalPart.lastIndexOf(n.digit))&&e.fractionalPart.substring(f).split("").filter(function(e){return a.indexOf(e)>-1}).length>0&&(t="D3091");var l="string"==typeof e.exponentPart;if(l&&e.exponentPart.length>0&&(-1!==o.indexOf(n.percent)||-1!==o.indexOf(n["per-mille"]))&&(t="D3092"),l&&(0===e.exponentPart.length||e.exponentPart.split("").filter(function(e){return-1===a.indexOf(e)}).length>0)&&(t="D3093"),t)throw{code:t,stack:(new Error).stack}});var f,l,p,h,d=c.map(function(e){var t=function(t,r){for(var o=[],i=t.indexOf(n["grouping-separator"]);-1!==i;){var s=(r?t.substring(0,i):t.substring(i)).split("").filter(function(e){return-1!==a.indexOf(e)||e===n.digit}).length;o.push(s),i=e.integerPart.indexOf(n["grouping-separator"],i+1)}return o},r=t(e.integerPart),o=function(e){if(0===e.length)return 0;for(var t=e.reduce(function e(t,r){return 0===r?t:e(r,t%r)}),r=1;r<=e.length;r++)if(-1===e.indexOf(r*t))return 0;return t}(r),i=t(e.fractionalPart,!0),s=e.integerPart.split("").filter(function(e){return-1!==a.indexOf(e)}).length,u=s,c=e.fractionalPart.split(""),f=c.filter(function(e){return-1!==a.indexOf(e)}).length,l=c.filter(function(e){return-1!==a.indexOf(e)||e===n.digit}).length,p="string"==typeof e.exponentPart;0===s&&0===l&&(p?(f=1,l=1):s=1),p&&0===s&&-1!==e.integerPart.indexOf(n.digit)&&(s=1),0===s&&0===f&&(f=1);var h=0;return p&&(h=e.exponentPart.split("").filter(function(e){return-1!==a.indexOf(e)}).length),{integerPartGroupingPositions:r,regularGrouping:o,minimumIntegerPartSize:s,scalingFactor:u,prefix:e.prefix,fractionalPartGroupingPositions:i,minimumFactionalPartSize:f,maximumFactionalPartSize:l,minimumExponentSize:h,suffix:e.suffix,picture:e.subpicture}}),v=n["minus-sign"],g=n["zero-digit"],b=n["decimal-separator"],y=n["grouping-separator"];if(1===d.length&&(d.push(JSON.parse(JSON.stringify(d[0]))),d[1].prefix=v+d[1].prefix),l=-1!==(f=e>=0?d[0]:d[1]).picture.indexOf(n.percent)?100*e:-1!==f.picture.indexOf(n["per-mille"])?1e3*e:e,0===f.minimumExponentSize)p=l;else{var m=Math.pow(10,f.scalingFactor),k=Math.pow(10,f.scalingFactor-1);for(p=l,h=0;p<k;)p*=10,h-=1;for(;p>m;)p/=10,h+=1}var x=function(e,t){var r=Math.abs(e).toFixed(t);return"0"!==g&&(r=r.split("").map(function(e){return e>="0"&&e<="9"?a[e.charCodeAt(0)-48]:e}).join("")),r},w=x(C(p,f.maximumFactionalPartSize),f.maximumFactionalPartSize),E=w.indexOf(".");for(-1===E?w+=b:w=w.replace(".",b);w.charAt(0)===g;)w=w.substring(1);for(;w.charAt(w.length-1)===g;)w=w.substring(0,w.length-1);E=w.indexOf(b);var A=f.minimumIntegerPartSize-E,S=f.minimumFactionalPartSize-(w.length-E-1);if(w=(A>0?new Array(A+1).join(g):"")+w,w+=S>0?new Array(S+1).join(g):"",E=w.indexOf(b),f.regularGrouping>0)for(var T=Math.floor((E-1)/f.regularGrouping),O=1;O<=T;O++)w=[w.slice(0,E-O*f.regularGrouping),y,w.slice(E-O*f.regularGrouping)].join("");else f.integerPartGroupingPositions.forEach(function(e){w=[w.slice(0,E-e),y,w.slice(E-e)].join(""),E++});if(E=w.indexOf(b),f.fractionalPartGroupingPositions.forEach(function(e){w=[w.slice(0,e+E+1),y,w.slice(e+E+1)].join("")}),E=w.indexOf(b),-1!==f.picture.indexOf(b)&&E!==w.length-1||(w=w.substring(0,w.length-1)),void 0!==h){var D=x(h,0);(A=f.minimumExponentSize-D.length)>0&&(D=new Array(A+1).join(g)+D),w=w+n["exponent-separator"]+(h<0?v:"")+D}return w=f.prefix+w+f.suffix}},formatBase:function(e,t){if(void 0!==e){if(e=C(e),(t=void 0===t?10:C(t))<2||t>36)throw{code:"D3100",stack:(new Error).stack,value:t};return e.toString(t)}},number:function(e){var t;if(void 0!==e){if("number"==typeof e)t=e;else if("string"==typeof e&&/^-?[0-9]+(\.[0-9]+)?([Ee][-+]?[0-9]+)?$/.test(e)&&!isNaN(parseFloat(e))&&isFinite(e))t=parseFloat(e);else if(!0===e)t=1;else{if(!1!==e)throw{code:"D3030",value:e,stack:(new Error).stack,index:1};t=0}return t}},floor:function(e){if(void 0!==e)return Math.floor(e)},ceil:function(e){if(void 0!==e)return Math.ceil(e)},round:C,abs:function(e){if(void 0!==e)return Math.abs(e)},sqrt:function(e){if(void 0!==e){if(e<0)throw{stack:(new Error).stack,code:"D3060",index:1,value:e};return Math.sqrt(e)}},power:function(e,t){var r;if(void 0!==e){if(r=Math.pow(e,t),!isFinite(r))throw{stack:(new Error).stack,code:"D3061",index:1,value:e,exp:t};return r}},random:function(){return Math.random()},boolean:F,not:function(e){if(void 0!==e)return!F(e)},map:U,zip:function(){for(var e=[],t=Array.prototype.slice.call(arguments),r=Math.min.apply(Math,t.map(function(e){return Array.isArray(e)?e.length:0})),n=0;n<r;n++){var a=t.map(function(e){return e[n]});e.push(a)}return e},filter:I,single:_,foldLeft:R,sift:q,keys:function e(t){var r=m();if(Array.isArray(t)){var a={};t.forEach(function(t){e(t).forEach(function(e){a[e]=!0})}),r=e(a)}else null===t||"object"!==n(t)||w(t)||Object.keys(t).forEach(function(e){return r.push(e)});return r},lookup:function e(t,r){var a;if(Array.isArray(t)){a=m();for(var o=0;o<t.length;o++){var i=e(t[o],r);void 0!==i&&(Array.isArray(i)?i.forEach(function(e){return a.push(e)}):a.push(i))}}else null!==t&&"object"===n(t)&&(a=t[r]);return a},append:G,exists:function(e){return void 0!==e},spread:function e(t){var r=m();if(Array.isArray(t))t.forEach(function(t){r=G(r,e(t))});else if(null===t||"object"!==n(t)||w(t))r=t;else for(var a in t){var o={};o[a]=t[a],r.push(o)}return r},merge:function(e){if(void 0!==e){var t={};return e.forEach(function(e){for(var r in e)t[r]=e[r]}),t}},reverse:function(e){if(void 0!==e){if(e.length<=1)return e;for(var t=e.length,r=new Array(t),n=0;n<t;n++)r[t-n-1]=e[n];return r}},each:z,error:function(e){throw{code:"D3137",stack:(new Error).stack,message:e||"$error() function evaluated"}},assert:function(e,t){if(!e)throw{code:"D3141",stack:(new Error).stack,message:t||"$assert() statement failed"}},type:function(e){if(void 0!==e)return null===e?"null":g(e)?"number":"string"==typeof e?"string":"boolean"==typeof e?"boolean":Array.isArray(e)?"array":x(e)?"function":"object"},sort:$,shuffle:function(e){if(void 0!==e){if(e.length<=1)return e;for(var t=new Array(e.length),r=0;r<e.length;r++){var n=Math.floor(Math.random()*(r+1));r!==n&&(t[r]=t[n]),t[n]=e[r]}return t}},distinct:function(e){if(void 0!==e){if(!Array.isArray(e)||e.length<=1)return e;for(var t=k(e)?m():[],r=0;r<e.length;r++){for(var n=e[r],a=!1,o=0;o<t.length;o++)if(S(n,t[o])){a=!0;break}a||t.push(n)}return t}},base64encode:function(e){if(void 0!==e)return("undefined"!=typeof window?window.btoa:function(e){return new r.Buffer.from(e,"binary").toString("base64")})(e)},base64decode:function(e){if(void 0!==e)return("undefined"!=typeof window?window.atob:function(e){return new r.Buffer.from(e,"base64").toString("binary")})(e)},encodeUrlComponent:function(e){if(void 0!==e){var t;try{t=encodeURIComponent(e)}catch(t){throw{code:"D3140",stack:(new Error).stack,value:e,functionName:"encodeUrlComponent"}}return t}},encodeUrl:function(e){if(void 0!==e){var t;try{t=encodeURI(e)}catch(t){throw{code:"D3140",stack:(new Error).stack,value:e,functionName:"encodeUrl"}}return t}},decodeUrlComponent:function(e){if(void 0!==e){var t;try{t=decodeURIComponent(e)}catch(t){throw{code:"D3140",stack:(new Error).stack,value:e,functionName:"decodeUrlComponent"}}return t}},decodeUrl:function(e){if(void 0!==e){var t;try{t=decodeURI(e)}catch(t){throw{code:"D3140",stack:(new Error).stack,value:e,functionName:"decodeUrl"}}return t}}}}();t.exports=i}).call(this)}).call(this,"undefined"!=typeof global?global:"undefined"!=typeof self?self:"undefined"!=typeof window?window:{})},{"./utils":6}],3:[function(e,t,r){"use strict";function n(e){"@babel/helpers - typeof";return(n="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":typeof e})(e)}function a(){a=function(){return t};var e,t={},r=Object.prototype,o=r.hasOwnProperty,i=Object.defineProperty||function(e,t,r){e[t]=r.value},s="function"==typeof Symbol?Symbol:{},u=s.iterator||"@@iterator",c=s.asyncIterator||"@@asyncIterator",f=s.toStringTag||"@@toStringTag";function l(e,t,r){return Object.defineProperty(e,t,{value:r,enumerable:!0,configurable:!0,writable:!0}),e[t]}try{l({},"")}catch(e){l=function(e,t,r){return e[t]=r}}function p(e,t,r,n){var a=t&&t.prototype instanceof m?t:m,o=Object.create(a.prototype),s=new L(n||[]);return i(o,"_invoke",{value:D(e,r,s)}),o}function h(e,t,r){try{return{type:"normal",arg:e.call(t,r)}}catch(e){return{type:"throw",arg:e}}}t.wrap=p;var d="suspendedStart",v="suspendedYield",g="executing",b="completed",y={};function m(){}function k(){}function x(){}var w={};l(w,u,function(){return this});var E=Object.getPrototypeOf,A=E&&E(E(M([])));A&&A!==r&&o.call(A,u)&&(w=A);var S=x.prototype=m.prototype=Object.create(w);function T(e){["next","throw","return"].forEach(function(t){l(e,t,function(e){return this._invoke(t,e)})})}function O(e,t){function r(a,i,s,u){var c=h(e[a],e,i);if("throw"!==c.type){var f=c.arg,l=f.value;return l&&"object"==n(l)&&o.call(l,"__await")?t.resolve(l.__await).then(function(e){r("next",e,s,u)},function(e){r("throw",e,s,u)}):t.resolve(l).then(function(e){f.value=e,s(f)},function(e){return r("throw",e,s,u)})}u(c.arg)}var a;i(this,"_invoke",{value:function(e,n){function o(){return new t(function(t,a){r(e,n,t,a)})}return a=a?a.then(o,o):o()}})}function D(t,r,n){var a=d;return function(o,i){if(a===g)throw new Error("Generator is already running");if(a===b){if("throw"===o)throw i;return{value:e,done:!0}}for(n.method=o,n.arg=i;;){var s=n.delegate;if(s){var u=j(s,n);if(u){if(u===y)continue;return u}}if("next"===n.method)n.sent=n._sent=n.arg;else if("throw"===n.method){if(a===d)throw a=b,n.arg;n.dispatchException(n.arg)}else"return"===n.method&&n.abrupt("return",n.arg);a=g;var c=h(t,r,n);if("normal"===c.type){if(a=n.done?b:v,c.arg===y)continue;return{value:c.arg,done:n.done}}"throw"===c.type&&(a=b,n.method="throw",n.arg=c.arg)}}}function j(t,r){var n=r.method,a=t.iterator[n];if(a===e)return r.delegate=null,"throw"===n&&t.iterator.return&&(r.method="return",r.arg=e,j(t,r),"throw"===r.method)||"return"!==n&&(r.method="throw",r.arg=new TypeError("The iterator does not provide a '"+n+"' method")),y;var o=h(a,t.iterator,r.arg);if("throw"===o.type)return r.method="throw",r.arg=o.arg,r.delegate=null,y;var i=o.arg;return i?i.done?(r[t.resultName]=i.value,r.next=t.nextLoc,"return"!==r.method&&(r.method="next",r.arg=e),r.delegate=null,y):i:(r.method="throw",r.arg=new TypeError("iterator result is not an object"),r.delegate=null,y)}function P(e){var t={tryLoc:e[0]};1 in e&&(t.catchLoc=e[1]),2 in e&&(t.finallyLoc=e[2],t.afterLoc=e[3]),this.tryEntries.push(t)}function Y(e){var t=e.completion||{};t.type="normal",delete t.arg,e.completion=t}function L(e){this.tryEntries=[{tryLoc:"root"}],e.forEach(P,this),this.reset(!0)}function M(t){if(t||""===t){var r=t[u];if(r)return r.call(t);if("function"==typeof t.next)return t;if(!isNaN(t.length)){var a=-1,i=function r(){for(;++a<t.length;)if(o.call(t,a))return r.value=t[a],r.done=!1,r;return r.value=e,r.done=!0,r};return i.next=i}}throw new TypeError(n(t)+" is not iterable")}return k.prototype=x,i(S,"constructor",{value:x,configurable:!0}),i(x,"constructor",{value:k,configurable:!0}),k.displayName=l(x,f,"GeneratorFunction"),t.isGeneratorFunction=function(e){var t="function"==typeof e&&e.constructor;return!!t&&(t===k||"GeneratorFunction"===(t.displayName||t.name))},t.mark=function(e){return Object.setPrototypeOf?Object.setPrototypeOf(e,x):(e.__proto__=x,l(e,f,"GeneratorFunction")),e.prototype=Object.create(S),e},t.awrap=function(e){return{__await:e}},T(O.prototype),l(O.prototype,c,function(){return this}),t.AsyncIterator=O,t.async=function(e,r,n,a,o){void 0===o&&(o=Promise);var i=new O(p(e,r,n,a),o);return t.isGeneratorFunction(r)?i:i.next().then(function(e){return e.done?e.value:i.next()})},T(S),l(S,f,"Generator"),l(S,u,function(){return this}),l(S,"toString",function(){return"[object Generator]"}),t.keys=function(e){var t=Object(e),r=[];for(var n in t)r.push(n);return r.reverse(),function e(){for(;r.length;){var n=r.pop();if(n in t)return e.value=n,e.done=!1,e}return e.done=!0,e}},t.values=M,L.prototype={constructor:L,reset:function(t){if(this.prev=0,this.next=0,this.sent=this._sent=e,this.done=!1,this.delegate=null,this.method="next",this.arg=e,this.tryEntries.forEach(Y),!t)for(var r in this)"t"===r.charAt(0)&&o.call(this,r)&&!isNaN(+r.slice(1))&&(this[r]=e)},stop:function(){this.done=!0;var e=this.tryEntries[0].completion;if("throw"===e.type)throw e.arg;return this.rval},dispatchException:function(t){if(this.done)throw t;var r=this;function n(n,a){return s.type="throw",s.arg=t,r.next=n,a&&(r.method="next",r.arg=e),!!a}for(var a=this.tryEntries.length-1;a>=0;--a){var i=this.tryEntries[a],s=i.completion;if("root"===i.tryLoc)return n("end");if(i.tryLoc<=this.prev){var u=o.call(i,"catchLoc"),c=o.call(i,"finallyLoc");if(u&&c){if(this.prev<i.catchLoc)return n(i.catchLoc,!0);if(this.prev<i.finallyLoc)return n(i.finallyLoc)}else if(u){if(this.prev<i.catchLoc)return n(i.catchLoc,!0)}else{if(!c)throw new Error("try statement without catch or finally");if(this.prev<i.finallyLoc)return n(i.finallyLoc)}}}},abrupt:function(e,t){for(var r=this.tryEntries.length-1;r>=0;--r){var n=this.tryEntries[r];if(n.tryLoc<=this.prev&&o.call(n,"finallyLoc")&&this.prev<n.finallyLoc){var a=n;break}}a&&("break"===e||"continue"===e)&&a.tryLoc<=t&&t<=a.finallyLoc&&(a=null);var i=a?a.completion:{};return i.type=e,i.arg=t,a?(this.method="next",this.next=a.finallyLoc,y):this.complete(i)},complete:function(e,t){if("throw"===e.type)throw e.arg;return"break"===e.type||"continue"===e.type?this.next=e.arg:"return"===e.type?(this.rval=this.arg=e.arg,this.method="return",this.next="end"):"normal"===e.type&&t&&(this.next=t),y},finish:function(e){for(var t=this.tryEntries.length-1;t>=0;--t){var r=this.tryEntries[t];if(r.finallyLoc===e)return this.complete(r.completion,r.afterLoc),Y(r),y}},catch:function(e){for(var t=this.tryEntries.length-1;t>=0;--t){var r=this.tryEntries[t];if(r.tryLoc===e){var n=r.completion;if("throw"===n.type){var a=n.arg;Y(r)}return a}}throw new Error("illegal catch attempt")},delegateYield:function(t,r,n){return this.delegate={iterator:M(t),resultName:r,nextLoc:n},"next"===this.method&&(this.arg=e),y}},t}var o=e("./datetime"),i=e("./functions"),s=e("./utils"),u=e("./parser"),c=e("./signature"),f=function(){var e=a().mark(R),t=a().mark(G),r=a().mark($),f=a().mark(q),l=a().mark(W),p=a().mark(H),h=a().mark(Z),d=a().mark(B),v=a().mark(ne),g=a().mark(ie),b=a().mark(ce),y=a().mark(fe),m=a().mark(le),k=a().mark(de),x=a().mark(be),w=a().mark(ye),E=a().mark(me),A=a().mark(ke),S=a().mark(we),T=a().mark(Ae),O=a().mark(Oe),D=a().mark(Pe),j=s.isNumeric,P=s.isArrayOfStrings,Y=s.isArrayOfNumbers,L=s.createSequence,M=s.isSequence,C=s.isFunction,F=s.isLambda,N=s.isIterable,U=s.getFunctionArity,I=s.isDeepEqual,_=Ye(null);function R(t,r,n){var o,i,s,u;return a().wrap(function(e){for(;;)switch(e.prev=e.next){case 0:(i=n.lookup("__evaluate_entry"))&&i(t,r,n),e.t0=t.type,e.next="path"===e.t0?5:"binary"===e.t0?8:"unary"===e.t0?11:"name"===e.t0?14:"string"===e.t0?16:"number"===e.t0?16:"value"===e.t0?16:"wildcard"===e.t0?18:"descendant"===e.t0?20:"parent"===e.t0?22:"condition"===e.t0?24:"block"===e.t0?27:"bind"===e.t0?30:"regex"===e.t0?33:"function"===e.t0?35:"variable"===e.t0?38:"lambda"===e.t0?40:"partial"===e.t0?42:"apply"===e.t0?45:"transform"===e.t0?48:50;break;case 5:return e.delegateYield(G(t,r,n),"t1",6);case 6:return o=e.t1,e.abrupt("break",50);case 8:return e.delegateYield(Z(t,r,n),"t2",9);case 9:return o=e.t2,e.abrupt("break",50);case 11:return e.delegateYield(B(t,r,n),"t3",12);case 12:return o=e.t3,e.abrupt("break",50);case 14:return o=J(t,r,n),e.abrupt("break",50);case 16:return o=X(t),e.abrupt("break",50);case 18:return o=Q(t,r),e.abrupt("break",50);case 20:return o=V(t,r),e.abrupt("break",50);case 22:return o=n.lookup(t.slot.label),e.abrupt("break",50);case 24:return e.delegateYield(fe(t,r,n),"t4",25);case 25:return o=e.t4,e.abrupt("break",50);case 27:return e.delegateYield(le(t,r,n),"t5",28);case 28:return o=e.t5,e.abrupt("break",50);case 30:return e.delegateYield(ce(t,r,n),"t6",31);case 31:return o=e.t6,e.abrupt("break",50);case 33:return o=pe(t),e.abrupt("break",50);case 35:return e.delegateYield(ye(t,r,n),"t7",36);case 36:return o=e.t7,e.abrupt("break",50);case 38:return o=he(t,r,n),e.abrupt("break",50);case 40:return o=xe(t,r,n),e.abrupt("break",50);case 42:return e.delegateYield(we(t,r,n),"t8",43);case 43:return o=e.t8,e.abrupt("break",50);case 45:return e.delegateYield(be(t,r,n),"t9",46);case 46:return o=e.t9,e.abrupt("break",50);case 48:return o=ve(t,r,n),e.abrupt("break",50);case 50:if(!n.async||null!=o&&"function"==typeof o.then||(o=Promise.resolve(o)),!n.async||"function"!=typeof o.then||!t.nextFunction||"function"!=typeof o[t.nextFunction]){e.next=54;break}e.next=57;break;case 54:return e.next=56,o;case 56:o=e.sent;case 57:if(!Object.prototype.hasOwnProperty.call(t,"predicate")){e.next=65;break}s=0;case 59:if(!(s<t.predicate.length)){e.next=65;break}return e.delegateYield(H(t.predicate[s].expr,o,n),"t10",61);case 61:o=e.t10;case 62:s++,e.next=59;break;case 65:if("path"===t.type||!Object.prototype.hasOwnProperty.call(t,"group")){e.next=68;break}return e.delegateYield(ie(t.group,o,n),"t11",67);case 67:o=e.t11;case 68:return(u=n.lookup("__evaluate_exit"))&&u(t,r,n,o),o&&M(o)&&!o.tupleStream&&(t.keepArray&&(o.keepSingleton=!0),0===o.length?o=void 0:1===o.length&&(o=o.keepSingleton?o:o[0])),e.abrupt("return",o);case 72:case"end":return e.stop()}},e)}function G(e,r,n){var o,i,s,u,c,f;return a().wrap(function(t){for(;;)switch(t.prev=t.next){case 0:o=Array.isArray(r)&&"variable"!==e.steps[0].type?r:L(r),s=!1,u=void 0,c=0;case 4:if(!(c<e.steps.length)){t.next=25;break}if((f=e.steps[c]).tuple&&(s=!0),0!==c||!f.consarray){t.next=12;break}return t.delegateYield(R(f,o,n),"t0",9);case 9:i=t.t0,t.next=19;break;case 12:if(!s){t.next=17;break}return t.delegateYield(W(f,o,u,n),"t1",14);case 14:u=t.t1,t.next=19;break;case 17:return t.delegateYield($(f,o,n,c===e.steps.length-1),"t2",18);case 18:i=t.t2;case 19:if(s||void 0!==i&&0!==i.length){t.next=21;break}return t.abrupt("break",25);case 21:void 0===f.focus&&(o=i);case 22:c++,t.next=4;break;case 25:if(s)if(e.tuple)i=u;else for(i=L(),c=0;c<u.length;c++)i.push(u[c]["@"]);if(e.keepSingletonArray&&(Array.isArray(i)&&i.cons&&!i.sequence&&(i=L(i)),i.keepSingleton=!0),!e.hasOwnProperty("group")){t.next=30;break}return t.delegateYield(ie(e.group,s?u:i,n),"t3",29);case 29:i=t.t3;case 30:return t.abrupt("return",i);case 31:case"end":return t.stop()}},t)}function z(e,t){var r=Ye(e);for(var n in t)r.bind(n,t[n]);return r}function $(e,t,n,o){var i,s,u,c,f;return a().wrap(function(r){for(;;)switch(r.prev=r.next){case 0:if("sort"!==e.type){r.next=7;break}return r.delegateYield(de(e,t,n),"t0",2);case 2:if(i=r.t0,!e.stages){r.next=6;break}return r.delegateYield(q(e.stages,i,n),"t1",5);case 5:i=r.t1;case 6:return r.abrupt("return",i);case 7:i=L(),s=0;case 9:if(!(s<t.length)){r.next=24;break}return r.delegateYield(R(e,t[s],n),"t2",11);case 11:if(u=r.t2,!e.stages){r.next=20;break}c=0;case 14:if(!(c<e.stages.length)){r.next=20;break}return r.delegateYield(H(e.stages[c].expr,u,n),"t3",16);case 16:u=r.t3;case 17:c++,r.next=14;break;case 20:void 0!==u&&i.push(u);case 21:s++,r.next=9;break;case 24:return f=L(),o&&1===i.length&&Array.isArray(i[0])&&!M(i[0])?f=i[0]:i.forEach(function(e){!Array.isArray(e)||e.cons?f.push(e):e.forEach(function(e){return f.push(e)})}),r.abrupt("return",f);case 27:case"end":return r.stop()}},r)}function q(e,t,r){var n,o,i,s;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:n=t,o=0;case 2:if(!(o<e.length)){a.next=15;break}i=e[o],a.t0=i.type,a.next="filter"===a.t0?7:"index"===a.t0?10:12;break;case 7:return a.delegateYield(H(i.expr,n,r),"t1",8);case 8:return n=a.t1,a.abrupt("break",12);case 10:for(s=0;s<n.length;s++)n[s][i.value]=s;return a.abrupt("break",12);case 12:o++,a.next=2;break;case 15:return a.abrupt("return",n);case 16:case"end":return a.stop()}},f)}function W(e,t,r,n){var o,i,s,u,c,f,p,h;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if("sort"!==e.type){a.next=15;break}if(!r){a.next=6;break}return a.delegateYield(de(e,r,n),"t0",3);case 3:o=a.t0,a.next=11;break;case 6:return a.delegateYield(de(e,t,n),"t1",7);case 7:for(i=a.t1,(o=L()).tupleStream=!0,s=0;s<i.length;s++)(u={"@":i[s]})[e.index]=s,o.push(u);case 11:if(!e.stages){a.next=14;break}return a.delegateYield(q(e.stages,o,n),"t2",13);case 13:o=a.t2;case 14:return a.abrupt("return",o);case 15:(o=L()).tupleStream=!0,c=n,void 0===r&&(r=t.map(function(e){return{"@":e}})),f=0;case 20:if(!(f<r.length)){a.next=28;break}return c=z(n,r[f]),a.delegateYield(R(e,r[f]["@"],c),"t3",23);case 23:if(void 0!==(p=a.t3))for(Array.isArray(p)||(p=[p]),h=0;h<p.length;h++)u={},Object.assign(u,r[f]),p.tupleStream?Object.assign(u,p[h]):(e.focus?(u[e.focus]=p[h],u["@"]=r[f]["@"]):u["@"]=p[h],e.index&&(u[e.index]=h),e.ancestor&&(u[e.ancestor.label]=r[f]["@"])),o.push(u);case 25:f++,a.next=20;break;case 28:if(!e.stages){a.next=31;break}return a.delegateYield(q(e.stages,o,n),"t4",30);case 30:o=a.t4;case 31:return a.abrupt("return",o);case 32:case"end":return a.stop()}},l)}function H(e,t,r){var n,o,s,u,c,f;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(n=L(),t&&t.tupleStream&&(n.tupleStream=!0),Array.isArray(t)||(t=L(t)),"number"!==e.type){a.next=10;break}(o=Math.floor(e.value))<0&&(o=t.length+o),void 0!==(s=t[o])&&(Array.isArray(s)?n=s:n.push(s)),a.next=23;break;case 10:o=0;case 11:if(!(o<t.length)){a.next=23;break}return s=t[o],u=s,c=r,t.tupleStream&&(u=s["@"],c=z(r,s)),a.delegateYield(R(e,u,c),"t0",17);case 17:f=a.t0,j(f)&&(f=[f]),Y(f)?f.forEach(function(e){var r=Math.floor(e);r<0&&(r=t.length+r),r===o&&n.push(s)}):i.boolean(f)&&n.push(s);case 20:o++,a.next=11;break;case 23:return a.abrupt("return",n);case 24:case"end":return a.stop()}},p)}function Z(e,t,r){var n,o,i,s,u;return a().wrap(function(c){for(;;)switch(c.prev=c.next){case 0:return c.delegateYield(R(e.lhs,t,r),"t0",1);case 1:if(o=c.t0,i=e.value,s=a().mark(function n(){return a().wrap(function(n){for(;;)switch(n.prev=n.next){case 0:return n.delegateYield(R(e.rhs,t,r),"t0",1);case 1:return n.abrupt("return",n.t0);case 2:case"end":return n.stop()}},n)}),"and"!==i&&"or"!==i){c.next=15;break}return c.prev=5,c.delegateYield(ne(o,s,i),"t1",7);case 7:return c.abrupt("return",c.t1);case 10:throw c.prev=10,c.t2=c.catch(5),c.t2.position=e.position,c.t2.token=i,c.t2;case 15:return c.delegateYield(s(),"t3",16);case 16:u=c.t3,c.prev=17,c.t4=i,c.next="+"===c.t4?21:"-"===c.t4?21:"*"===c.t4?21:"/"===c.t4?21:"%"===c.t4?21:"="===c.t4?23:"!="===c.t4?23:"<"===c.t4?25:"<="===c.t4?25:">"===c.t4?25:">="===c.t4?25:"&"===c.t4?27:".."===c.t4?29:"in"===c.t4?31:33;break;case 21:return n=K(o,u,i),c.abrupt("break",33);case 23:return n=ee(o,u,i),c.abrupt("break",33);case 25:return n=te(o,u,i),c.abrupt("break",33);case 27:return n=oe(o,u),c.abrupt("break",33);case 29:return n=ue(o,u),c.abrupt("break",33);case 31:return n=re(o,u),c.abrupt("break",33);case 33:c.next=40;break;case 35:throw c.prev=35,c.t5=c.catch(17),c.t5.position=e.position,c.t5.token=i,c.t5;case 40:return c.abrupt("return",n);case 41:case"end":return c.stop()}},h,null,[[5,10],[17,35]])}function B(e,t,r){var n,o,s,u;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:a.t0=e.value,a.next="-"===a.t0?3:"["===a.t0?15:"{"===a.t0?27:30;break;case 3:return a.delegateYield(R(e.expression,t,r),"t1",4);case 4:if(void 0!==(n=a.t1)){a.next=9;break}n=void 0,a.next=14;break;case 9:if(!j(n)){a.next=13;break}n=-n,a.next=14;break;case 13:throw{code:"D1002",stack:(new Error).stack,position:e.position,token:e.value,value:n};case 14:return a.abrupt("break",30);case 15:n=[],o=0;case 17:if(!(o<e.expressions.length)){a.next=25;break}return s=e.expressions[o],a.delegateYield(R(s,t,r),"t2",20);case 20:void 0!==(u=a.t2)&&("["===s.value?n.push(u):n=i.append(n,u));case 22:o++,a.next=17;break;case 25:return e.consarray&&Object.defineProperty(n,"cons",{enumerable:!1,configurable:!1,value:!0}),a.abrupt("break",30);case 27:return a.delegateYield(ie(e,t,r),"t3",28);case 28:return n=a.t3,a.abrupt("break",30);case 30:return a.abrupt("return",n);case 31:case"end":return a.stop()}},d)}function J(e,t,r){return i.lookup(t,e.value)}function X(e){return e.value}function Q(e,t){var r=L();return Array.isArray(t)&&t.outerWrapper&&t.length>0&&(t=t[0]),null!==t&&"object"===n(t)&&Object.keys(t).forEach(function(e){var n=t[e];Array.isArray(n)?(n=function e(t,r){void 0===r&&(r=[]);Array.isArray(t)?t.forEach(function(t){e(t,r)}):r.push(t);return r}(n),r=i.append(r,n)):r.push(n)}),r}function V(e,t){var r,a=L();return void 0!==t&&(!function e(t,r){Array.isArray(t)||r.push(t);Array.isArray(t)?t.forEach(function(t){e(t,r)}):null!==t&&"object"===n(t)&&Object.keys(t).forEach(function(n){e(t[n],r)})}(t,a),r=1===a.length?a[0]:a),r}function K(e,t,r){var n;if(void 0!==e&&!j(e))throw{code:"T2001",stack:(new Error).stack,value:e};if(void 0!==t&&!j(t))throw{code:"T2002",stack:(new Error).stack,value:t};if(void 0===e||void 0===t)return n;switch(r){case"+":n=e+t;break;case"-":n=e-t;break;case"*":n=e*t;break;case"/":n=e/t;break;case"%":n=e%t}return n}function ee(e,t,r){var a,o=n(e),i=n(t);if("undefined"===o||"undefined"===i)return!1;switch(r){case"=":a=I(e,t);break;case"!=":a=!I(e,t)}return a}function te(e,t,r){var a,o=n(e),i=n(t);if(!("undefined"===o||"string"===o||"number"===o)||!("undefined"===i||"string"===i||"number"===i))throw{code:"T2010",stack:(new Error).stack,value:"string"!==o&&"number"!==o?e:t};if("undefined"!==o&&"undefined"!==i){if(o!==i)throw{code:"T2009",stack:(new Error).stack,value:e,value2:t};switch(r){case"<":a=e<t;break;case"<=":a=e<=t;break;case">":a=e>t;break;case">=":a=e>=t}return a}}function re(e,t){var r=!1;if(void 0===e||void 0===t)return!1;Array.isArray(t)||(t=[t]);for(var n=0;n<t.length;n++)if(t[n]===e){r=!0;break}return r}function ne(e,t,r){var n,o;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:o=ae(e),a.t0=r,a.next="and"===a.t0?4:"or"===a.t0?12:20;break;case 4:if(a.t1=o,!a.t1){a.next=10;break}return a.t2=ae,a.delegateYield(t(),"t3",8);case 8:a.t4=a.t3,a.t1=(0,a.t2)(a.t4);case 10:return n=a.t1,a.abrupt("break",20);case 12:if(a.t5=o,a.t5){a.next=18;break}return a.t6=ae,a.delegateYield(t(),"t7",16);case 16:a.t8=a.t7,a.t5=(0,a.t6)(a.t8);case 18:return n=a.t5,a.abrupt("break",20);case 20:return a.abrupt("return",n);case 21:case"end":return a.stop()}},v)}function ae(e){var t=i.boolean(e);return void 0!==t&&t}function oe(e,t){var r="",n="";return void 0!==e&&(r=i.string(e)),void 0!==t&&(n=i.string(t)),r.concat(n)}function ie(e,t,r){var n,o,s,u,c,f,l,p,h,d,v,b,y;return a().wrap(function(g){for(;;)switch(g.prev=g.next){case 0:n={},o={},s=!(!t||!t.tupleStream),Array.isArray(t)||(t=L(t)),0===t.length&&t.push(void 0),u=0;case 6:if(!(u<t.length)){g.next=31;break}c=t[u],f=s?z(r,c):r,l=0;case 10:if(!(l<e.lhs.length)){g.next=28;break}return p=e.lhs[l],g.delegateYield(R(p[0],s?c["@"]:c,f),"t0",13);case 13:if("string"==typeof(h=g.t0)||void 0===h){g.next=16;break}throw{code:"T1003",stack:(new Error).stack,position:e.position,value:h};case 16:if(void 0===h){g.next=25;break}if(d={data:c,exprIndex:l},!o.hasOwnProperty(h)){g.next=24;break}if(o[h].exprIndex===l){g.next=21;break}throw{code:"D1009",stack:(new Error).stack,position:e.position,value:h};case 21:o[h].data=i.append(o[h].data,c),g.next=25;break;case 24:o[h]=d;case 25:l++,g.next=10;break;case 28:u++,g.next=6;break;case 31:g.t1=a().keys(o);case 32:if((g.t2=g.t1()).done){g.next=43;break}return h=g.t2.value,d=o[h],v=d.data,f=r,s&&(b=se(d.data),v=b["@"],delete b["@"],f=z(r,b)),g.delegateYield(R(e.lhs[d.exprIndex][1],v,f),"t3",39);case 39:void 0!==(y=g.t3)&&(n[h]=y),g.next=32;break;case 43:return g.abrupt("return",n);case 44:case"end":return g.stop()}},g)}function se(e){if(!Array.isArray(e))return e;var t={};Object.assign(t,e[0]);for(var r=1;r<e.length;r++)for(var n in e[r])t[n]=i.append(t[n],e[r][n]);return t}function ue(e,t){var r;if(void 0!==e&&!Number.isInteger(e))throw{code:"T2003",stack:(new Error).stack,value:e};if(void 0!==t&&!Number.isInteger(t))throw{code:"T2004",stack:(new Error).stack,value:t};if(void 0===e||void 0===t)return r;if(e>t)return r;var n=t-e+1;if(n>1e7)throw{code:"D2014",stack:(new Error).stack,value:n};r=new Array(n);for(var a=e,o=0;a<=t;a++,o++)r[o]=a;return r.sequence=!0,r}function ce(e,t,r){var n;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return a.delegateYield(R(e.rhs,t,r),"t0",1);case 1:return n=a.t0,r.bind(e.lhs.value,n),a.abrupt("return",n);case 4:case"end":return a.stop()}},b)}function fe(e,t,r){var n,o;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return a.delegateYield(R(e.condition,t,r),"t0",1);case 1:if(o=a.t0,!i.boolean(o)){a.next=7;break}return a.delegateYield(R(e.then,t,r),"t1",4);case 4:n=a.t1,a.next=10;break;case 7:if(void 0===e.else){a.next=10;break}return a.delegateYield(R(e.else,t,r),"t2",9);case 9:n=a.t2;case 10:return a.abrupt("return",n);case 11:case"end":return a.stop()}},y)}function le(e,t,r){var n,o,i;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:o=Ye(r),i=0;case 2:if(!(i<e.expressions.length)){a.next=8;break}return a.delegateYield(R(e.expressions[i],t,o),"t0",4);case 4:n=a.t0;case 5:i++,a.next=2;break;case 8:return a.abrupt("return",n);case 9:case"end":return a.stop()}},m)}function pe(e){var t=new Ce.RegexEngine(e.value);return function r(n,a){var o;t.lastIndex=a||0;var i=t.exec(n);if(null!==i){if(o={match:i[0],start:i.index,end:i.index+i[0].length,groups:[]},i.length>1)for(var s=1;s<i.length;s++)o.groups.push(i[s]);o.next=function(){if(!(t.lastIndex>=n.length)){var a=r(n,t.lastIndex);if(a&&""===a.match)throw{code:"D1004",stack:(new Error).stack,position:e.position,value:e.value.source};return a}}}return o}}function he(e,t,r){return""===e.value?t&&t.outerWrapper?t[0]:t:r.lookup(e.value)}function de(e,t,r){var o,s,u,c,f;return a().wrap(function(l){for(;;)switch(l.prev=l.next){case 0:return s=t,u=!!t.tupleStream,c=a().mark(function t(o,i){var s,c,f,l,p,h,d,v,g;return a().wrap(function(t){for(;;)switch(t.prev=t.next){case 0:s=0,c=0;case 2:if(!(0===s&&c<e.terms.length)){t.next=35;break}return f=e.terms[c],l=o,p=r,u&&(l=o["@"],p=z(r,o)),t.delegateYield(R(f.expression,l,p),"t0",8);case 8:return h=t.t0,l=i,p=r,u&&(l=i["@"],p=z(r,i)),t.delegateYield(R(f.expression,l,p),"t1",13);case 13:if(d=t.t1,v=n(h),g=n(d),"undefined"!==v){t.next=19;break}return s="undefined"===g?0:1,t.abrupt("continue",32);case 19:if("undefined"!==g){t.next=22;break}return s=-1,t.abrupt("continue",32);case 22:if(!("string"!==v&&"number"!==v||"string"!==g&&"number"!==g)){t.next=24;break}throw{code:"T2008",stack:(new Error).stack,position:e.position,value:"string"!==v&&"number"!==v?h:d};case 24:if(v===g){t.next=26;break}throw{code:"T2007",stack:(new Error).stack,position:e.position,value:h,value2:d};case 26:if(h!==d){t.next=30;break}return t.abrupt("continue",32);case 30:s=h<d?-1:1;case 31:!0===f.descending&&(s=-s);case 32:c++,t.next=2;break;case 35:return t.abrupt("return",1===s);case 36:case"end":return t.stop()}},t)}),f={environment:r,input:t},l.delegateYield(i.sort.apply(f,[s,c]),"t0",5);case 5:return o=l.t0,l.abrupt("return",o);case 7:case"end":return l.stop()}},k)}function ve(e,t,r){return je(a().mark(function t(o){var i,s,u,c,f,l,p,h,d,v,g;return a().wrap(function(t){for(;;)switch(t.prev=t.next){case 0:if(void 0!==o){t.next=2;break}return t.abrupt("return",void 0);case 2:if(i=r.lookup("clone"),C(i)){t.next=5;break}throw{code:"T2013",stack:(new Error).stack,position:e.position};case 5:return t.delegateYield(me(i,[o],null,r),"t0",6);case 6:return s=t.t0,t.delegateYield(R(e.pattern,s,r),"t1",8);case 8:if(void 0===(u=t.t1)){t.next=35;break}Array.isArray(u)||(u=[u]),c=0;case 12:if(!(c<u.length)){t.next=35;break}if(!(f=u[c])||!(f.isPrototypeOf(s)||f instanceof Object.constructor)){t.next=16;break}throw{code:"D1010",stack:(new Error).stack,position:e.position};case 16:return t.delegateYield(R(e.update,f,r),"t2",17);case 17:if(l=t.t2,"undefined"===(p=n(l))){t.next=23;break}if("object"===p&&null!==l&&!Array.isArray(l)){t.next=22;break}throw{code:"T2011",stack:(new Error).stack,position:e.update.position,value:l};case 22:for(h in l)f[h]=l[h];case 23:if(void 0===e.delete){t.next=32;break}return t.delegateYield(R(e.delete,f,r),"t3",25);case 25:if(void 0===(d=t.t3)){t.next=32;break}if(v=d,Array.isArray(d)||(d=[d]),P(d)){t.next=31;break}throw{code:"T2012",stack:(new Error).stack,position:e.delete.position,value:v};case 31:for(g=0;g<d.length;g++)"object"===n(f)&&null!==f&&delete f[d[g]];case 32:c++,t.next=12;break;case 35:return t.abrupt("return",s);case 36:case"end":return t.stop()}},t)}),"<(oa):o>")}var ge=u("function($f, $g) { function($x){ $g($f($x)) } }");function be(e,t,r){var n,o,i,s;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return a.delegateYield(R(e.lhs,t,r),"t0",1);case 1:if(o=a.t0,"function"!==e.rhs.type){a.next=7;break}return a.delegateYield(ye(e.rhs,t,r,{context:o}),"t1",4);case 4:n=a.t1,a.next=20;break;case 7:return a.delegateYield(R(e.rhs,t,r),"t2",8);case 8:if(i=a.t2,C(i)){a.next=11;break}throw{code:"T2006",stack:(new Error).stack,position:e.position,value:i};case 11:if(!C(o)){a.next=18;break}return a.delegateYield(R(ge,null,r),"t3",13);case 13:return s=a.t3,a.delegateYield(me(s,[o,i],null,r),"t4",15);case 15:n=a.t4,a.next=20;break;case 18:return a.delegateYield(me(i,[o],null,r),"t5",19);case 19:n=a.t5;case 20:return a.abrupt("return",n);case 21:case"end":return a.stop()}},x)}function ye(e,t,r,o){var i,s,u,c,f,l;return a().wrap(function(p){for(;;)switch(p.prev=p.next){case 0:return p.delegateYield(R(e.procedure,t,r),"t0",1);case 1:if(void 0!==(s=p.t0)||"path"!==e.procedure.type||!r.lookup(e.procedure.steps[0].value)){p.next=4;break}throw{code:"T1005",stack:(new Error).stack,position:e.position,token:e.procedure.steps[0].value};case 4:u=[],void 0!==o&&u.push(o.context),c=a().mark(function n(){var o,i;return a().wrap(function(n){for(;;)switch(n.prev=n.next){case 0:return n.delegateYield(R(e.arguments[f],t,r),"t0",1);case 1:o=n.t0,C(o)?((i=a().mark(function e(){var t,n,i,s=arguments;return a().wrap(function(e){for(;;)switch(e.prev=e.next){case 0:for(t=s.length,n=new Array(t),i=0;i<t;i++)n[i]=s[i];return e.delegateYield(me(o,n,null,r),"t0",2);case 2:return e.abrupt("return",e.t0);case 3:case"end":return e.stop()}},e)})).arity=U(o),u.push(i)):u.push(o);case 3:case"end":return n.stop()}},n)}),f=0;case 8:if(!(f<e.arguments.length)){p.next=13;break}return p.delegateYield(c(),"t1",10);case 10:f++,p.next=8;break;case 13:return l="path"===e.procedure.type?e.procedure.steps[0].value:e.procedure.value,p.prev=14,"object"===n(s)&&(s.token=l,s.position=e.position),p.delegateYield(me(s,u,t,r),"t2",17);case 17:i=p.t2,p.next=25;break;case 20:throw p.prev=20,p.t3=p.catch(14),p.t3.position||(p.t3.position=e.position),p.t3.token||(p.t3.token=l),p.t3;case 25:return p.abrupt("return",i);case 26:case"end":return p.stop()}},w,null,[[14,20]])}function me(e,t,r,n){var o,i,s,u;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return a.delegateYield(ke(e,t,r,n),"t0",1);case 1:o=a.t0;case 2:if(!F(o)||!0!==o.thunk){a.next=21;break}return a.delegateYield(R(o.body.procedure,o.input,o.environment),"t1",4);case 4:i=a.t1,"variable"===o.body.procedure.type&&(i.token=o.body.procedure.value),i.position=o.body.procedure.position,s=[],u=0;case 9:if(!(u<o.body.arguments.length)){a.next=17;break}return a.t2=s,a.delegateYield(R(o.body.arguments[u],o.input,o.environment),"t3",12);case 12:a.t4=a.t3,a.t2.push.call(a.t2,a.t4);case 14:u++,a.next=9;break;case 17:return a.delegateYield(ke(i,s,r,n),"t5",18);case 18:o=a.t5,a.next=2;break;case 21:return a.abrupt("return",o);case 22:case"end":return a.stop()}},E)}function ke(e,t,r,n){var o,i,s;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(a.prev=0,i=t,e&&(i=Ee(e.signature,t,r)),!F(e)){a.next=8;break}return a.delegateYield(Ae(e,i),"t0",5);case 5:o=a.t0,a.next=24;break;case 8:if(!e||!0!==e._jsonata_function){a.next=16;break}if(s={environment:n,input:r},o=e.implementation.apply(s,i),!N(o)){a.next=14;break}return a.delegateYield(o,"t1",13);case 13:o=a.t1;case 14:a.next=24;break;case 16:if("function"!=typeof e){a.next=23;break}if(o=e.apply(r,i),!N(o)){a.next=21;break}return a.delegateYield(o,"t2",20);case 20:o=a.t2;case 21:a.next=24;break;case 23:throw{code:"T1006",stack:(new Error).stack};case 24:a.next=30;break;case 26:throw a.prev=26,a.t3=a.catch(0),e&&(void 0===a.t3.token&&void 0!==e.token&&(a.t3.token=e.token),a.t3.position=e.position||a.t3.position),a.t3;case 30:return a.abrupt("return",o);case 31:case"end":return a.stop()}},A,null,[[0,26]])}function xe(e,t,r){var n={_jsonata_lambda:!0,input:t,environment:r,arguments:e.arguments,signature:e.signature,body:e.body};return!0===e.thunk&&(n.thunk=!0),n.apply=a().mark(function e(r,o){return a().wrap(function(e){for(;;)switch(e.prev=e.next){case 0:return e.delegateYield(me(n,o,t,r.environment),"t0",1);case 1:return e.abrupt("return",e.t0);case 2:case"end":return e.stop()}},e)}),n}function we(e,t,r){var n,o,i,s,u;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:o=[],i=0;case 2:if(!(i<e.arguments.length)){a.next=15;break}if("operator"!==(s=e.arguments[i]).type||"?"!==s.value){a.next=8;break}o.push(s),a.next=12;break;case 8:return a.t0=o,a.delegateYield(R(s,t,r),"t1",10);case 10:a.t2=a.t1,a.t0.push.call(a.t0,a.t2);case 12:i++,a.next=2;break;case 15:return a.delegateYield(R(e.procedure,t,r),"t3",16);case 16:if(void 0!==(u=a.t3)||"path"!==e.procedure.type||!r.lookup(e.procedure.steps[0].value)){a.next=19;break}throw{code:"T1007",stack:(new Error).stack,position:e.position,token:e.procedure.steps[0].value};case 19:if(!F(u)){a.next=23;break}n=Se(u,o),a.next=32;break;case 23:if(!u||!0!==u._jsonata_function){a.next=27;break}n=Te(u.implementation,o),a.next=32;break;case 27:if("function"!=typeof u){a.next=31;break}n=Te(u,o),a.next=32;break;case 31:throw{code:"T1008",stack:(new Error).stack,position:e.position,token:"path"===e.procedure.type?e.procedure.steps[0].value:e.procedure.value};case 32:return a.abrupt("return",n);case 33:case"end":return a.stop()}},S)}function Ee(e,t,r){return void 0===e?t:e.validate(t,r)}function Ae(e,t){var r,n;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(n=Ye(e.environment),e.arguments.forEach(function(e,r){n.bind(e.value,t[r])}),"function"!=typeof e.body){a.next=7;break}return a.delegateYield(Oe(e.body,n),"t0",4);case 4:r=a.t0,a.next=9;break;case 7:return a.delegateYield(R(e.body,e.input,n),"t1",8);case 8:r=a.t1;case 9:return a.abrupt("return",r);case 10:case"end":return a.stop()}},T)}function Se(e,t){var r=Ye(e.environment),n=[];return e.arguments.forEach(function(e,a){var o=t[a];o&&"operator"===o.type&&"?"===o.value?n.push(e):r.bind(e.value,o)}),{_jsonata_lambda:!0,input:e.input,environment:r,arguments:n,body:e.body}}function Te(e,t){var r=De(e),n="function("+(r=r.map(function(e){return"$"+e.trim()})).join(", ")+"){ _ }",a=u(n);return a.body=e,Se(a,t)}function Oe(e,t){var r,n,o,i;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(r=De(e),n=r.map(function(e){return t.lookup(e.trim())}),o={environment:t},i=e.apply(o,n),!N(i)){a.next=7;break}return a.delegateYield(i,"t0",6);case 6:i=a.t0;case 7:return a.abrupt("return",i);case 8:case"end":return a.stop()}},O)}function De(e){var t=e.toString();return/\(([^)]*)\)/.exec(t)[1].split(",")}function je(e,t){var r={_jsonata_function:!0,implementation:e};return void 0!==t&&(r.signature=c(t)),r}function Pe(e,t){var r,n,o;return a().wrap(function(a){for(;;)switch(a.prev=a.next){case 0:if(void 0!==e){a.next=2;break}return a.abrupt("return",void 0);case 2:r=this.input,void 0!==t&&(r=t,Array.isArray(r)&&!M(r)&&((r=L(r)).outerWrapper=!0)),a.prev=4,n=u(e,!1),a.next=12;break;case 8:throw a.prev=8,a.t0=a.catch(4),Me(a.t0),{stack:(new Error).stack,code:"D3120",value:a.t0.message,error:a.t0};case 12:return a.prev=12,a.delegateYield(R(n,r,this.environment),"t1",14);case 14:o=a.t1,a.next=21;break;case 17:throw a.prev=17,a.t2=a.catch(12),Me(a.t2),{stack:(new Error).stack,code:"D3121",value:a.t2.message,error:a.t2};case 21:return a.abrupt("return",o);case 22:case"end":return a.stop()}},D,this,[[4,8],[12,17]])}function Ye(e){var t={};return{bind:function(e,r){t[e]=r},lookup:function(r){var n;return t.hasOwnProperty(r)?n=t[r]:e&&(n=e.lookup(r)),n},timestamp:e?e.timestamp:null,async:!!e&&e.async,global:e?e.global:{ancestry:[null]}}}_.bind("sum",je(i.sum,"<a<n>:n>")),_.bind("count",je(i.count,"<a:n>")),_.bind("max",je(i.max,"<a<n>:n>")),_.bind("min",je(i.min,"<a<n>:n>")),_.bind("average",je(i.average,"<a<n>:n>")),_.bind("string",je(i.string,"<x-b?:s>")),_.bind("substring",je(i.substring,"<s-nn?:s>")),_.bind("substringBefore",je(i.substringBefore,"<s-s:s>")),_.bind("substringAfter",je(i.substringAfter,"<s-s:s>")),_.bind("lowercase",je(i.lowercase,"<s-:s>")),_.bind("uppercase",je(i.uppercase,"<s-:s>")),_.bind("length",je(i.length,"<s-:n>")),_.bind("trim",je(i.trim,"<s-:s>")),_.bind("pad",je(i.pad,"<s-ns?:s>")),_.bind("match",je(i.match,"<s-f<s:o>n?:a<o>>")),_.bind("contains",je(i.contains,"<s-(sf):b>")),_.bind("replace",je(i.replace,"<s-(sf)(sf)n?:s>")),_.bind("split",je(i.split,"<s-(sf)n?:a<s>>")),_.bind("join",je(i.join,"<a<s>s?:s>")),_.bind("formatNumber",je(i.formatNumber,"<n-so?:s>")),_.bind("formatBase",je(i.formatBase,"<n-n?:s>")),_.bind("formatInteger",je(o.formatInteger,"<n-s:s>")),_.bind("parseInteger",je(o.parseInteger,"<s-s:n>")),_.bind("number",je(i.number,"<(nsb)-:n>")),_.bind("floor",je(i.floor,"<n-:n>")),_.bind("ceil",je(i.ceil,"<n-:n>")),_.bind("round",je(i.round,"<n-n?:n>")),_.bind("abs",je(i.abs,"<n-:n>")),_.bind("sqrt",je(i.sqrt,"<n-:n>")),_.bind("power",je(i.power,"<n-n:n>")),_.bind("random",je(i.random,"<:n>")),_.bind("boolean",je(i.boolean,"<x-:b>")),_.bind("not",je(i.not,"<x-:b>")),_.bind("map",je(i.map,"<af>")),_.bind("zip",je(i.zip,"<a+>")),_.bind("filter",je(i.filter,"<af>")),_.bind("single",je(i.single,"<af?>")),_.bind("reduce",je(i.foldLeft,"<afj?:j>")),_.bind("sift",je(i.sift,"<o-f?:o>")),_.bind("keys",je(i.keys,"<x-:a<s>>")),_.bind("lookup",je(i.lookup,"<x-s:x>")),_.bind("append",je(i.append,"<xx:a>")),_.bind("exists",je(i.exists,"<x:b>")),_.bind("spread",je(i.spread,"<x-:a<o>>")),_.bind("merge",je(i.merge,"<a<o>:o>")),_.bind("reverse",je(i.reverse,"<a:a>")),_.bind("each",je(i.each,"<o-f:a>")),_.bind("error",je(i.error,"<s?:x>")),_.bind("assert",je(i.assert,"<bs?:x>")),_.bind("type",je(i.type,"<x:s>")),_.bind("sort",je(i.sort,"<af?:a>")),_.bind("shuffle",je(i.shuffle,"<a:a>")),_.bind("distinct",je(i.distinct,"<x:x>")),_.bind("base64encode",je(i.base64encode,"<s-:s>")),_.bind("base64decode",je(i.base64decode,"<s-:s>")),_.bind("encodeUrlComponent",je(i.encodeUrlComponent,"<s-:s>")),_.bind("encodeUrl",je(i.encodeUrl,"<s-:s>")),_.bind("decodeUrlComponent",je(i.decodeUrlComponent,"<s-:s>")),_.bind("decodeUrl",je(i.decodeUrl,"<s-:s>")),_.bind("eval",je(Pe,"<sx?:x>")),_.bind("toMillis",je(o.toMillis,"<s-s?:n>")),_.bind("fromMillis",je(o.fromMillis,"<n-s?s?:s>")),_.bind("clone",je(function(e){if(void 0!==e)return JSON.parse(i.string(e))},"<(oa)-:o>"));var Le={S0101:"String literal must be terminated by a matching quote",S0102:"Number out of range: {{token}}",S0103:"Unsupported escape sequence: \\{{token}}",S0104:"The escape sequence \\u must be followed by 4 hex digits",S0105:"Quoted property name must be terminated with a backquote (`)",S0106:"Comment has no closing tag",S0201:"Syntax error: {{token}}",S0202:"Expected {{value}}, got {{token}}",S0203:"Expected {{value}} before end of expression",S0204:"Unknown operator: {{token}}",S0205:"Unexpected token: {{token}}",S0206:"Unknown expression type: {{token}}",S0207:"Unexpected end of expression",S0208:"Parameter {{value}} of function definition must be a variable name (start with $)",S0209:"A predicate cannot follow a grouping expression in a step",S0210:"Each step can only have one grouping expression",S0211:"The symbol {{token}} cannot be used as a unary operator",S0212:"The left side of := must be a variable name (start with $)",S0213:"The literal value {{value}} cannot be used as a step within a path expression",S0214:"The right side of {{token}} must be a variable name (start with $)",S0215:"A context variable binding must precede any predicates on a step",S0216:"A context variable binding must precede the 'order-by' clause on a step",S0217:"The object representing the 'parent' cannot be derived from this expression",S0301:"Empty regular expressions are not allowed",S0302:"No terminating / in regular expression",S0402:"Choice groups containing parameterized types are not supported",S0401:"Type parameters can only be applied to functions and arrays",S0500:"Attempted to evaluate an expression containing syntax error(s)",T0410:"Argument {{index}} of function {{token}} does not match function signature",T0411:"Context value is not a compatible type with argument {{index}} of function {{token}}",T0412:"Argument {{index}} of function {{token}} must be an array of {{type}}",D1001:"Number out of range: {{value}}",D1002:"Cannot negate a non-numeric value: {{value}}",T1003:"Key in object structure must evaluate to a string; got: {{value}}",D1004:"Regular expression matches zero length string",T1005:"Attempted to invoke a non-function. Did you mean ${{{token}}}?",T1006:"Attempted to invoke a non-function",T1007:"Attempted to partially apply a non-function. Did you mean ${{{token}}}?",T1008:"Attempted to partially apply a non-function",D1009:"Multiple key definitions evaluate to same key: {{value}}",D1010:"Attempted to access the Javascript object prototype",T1010:"The matcher function argument passed to function {{token}} does not return the correct object structure",T2001:"The left side of the {{token}} operator must evaluate to a number",T2002:"The right side of the {{token}} operator must evaluate to a number",T2003:"The left side of the range operator (..) must evaluate to an integer",T2004:"The right side of the range operator (..) must evaluate to an integer",D2005:"The left side of := must be a variable name (start with $)",T2006:"The right side of the function application operator ~> must be a function",T2007:"Type mismatch when comparing values {{value}} and {{value2}} in order-by clause",T2008:"The expressions within an order-by clause must evaluate to numeric or string values",T2009:"The values {{value}} and {{value2}} either side of operator {{token}} must be of the same data type",T2010:"The expressions either side of operator {{token}} must evaluate to numeric or string values",T2011:"The insert/update clause of the transform expression must evaluate to an object: {{value}}",T2012:"The delete clause of the transform expression must evaluate to a string or array of strings: {{value}}",T2013:"The transform expression clones the input object using the $clone() function.  This has been overridden in the current scope by a non-function.",D2014:"The size of the sequence allocated by the range operator (..) must not exceed 1e6.  Attempted to allocate {{value}}.",D3001:"Attempting to invoke string function on Infinity or NaN",D3010:"Second argument of replace function cannot be an empty string",D3011:"Fourth argument of replace function must evaluate to a positive number",D3012:"Attempted to replace a matched string with a non-string value",D3020:"Third argument of split function must evaluate to a positive number",D3030:"Unable to cast value to a number: {{value}}",D3040:"Third argument of match function must evaluate to a positive number",D3050:"The second argument of reduce function must be a function with at least two arguments",D3060:"The sqrt function cannot be applied to a negative number: {{value}}",D3061:"The power function has resulted in a value that cannot be represented as a JSON number: base={{value}}, exponent={{exp}}",D3070:"The single argument form of the sort function can only be applied to an array of strings or an array of numbers.  Use the second argument to specify a comparison function",D3080:"The picture string must only contain a maximum of two sub-pictures",D3081:"The sub-picture must not contain more than one instance of the 'decimal-separator' character",D3082:"The sub-picture must not contain more than one instance of the 'percent' character",D3083:"The sub-picture must not contain more than one instance of the 'per-mille' character",D3084:"The sub-picture must not contain both a 'percent' and a 'per-mille' character",D3085:"The mantissa part of a sub-picture must contain at least one character that is either an 'optional digit character' or a member of the 'decimal digit family'",D3086:"The sub-picture must not contain a passive character that is preceded by an active character and that is followed by another active character",D3087:"The sub-picture must not contain a 'grouping-separator' character that appears adjacent to a 'decimal-separator' character",D3088:"The sub-picture must not contain a 'grouping-separator' at the end of the integer part",D3089:"The sub-picture must not contain two adjacent instances of the 'grouping-separator' character",D3090:"The integer part of the sub-picture must not contain a member of the 'decimal digit family' that is followed by an instance of the 'optional digit character'",D3091:"The fractional part of the sub-picture must not contain an instance of the 'optional digit character' that is followed by a member of the 'decimal digit family'",D3092:"A sub-picture that contains a 'percent' or 'per-mille' character must not contain a character treated as an 'exponent-separator'",D3093:"The exponent part of the sub-picture must comprise only of one or more characters that are members of the 'decimal digit family'",D3100:"The radix of the formatBase function must be between 2 and 36.  It was given {{value}}",D3110:"The argument of the toMillis function must be an ISO 8601 formatted timestamp. Given {{value}}",D3120:"Syntax error in expression passed to function eval: {{value}}",D3121:"Dynamic error evaluating the expression passed to function eval: {{value}}",D3130:"Formatting or parsing an integer as a sequence starting with {{value}} is not supported by this implementation",D3131:"In a decimal digit pattern, all digits must be from the same decimal group",D3132:"Unknown component specifier {{value}} in date/time picture string",D3133:"The 'name' modifier can only be applied to months and days in the date/time picture string, not {{value}}",D3134:"The timezone integer format specifier cannot have more than four digits",D3135:"No matching closing bracket ']' in date/time picture string",D3136:"The date/time picture string is missing specifiers required to parse the timestamp",D3137:"{{{message}}}",D3138:"The $single() function expected exactly 1 matching result.  Instead it matched more.",D3139:"The $single() function expected exactly 1 matching result.  Instead it matched 0.",D3140:"Malformed URL passed to ${{{functionName}}}(): {{value}}",D3141:"{{{message}}}"};function Me(e){var t=Le[e.code];if(void 0!==t){var r=t.replace(/\{\{\{([^}]+)}}}/g,function(){return e[arguments[1]]});r=r.replace(/\{\{([^}]+)}}/g,function(){return JSON.stringify(e[arguments[1]])}),e.message=r}}function Ce(e,t){var r,n;try{r=u(e,t&&t.recover),n=r.errors,delete r.errors}catch(e){throw Me(e),e}var a=Ye(_),i=new Date;return a.bind("now",je(function(e,t){return o.fromMillis(i.getTime(),e,t)},"<s?s?:s>")),a.bind("millis",je(function(){return i.getTime()},"<:n>")),t&&t.RegexEngine?Ce.RegexEngine=t.RegexEngine:Ce.RegexEngine=RegExp,{evaluate:function(e,t,o){if(void 0!==n){var s={code:"S0500",position:0};throw Me(s),s}var u,c,f;if(void 0!==t)for(var l in u=Ye(a),t)u.bind(l,t[l]);else u=a;if(u.bind("$",e),i=new Date,u.timestamp=i,Array.isArray(e)&&!M(e)&&((e=L(e)).outerWrapper=!0),"function"==typeof o){u.async=!0;var p=function(e){Me(e),o(e,null)};f=R(r,e,u),(c=f.next()).value.then(function e(t){(c=f.next(t)).done?o(null,c.value):c.value.then(e).catch(p)}).catch(p)}else try{for(f=R(r,e,u),c=f.next();!c.done;)c=f.next(c.value);return c.value}catch(s){throw Me(s),s}},assign:function(e,t){a.bind(e,t)},registerFunction:function(e,t,r){var n=je(t,r);a.bind(e,n)},ast:function(){return r},errors:function(){return n}}}return Ce.parser=u,Ce}();t.exports=f},{"./datetime":1,"./functions":2,"./parser":4,"./signature":5,"./utils":6}],4:[function(e,t,r){"use strict";var n,a,o,i=e("./signature"),s=(n={".":75,"[":80,"]":0,"{":70,"}":0,"(":80,")":0,",":0,"@":80,"#":80,";":80,":":80,"?":20,"+":50,"-":50,"*":60,"/":60,"%":60,"|":20,"=":40,"<":40,">":40,"^":40,"**":60,"..":20,":=":10,"!=":40,"<=":40,">=":40,"~>":40,and:30,or:25,in:40,"&":50,"!":0,"~":0},a={'"':'"',"\\":"\\","/":"/",b:"\b",f:"\f",n:"\n",r:"\r",t:"\t"},o=function(e){var t=0,r=e.length,o=function(e,r){return{type:e,value:r,position:t}};return function i(s){if(t>=r)return null;for(var u=e.charAt(t);t<r&&" \t\n\r\v".indexOf(u)>-1;)t++,u=e.charAt(t);if("/"===u&&"*"===e.charAt(t+1)){var c=t;for(t+=2,u=e.charAt(t);"*"!==u||"/"!==e.charAt(t+1);)if(u=e.charAt(++t),t>=r)throw{code:"S0106",stack:(new Error).stack,position:c};return t+=2,u=e.charAt(t),i(s)}if(!0!==s&&"/"===u)return t++,o("regex",function(){for(var n,a,o=t,i=0;t<r;){var s=e.charAt(t);if("/"===s&&"\\"!==e.charAt(t-1)&&0===i){if(""===(n=e.substring(o,t)))throw{code:"S0301",stack:(new Error).stack,position:t};for(t++,s=e.charAt(t),o=t;"i"===s||"m"===s;)t++,s=e.charAt(t);return a=e.substring(o,t)+"g",new RegExp(n,a)}"("!==s&&"["!==s&&"{"!==s||"\\"===e.charAt(t-1)||i++,")"!==s&&"]"!==s&&"}"!==s||"\\"===e.charAt(t-1)||i--,t++}throw{code:"S0302",stack:(new Error).stack,position:t}}());if("."===u&&"."===e.charAt(t+1))return t+=2,o("operator","..");if(":"===u&&"="===e.charAt(t+1))return t+=2,o("operator",":=");if("!"===u&&"="===e.charAt(t+1))return t+=2,o("operator","!=");if(">"===u&&"="===e.charAt(t+1))return t+=2,o("operator",">=");if("<"===u&&"="===e.charAt(t+1))return t+=2,o("operator","<=");if("*"===u&&"*"===e.charAt(t+1))return t+=2,o("operator","**");if("~"===u&&">"===e.charAt(t+1))return t+=2,o("operator","~>");if(Object.prototype.hasOwnProperty.call(n,u))return t++,o("operator",u);if('"'===u||"'"===u){var f=u;t++;for(var l="";t<r;){if("\\"===(u=e.charAt(t)))if(t++,u=e.charAt(t),Object.prototype.hasOwnProperty.call(a,u))l+=a[u];else{if("u"!==u)throw{code:"S0103",stack:(new Error).stack,position:t,token:u};var p=e.substr(t+1,4);if(!/^[0-9a-fA-F]+$/.test(p))throw{code:"S0104",stack:(new Error).stack,position:t};var h=parseInt(p,16);l+=String.fromCharCode(h),t+=4}else{if(u===f)return t++,o("string",l);l+=u}t++}throw{code:"S0101",stack:(new Error).stack,position:t}}var d,v=/^-?(0|([1-9][0-9]*))(\.[0-9]+)?([Ee][-+]?[0-9]+)?/.exec(e.substring(t));if(null!==v){var g=parseFloat(v[0]);if(!isNaN(g)&&isFinite(g))return t+=v[0].length,o("number",g);throw{code:"S0102",stack:(new Error).stack,position:t,token:v[0]}}if("`"===u){t++;var b=e.indexOf("`",t);if(-1!==b)return d=e.substring(t,b),t=b+1,o("name",d);throw t=r,{code:"S0105",stack:(new Error).stack,position:t}}for(var y,m=t;;)if(y=e.charAt(m),m===r||" \t\n\r\v".indexOf(y)>-1||Object.prototype.hasOwnProperty.call(n,y)){if("$"===e.charAt(t))return d=e.substring(t+1,m),t=m,o("variable",d);switch(d=e.substring(t,m),t=m,d){case"or":case"in":case"and":return o("operator",d);case"true":return o("value",!0);case"false":return o("value",!1);case"null":return o("value",null);default:return t===r&&""===d?null:o("name",d)}}else m++}},function(e,t){var r,a,s={},u=[],c=function(){var e=[];"(end)"!==r.id&&e.push({type:r.type,value:r.value,position:r.position});for(var t=a();null!==t;)e.push(t),t=a();return e},f={nud:function(){var e={code:"S0211",token:this.value,position:this.position};if(t)return e.remaining=c(),e.type="error",u.push(e),e;throw e.stack=(new Error).stack,e}},l=function(e,t){var r=s[e];return t=t||0,r?t>=r.lbp&&(r.lbp=t):((r=Object.create(f)).id=r.value=e,r.lbp=t,s[e]=r),r},p=function(e){if(t){e.remaining=c(),u.push(e);var n=s["(error)"];return(r=Object.create(n)).error=e,r.type="(error)",r}throw e.stack=(new Error).stack,e},h=function(t,n){if(t&&r.id!==t){var o={code:"(end)"===r.id?"S0203":"S0202",position:r.position,token:r.value,value:t};return p(o)}var i=a(n);if(null===i)return(r=s["(end)"]).position=e.length,r;var u,c=i.value,f=i.type;switch(f){case"name":case"variable":u=s["(name)"];break;case"operator":if(!(u=s[c]))return p({code:"S0204",stack:(new Error).stack,position:i.position,token:c});break;case"string":case"number":case"value":u=s["(literal)"];break;case"regex":f="regex",u=s["(regex)"];break;default:return p({code:"S0205",stack:(new Error).stack,position:i.position,token:c})}return(r=Object.create(u)).value=c,r.type=f,r.position=i.position,r},d=function(e){var t,n=r;for(h(null,!0),t=n.nud();e<r.lbp;)n=r,h(),t=n.led(t);return t},v=function(e){l(e,0).nud=function(){return this}},g=function(e,t,r){var a=t||n[e],o=l(e,a);return o.led=r||function(e){return this.lhs=e,this.rhs=d(a),this.type="binary",this},o},b=function(e,t,r){var n=l(e,t);return n.led=r,n},y=function(e,t){var r=l(e);return r.nud=t||function(){return this.expression=d(70),this.type="unary",this},r};v("(end)"),v("(name)"),v("(literal)"),v("(regex)"),l(":"),l(";"),l(","),l(")"),l("]"),l("}"),l(".."),g("."),g("+"),g("-"),g("*"),g("/"),g("%"),g("="),g("<"),g(">"),g("!="),g("<="),g(">="),g("&"),g("and"),g("or"),g("in"),v("and"),v("or"),v("in"),y("-"),g("~>"),b("(error)",10,function(e){return this.lhs=e,this.error=r.error,this.remaining=c(),this.type="error",this}),y("*",function(){return this.type="wildcard",this}),y("**",function(){return this.type="descendant",this}),y("%",function(){return this.type="parent",this}),g("(",n["("],function(e){if(this.procedure=e,this.type="function",this.arguments=[],")"!==r.id)for(;"operator"===r.type&&"?"===r.id?(this.type="partial",this.arguments.push(r),h("?")):this.arguments.push(d(0)),","===r.id;)h(",");if(h(")",!0),"name"===e.type&&("function"===e.value||"λ"===e.value)){if(this.arguments.forEach(function(e,t){if("variable"!==e.type)return p({code:"S0208",stack:(new Error).stack,position:e.position,token:e.value,value:t+1})}),this.type="lambda","<"===r.id){for(var t=r.position,n=1,a="<";n>0&&"{"!==r.id&&"(end)"!==r.id;){var o=h();">"===o.id?n--:"<"===o.id&&n++,a+=o.value}h(">");try{this.signature=i(a)}catch(e){return e.position=t+e.offset,p(e)}}h("{"),this.body=d(0),h("}")}return this}),y("(",function(){for(var e=[];")"!==r.id&&(e.push(d(0)),";"===r.id);)h(";");return h(")",!0),this.type="block",this.expressions=e,this}),y("[",function(){var e=[];if("]"!==r.id)for(;;){var t=d(0);if(".."===r.id){var n={type:"binary",value:"..",position:r.position,lhs:t};h(".."),n.rhs=d(0),t=n}if(e.push(t),","!==r.id)break;h(",")}return h("]",!0),this.expressions=e,this.type="unary",this}),g("[",n["["],function(e){if("]"===r.id){for(var t=e;t&&"binary"===t.type&&"["===t.value;)t=t.lhs;return t.keepArray=!0,h("]"),e}return this.lhs=e,this.rhs=d(n["]"]),this.type="binary",h("]",!0),this}),g("^",n["^"],function(e){h("(");for(var t=[];;){var n={descending:!1};if("<"===r.id?h("<"):">"===r.id&&(n.descending=!0,h(">")),n.expression=d(0),t.push(n),","!==r.id)break;h(",")}return h(")"),this.lhs=e,this.rhs=t,this.type="binary",this});var m=function(e){var t=[];if("}"!==r.id)for(;;){var n=d(0);h(":");var a=d(0);if(t.push([n,a]),","!==r.id)break;h(",")}return h("}",!0),void 0===e?(this.lhs=t,this.type="unary"):(this.lhs=e,this.rhs=t,this.type="binary"),this};y("{",m),g("{",n["{"],m),b(":=",n[":="],function(e){return"variable"!==e.type?p({code:"S0212",stack:(new Error).stack,position:e.position,token:e.value}):(this.lhs=e,this.rhs=d(n[":="]-1),this.type="binary",this)}),g("@",n["@"],function(e){return this.lhs=e,this.rhs=d(n["@"]),"variable"!==this.rhs.type?p({code:"S0214",stack:(new Error).stack,position:this.rhs.position,token:"@"}):(this.type="binary",this)}),g("#",n["#"],function(e){return this.lhs=e,this.rhs=d(n["#"]),"variable"!==this.rhs.type?p({code:"S0214",stack:(new Error).stack,position:this.rhs.position,token:"#"}):(this.type="binary",this)}),g("?",n["?"],function(e){return this.type="condition",this.condition=e,this.then=d(0),":"===r.id&&(h(":"),this.else=d(0)),this}),y("|",function(){return this.type="transform",this.pattern=d(0),h("|"),this.update=d(0),","===r.id&&(h(","),this.delete=d(0)),h("|"),this});var k=0,x=0,w=[],E=function e(t,r){switch(t.type){case"name":case"wildcard":r.level--,0===r.level&&(void 0===t.ancestor?t.ancestor=r:(w[r.index].slot.label=t.ancestor.label,t.ancestor=r),t.tuple=!0);break;case"parent":r.level++;break;case"block":t.expressions.length>0&&(t.tuple=!0,r=e(t.expressions[t.expressions.length-1],r));break;case"path":t.tuple=!0;var n=t.steps.length-1;for(r=e(t.steps[n--],r);r.level>0&&n>=0;)r=e(t.steps[n--],r);break;default:throw{code:"S0217",token:t.type,position:t.position}}return r},A=function(e,t){if(void 0!==t.seekingParent||"parent"===t.type){var r=void 0!==t.seekingParent?t.seekingParent:[];"parent"===t.type&&r.push(t.slot),void 0===e.seekingParent?e.seekingParent=r:Array.prototype.push.apply(e.seekingParent,r)}},S=function(e){var t=e.steps.length-1,r=e.steps[t],n=void 0!==r.seekingParent?r.seekingParent:[];"parent"===r.type&&n.push(r.slot);for(var a=0;a<n.length;a++){var o=n[a];for(t=e.steps.length-2;o.level>0;){if(t<0){void 0===e.seekingParent?e.seekingParent=[o]:e.seekingParent.push(o);break}for(var i=e.steps[t--];t>=0&&i.focus&&e.steps[t].focus;)i=e.steps[t--];o=E(i,o)}}};a=o(e),h();var T=d(0);if("(end)"!==r.id){var O={code:"S0201",position:r.position,token:r.value};p(O)}if("parent"===(T=function e(r){var n;switch(r.type){case"binary":switch(r.value){case".":var a=e(r.lhs);n="path"===a.type?a:{type:"path",steps:[a]},"parent"===a.type&&(n.seekingParent=[a.slot]);var o=e(r.rhs);"function"===o.type&&"path"===o.procedure.type&&1===o.procedure.steps.length&&"name"===o.procedure.steps[0].type&&"function"===n.steps[n.steps.length-1].type&&(n.steps[n.steps.length-1].nextFunction=o.procedure.steps[0].value),"path"===o.type?Array.prototype.push.apply(n.steps,o.steps):(void 0!==o.predicate&&(o.stages=o.predicate,delete o.predicate),n.steps.push(o)),n.steps.filter(function(e){if("number"===e.type||"value"===e.type)throw{code:"S0213",stack:(new Error).stack,position:e.position,value:e.value};return"string"===e.type}).forEach(function(e){e.type="name"}),n.steps.filter(function(e){return!0===e.keepArray}).length>0&&(n.keepSingletonArray=!0);var i=n.steps[0];"unary"===i.type&&"["===i.value&&(i.consarray=!0);var s=n.steps[n.steps.length-1];"unary"===s.type&&"["===s.value&&(s.consarray=!0),S(n);break;case"[":var c=n=e(r.lhs),f="predicate";if("path"===n.type&&(c=n.steps[n.steps.length-1],f="stages"),void 0!==c.group)throw{code:"S0209",stack:(new Error).stack,position:r.position};void 0===c[f]&&(c[f]=[]);var l=e(r.rhs);void 0!==l.seekingParent&&(l.seekingParent.forEach(function(e){1===e.level?E(c,e):e.level--}),A(c,l)),c[f].push({type:"filter",expr:l,position:r.position});break;case"{":if(void 0!==(n=e(r.lhs)).group)throw{code:"S0210",stack:(new Error).stack,position:r.position};n.group={lhs:r.rhs.map(function(t){return[e(t[0]),e(t[1])]}),position:r.position};break;case"^":"path"!==(n=e(r.lhs)).type&&(n={type:"path",steps:[n]});var p={type:"sort",position:r.position};p.terms=r.rhs.map(function(t){var r=e(t.expression);return A(p,r),{descending:t.descending,expression:r}}),n.steps.push(p),S(n);break;case":=":(n={type:"bind",value:r.value,position:r.position}).lhs=e(r.lhs),n.rhs=e(r.rhs),A(n,n.rhs);break;case"@":if(n=e(r.lhs),c=n,"path"===n.type&&(c=n.steps[n.steps.length-1]),void 0!==c.stages||void 0!==c.predicate)throw{code:"S0215",stack:(new Error).stack,position:r.position};if("sort"===c.type)throw{code:"S0216",stack:(new Error).stack,position:r.position};r.keepArray&&(c.keepArray=!0),c.focus=r.rhs.value,c.tuple=!0;break;case"#":n=e(r.lhs),c=n,"path"===n.type?c=n.steps[n.steps.length-1]:(n={type:"path",steps:[n]},void 0!==c.predicate&&(c.stages=c.predicate,delete c.predicate)),void 0===c.stages?c.index=r.rhs.value:c.stages.push({type:"index",value:r.rhs.value,position:r.position}),c.tuple=!0;break;case"~>":(n={type:"apply",value:r.value,position:r.position}).lhs=e(r.lhs),n.rhs=e(r.rhs);break;default:(n={type:r.type,value:r.value,position:r.position}).lhs=e(r.lhs),n.rhs=e(r.rhs),A(n,n.lhs),A(n,n.rhs)}break;case"unary":n={type:r.type,value:r.value,position:r.position},"["===r.value?n.expressions=r.expressions.map(function(t){var r=e(t);return A(n,r),r}):"{"===r.value?n.lhs=r.lhs.map(function(t){var r=e(t[0]);A(n,r);var a=e(t[1]);return A(n,a),[r,a]}):(n.expression=e(r.expression),"-"===r.value&&"number"===n.expression.type?(n=n.expression).value=-n.value:A(n,n.expression));break;case"function":case"partial":(n={type:r.type,name:r.name,value:r.value,position:r.position}).arguments=r.arguments.map(function(t){var r=e(t);return A(n,r),r}),n.procedure=e(r.procedure);break;case"lambda":n={type:r.type,arguments:r.arguments,signature:r.signature,position:r.position};var h=e(r.body);n.body=function e(t){var r;if("function"!==t.type||t.predicate)if("condition"===t.type)t.then=e(t.then),void 0!==t.else&&(t.else=e(t.else)),r=t;else if("block"===t.type){var n=t.expressions.length;n>0&&(t.expressions[n-1]=e(t.expressions[n-1])),r=t}else r=t;else{var a={type:"lambda",thunk:!0,arguments:[],position:t.position};a.body=t,r=a}return r}(h);break;case"condition":(n={type:r.type,position:r.position}).condition=e(r.condition),A(n,n.condition),n.then=e(r.then),A(n,n.then),void 0!==r.else&&(n.else=e(r.else),A(n,n.else));break;case"transform":(n={type:r.type,position:r.position}).pattern=e(r.pattern),n.update=e(r.update),void 0!==r.delete&&(n.delete=e(r.delete));break;case"block":(n={type:r.type,position:r.position}).expressions=r.expressions.map(function(t){var r=e(t);return A(n,r),(r.consarray||"path"===r.type&&r.steps[0].consarray)&&(n.consarray=!0),r});break;case"name":n={type:"path",steps:[r]},r.keepArray&&(n.keepSingletonArray=!0);break;case"parent":n={type:"parent",slot:{label:"!"+k++,level:1,index:x++}},w.push(n);break;case"string":case"number":case"value":case"wildcard":case"descendant":case"variable":case"regex":n=r;break;case"operator":if("and"===r.value||"or"===r.value||"in"===r.value)r.type="name",n=e(r);else{if("?"!==r.value)throw{code:"S0201",stack:(new Error).stack,position:r.position,token:r.value};n=r}break;case"error":n=r,r.lhs&&(n=e(r.lhs));break;default:var d="S0206";"(end)"===r.id&&(d="S0207");var v={code:d,position:r.position,token:r.value};if(t)return u.push(v),{type:"error",error:v};throw v.stack=(new Error).stack,v}return r.keepArray&&(n.keepArray=!0),n}(T)).type||void 0!==T.seekingParent)throw{code:"S0217",token:T.type,position:T.position};return u.length>0&&(T.errors=u),T});t.exports=s},{"./signature":5}],5:[function(e,t,r){"use strict";function n(e){"@babel/helpers - typeof";return(n="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":typeof e})(e)}var a=e("./utils"),o=function(){var e={a:"arrays",b:"booleans",f:"functions",n:"numbers",o:"objects",s:"strings"};return function(t){for(var r=1,o=[],i={},s=i;r<t.length;){var u=t.charAt(r);if(":"===u)break;var c=function(){o.push(i),s=i,i={}},f=function(e,t,r,n){for(var a=1,o=t;o<e.length;)if(o++,(u=e.charAt(o))===n){if(0==--a)break}else u===r&&a++;return o};switch(u){case"s":case"n":case"b":case"l":case"o":i.regex="["+u+"m]",i.type=u,c();break;case"a":i.regex="[asnblfom]",i.type=u,i.array=!0,c();break;case"f":i.regex="f",i.type=u,c();break;case"j":i.regex="[asnblom]",i.type=u,c();break;case"x":i.regex="[asnblfom]",i.type=u,c();break;case"-":s.context=!0,s.contextRegex=new RegExp(s.regex),s.regex+="?";break;case"?":case"+":s.regex+=u;break;case"(":var l=f(t,r,"(",")"),p=t.substring(r+1,l);if(-1!==p.indexOf("<"))throw{code:"S0402",stack:(new Error).stack,value:p,offset:r};i.regex="["+p+"m]",i.type="("+p+")",r=l,c();break;case"<":if("a"!==s.type&&"f"!==s.type)throw{code:"S0401",stack:(new Error).stack,value:s.type,offset:r};var h=f(t,r,"<",">");s.subtype=t.substring(r+1,h),r=h}r++}var d="^"+o.map(function(e){return"("+e.regex+")"}).join("")+"$",v=new RegExp(d),g=function(e){var t;if(a.isFunction(e))t="f";else switch(n(e)){case"string":t="s";break;case"number":t="n";break;case"boolean":t="b";break;case"object":t=null===e?"l":Array.isArray(e)?"a":"o";break;case"undefined":default:t="m"}return t};return{definition:t,validate:function(t,r){var n="";t.forEach(function(e){n+=g(e)});var a=v.exec(n);if(a){var i=[],s=0;return o.forEach(function(n,o){var u=t[s],c=a[o+1];if(""===c)if(n.context&&n.contextRegex){var f=g(r);if(!n.contextRegex.test(f))throw{code:"T0411",stack:(new Error).stack,value:r,index:s+1};i.push(r)}else i.push(u),s++;else c.split("").forEach(function(r){if("a"===n.type){if("m"===r)u=void 0;else{u=t[s];var a=!0;if(void 0!==n.subtype)if("a"!==r&&c!==n.subtype)a=!1;else if("a"===r&&u.length>0){var o=g(u[0]);a=o===n.subtype.charAt(0)&&0===u.filter(function(e){return g(e)!==o}).length}if(!a)throw{code:"T0412",stack:(new Error).stack,value:u,index:s+1,type:e[n.subtype]};"a"!==r&&(u=[u])}i.push(u),s++}else i.push(u),s++})}),i}!function(e,t){for(var r="^",n=0,a=0;a<o.length;a++){r+=o[a].regex;var i=t.match(r);if(null===i)throw{code:"T0410",stack:(new Error).stack,value:e[n],index:n+1};n=i[0].length}throw{code:"T0410",stack:(new Error).stack,value:e[n],index:n+1}}(t,n)}}}}();t.exports=o},{"./utils":6}],6:[function(e,t,r){"use strict";function n(e,t){var r="undefined"!=typeof Symbol&&e[Symbol.iterator]||e["@@iterator"];if(!r){if(Array.isArray(e)||(r=function(e,t){if(!e)return;if("string"==typeof e)return a(e,t);var r=Object.prototype.toString.call(e).slice(8,-1);"Object"===r&&e.constructor&&(r=e.constructor.name);if("Map"===r||"Set"===r)return Array.from(e);if("Arguments"===r||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r))return a(e,t)}(e))||t&&e&&"number"==typeof e.length){r&&(e=r);var n=0,o=function(){};return{s:o,n:function(){return n>=e.length?{done:!0}:{done:!1,value:e[n++]}},e:function(e){throw e},f:o}}throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}var i,s=!0,u=!1;return{s:function(){r=r.call(e)},n:function(){var e=r.next();return s=e.done,e},e:function(e){u=!0,i=e},f:function(){try{s||null==r.return||r.return()}finally{if(u)throw i}}}}function a(e,t){(null==t||t>e.length)&&(t=e.length);for(var r=0,n=new Array(t);r<t;r++)n[r]=e[r];return n}function o(e){"@babel/helpers - typeof";return(o="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":typeof e})(e)}var i=function(){function e(e){var t=!1;if("number"==typeof e&&(t=!isNaN(e))&&!isFinite(e))throw{code:"D1001",value:e,stack:(new Error).stack};return t}var t=("function"==typeof Symbol?Symbol:{}).iterator||"@@iterator";return{isNumeric:e,isArrayOfStrings:function(e){var t=!1;return Array.isArray(e)&&(t=0===e.filter(function(e){return"string"!=typeof e}).length),t},isArrayOfNumbers:function(t){var r=!1;return Array.isArray(t)&&(r=0===t.filter(function(t){return!e(t)}).length),r},createSequence:function(){var e=[];return e.sequence=!0,1===arguments.length&&e.push(arguments[0]),e},isSequence:function(e){return!0===e.sequence&&Array.isArray(e)},isFunction:function(e){return e&&(!0===e._jsonata_function||!0===e._jsonata_lambda)||"function"==typeof e},isLambda:function(e){return e&&!0===e._jsonata_lambda},isIterable:function(e){return"object"===o(e)&&null!==e&&t in e&&"next"in e&&"function"==typeof e.next},getFunctionArity:function(e){return"number"==typeof e.arity?e.arity:"function"==typeof e.implementation?e.implementation.length:"number"==typeof e.length?e.length:e.arguments.length},isDeepEqual:function e(t,r){if(t===r)return!0;if("object"===o(t)&&"object"===o(r)&&null!==t&&null!==r){if(Array.isArray(t)&&Array.isArray(r)){if(t.length!==r.length)return!1;for(var n=0;n<t.length;n++)if(!e(t[n],r[n]))return!1;return!0}var a=Object.getOwnPropertyNames(t),i=Object.getOwnPropertyNames(r);if(a.length!==i.length)return!1;for(a=a.sort(),i=i.sort(),n=0;n<a.length;n++)if(a[n]!==i[n])return!1;for(n=0;n<a.length;n++){var s=a[n];if(!e(t[s],r[s]))return!1}return!0}return!1},stringToArray:function(e){var t,r=[],a=n(e);try{for(a.s();!(t=a.n()).done;){var o=t.value;r.push(o)}}catch(e){a.e(e)}finally{a.f()}return r}}}();t.exports=i},{}]},{},[3])(3)});;// jsonata-es5.min.js is prepended to this file as part of the Grunt build

;(function(window) {
    if (typeof window.window != "undefined" && window.document)
    return;
    if (window.require && window.define)
    return;

    if (!window.console) {
        window.console = function() {
            var msgs = Array.prototype.slice.call(arguments, 0);
            postMessage({type: "log", data: msgs});
        };
        window.console.error =
        window.console.warn =
        window.console.log =
        window.console.trace = window.console;
    }
    window.window = window;
    window.ace = window;
    window.onerror = function(message, file, line, col, err) {
        postMessage({type: "error", data: {
            message: message,
            data: err.data,
            file: file,
            line: line,
            col: col,
            stack: err.stack
        }});
    };

    window.normalizeModule = function(parentId, moduleName) {
        // normalize plugin requires
        if (moduleName.indexOf("!") !== -1) {
            var chunks = moduleName.split("!");
            return window.normalizeModule(parentId, chunks[0]) + "!" + window.normalizeModule(parentId, chunks[1]);
        }
        // normalize relative requires
        if (moduleName.charAt(0) == ".") {
            var base = parentId.split("/").slice(0, -1).join("/");
            moduleName = (base ? base + "/" : "") + moduleName;

            while (moduleName.indexOf(".") !== -1 && previous != moduleName) {
                var previous = moduleName;
                moduleName = moduleName.replace(/^\.\//, "").replace(/\/\.\//, "/").replace(/[^\/]+\/\.\.\//, "");
            }
        }

        return moduleName;
    };

    window.require = function require(parentId, id) {
        if (!id) {
            id = parentId;
            parentId = null;
        }
        if (!id.charAt)
        throw new Error("worker.js require() accepts only (parentId, id) as arguments");

        id = window.normalizeModule(parentId, id);

        var module = window.require.modules[id];
        if (module) {
            if (!module.initialized) {
                module.initialized = true;
                module.exports = module.factory().exports;
            }
            return module.exports;
        }

        if (!window.require.tlns)
        return console.log("unable to load " + id);

        var path = resolveModuleId(id, window.require.tlns);
        if (path.slice(-3) != ".js") path += ".js";

        window.require.id = id;
        window.require.modules[id] = {}; // prevent infinite loop on broken modules
        importScripts(path);
        return window.require(parentId, id);
    };
    function resolveModuleId(id, paths) {
        var testPath = id, tail = "";
        while (testPath) {
            var alias = paths[testPath];
            if (typeof alias == "string") {
                return alias + tail;
            } else if (alias) {
                return  alias.location.replace(/\/*$/, "/") + (tail || alias.main || alias.name);
            } else if (alias === false) {
                return "";
            }
            var i = testPath.lastIndexOf("/");
            if (i === -1) break;
            tail = testPath.substr(i) + tail;
            testPath = testPath.slice(0, i);
        }
        return id;
    }
    window.require.modules = {};
    window.require.tlns = {};

    window.define = function(id, deps, factory) {
        if (arguments.length == 2) {
            factory = deps;
            if (typeof id != "string") {
                deps = id;
                id = window.require.id;
            }
        } else if (arguments.length == 1) {
            factory = id;
            deps = [];
            id = window.require.id;
        }

        if (typeof factory != "function") {
            window.require.modules[id] = {
                exports: factory,
                initialized: true
            };
            return;
        }

        if (!deps.length)
        // If there is no dependencies, we inject "require", "exports" and
        // "module" as dependencies, to provide CommonJS compatibility.
        deps = ["require", "exports", "module"];

        var req = function(childId) {
            return window.require(id, childId);
        };

        window.require.modules[id] = {
            exports: {},
            factory: function() {
                var module = this;
                var returnExports = factory.apply(this, deps.map(function(dep) {
                    switch (dep) {
                        // Because "require", "exports" and "module" aren't actual
                        // dependencies, we must handle them seperately.
                        case "require": return req;
                        case "exports": return module.exports;
                        case "module":  return module;
                        // But for all other dependencies, we can just go ahead and
                        // require them.
                        default:        return req(dep);
                    }
                }));
                if (returnExports)
                module.exports = returnExports;
                return module;
            }
        };
    };
    window.define.amd = {};
    require.tlns = {};
    window.initBaseUrls  = function initBaseUrls(topLevelNamespaces) {
        for (var i in topLevelNamespaces)
        require.tlns[i] = topLevelNamespaces[i];
    };

    window.initSender = function initSender() {

        var EventEmitter = window.require("ace/lib/event_emitter").EventEmitter;
        var oop = window.require("ace/lib/oop");

        var Sender = function() {};

        (function() {

            oop.implement(this, EventEmitter);

            this.callback = function(data, callbackId) {
                postMessage({
                    type: "call",
                    id: callbackId,
                    data: data
                });
            };

            this.emit = function(name, data) {
                postMessage({
                    type: "event",
                    name: name,
                    data: data
                });
            };

        }).call(Sender.prototype);

        return new Sender();
    };

    var main = window.main = null;
    var sender = window.sender = null;

    window.onmessage = function(e) {
        var msg = e.data;
        if (msg.event && sender) {
            sender._signal(msg.event, msg.data);
        }
        else if (msg.command) {
            if (main[msg.command])
            main[msg.command].apply(main, msg.args);
            else if (window[msg.command])
            window[msg.command].apply(window, msg.args);
            else
            throw new Error("Unknown command:" + msg.command);
        }
        else if (msg.init) {
            window.initBaseUrls(msg.tlns);
            require("ace/lib/es5-shim");
            sender = window.sender = window.initSender();
            var clazz = require(msg.module)[msg.classname];
            main = window.main = new clazz(sender);
        }
    };
})(this);

define("ace/lib/oop",["require","exports","module"], function(require, exports, module) {
    "use strict";

    exports.inherits = function(ctor, superCtor) {
        ctor.super_ = superCtor;
        ctor.prototype = Object.create(superCtor.prototype, {
            constructor: {
                value: ctor,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
    };

    exports.mixin = function(obj, mixin) {
        for (var key in mixin) {
            obj[key] = mixin[key];
        }
        return obj;
    };

    exports.implement = function(proto, mixin) {
        exports.mixin(proto, mixin);
    };

});

define("ace/range",["require","exports","module"], function(require, exports, module) {
    "use strict";
    var comparePoints = function(p1, p2) {
        return p1.row - p2.row || p1.column - p2.column;
    };
    var Range = function(startRow, startColumn, endRow, endColumn) {
        this.start = {
            row: startRow,
            column: startColumn
        };

        this.end = {
            row: endRow,
            column: endColumn
        };
    };

    (function() {
        this.isEqual = function(range) {
            return this.start.row === range.start.row &&
            this.end.row === range.end.row &&
            this.start.column === range.start.column &&
            this.end.column === range.end.column;
        };
        this.toString = function() {
            return ("Range: [" + this.start.row + "/" + this.start.column +
            "] -> [" + this.end.row + "/" + this.end.column + "]");
        };

        this.contains = function(row, column) {
            return this.compare(row, column) == 0;
        };
        this.compareRange = function(range) {
            var cmp,
            end = range.end,
            start = range.start;

            cmp = this.compare(end.row, end.column);
            if (cmp == 1) {
                cmp = this.compare(start.row, start.column);
                if (cmp == 1) {
                    return 2;
                } else if (cmp == 0) {
                    return 1;
                } else {
                    return 0;
                }
            } else if (cmp == -1) {
                return -2;
            } else {
                cmp = this.compare(start.row, start.column);
                if (cmp == -1) {
                    return -1;
                } else if (cmp == 1) {
                    return 42;
                } else {
                    return 0;
                }
            }
        };
        this.comparePoint = function(p) {
            return this.compare(p.row, p.column);
        };
        this.containsRange = function(range) {
            return this.comparePoint(range.start) == 0 && this.comparePoint(range.end) == 0;
        };
        this.intersects = function(range) {
            var cmp = this.compareRange(range);
            return (cmp == -1 || cmp == 0 || cmp == 1);
        };
        this.isEnd = function(row, column) {
            return this.end.row == row && this.end.column == column;
        };
        this.isStart = function(row, column) {
            return this.start.row == row && this.start.column == column;
        };
        this.setStart = function(row, column) {
            if (typeof row == "object") {
                this.start.column = row.column;
                this.start.row = row.row;
            } else {
                this.start.row = row;
                this.start.column = column;
            }
        };
        this.setEnd = function(row, column) {
            if (typeof row == "object") {
                this.end.column = row.column;
                this.end.row = row.row;
            } else {
                this.end.row = row;
                this.end.column = column;
            }
        };
        this.inside = function(row, column) {
            if (this.compare(row, column) == 0) {
                if (this.isEnd(row, column) || this.isStart(row, column)) {
                    return false;
                } else {
                    return true;
                }
            }
            return false;
        };
        this.insideStart = function(row, column) {
            if (this.compare(row, column) == 0) {
                if (this.isEnd(row, column)) {
                    return false;
                } else {
                    return true;
                }
            }
            return false;
        };
        this.insideEnd = function(row, column) {
            if (this.compare(row, column) == 0) {
                if (this.isStart(row, column)) {
                    return false;
                } else {
                    return true;
                }
            }
            return false;
        };
        this.compare = function(row, column) {
            if (!this.isMultiLine()) {
                if (row === this.start.row) {
                    return column < this.start.column ? -1 : (column > this.end.column ? 1 : 0);
                }
            }

            if (row < this.start.row)
            return -1;

            if (row > this.end.row)
            return 1;

            if (this.start.row === row)
            return column >= this.start.column ? 0 : -1;

            if (this.end.row === row)
            return column <= this.end.column ? 0 : 1;

            return 0;
        };
        this.compareStart = function(row, column) {
            if (this.start.row == row && this.start.column == column) {
                return -1;
            } else {
                return this.compare(row, column);
            }
        };
        this.compareEnd = function(row, column) {
            if (this.end.row == row && this.end.column == column) {
                return 1;
            } else {
                return this.compare(row, column);
            }
        };
        this.compareInside = function(row, column) {
            if (this.end.row == row && this.end.column == column) {
                return 1;
            } else if (this.start.row == row && this.start.column == column) {
                return -1;
            } else {
                return this.compare(row, column);
            }
        };
        this.clipRows = function(firstRow, lastRow) {
            if (this.end.row > lastRow)
            var end = {row: lastRow + 1, column: 0};
            else if (this.end.row < firstRow)
            var end = {row: firstRow, column: 0};

            if (this.start.row > lastRow)
            var start = {row: lastRow + 1, column: 0};
            else if (this.start.row < firstRow)
            var start = {row: firstRow, column: 0};

            return Range.fromPoints(start || this.start, end || this.end);
        };
        this.extend = function(row, column) {
            var cmp = this.compare(row, column);

            if (cmp == 0)
            return this;
            else if (cmp == -1)
            var start = {row: row, column: column};
            else
            var end = {row: row, column: column};

            return Range.fromPoints(start || this.start, end || this.end);
        };

        this.isEmpty = function() {
            return (this.start.row === this.end.row && this.start.column === this.end.column);
        };
        this.isMultiLine = function() {
            return (this.start.row !== this.end.row);
        };
        this.clone = function() {
            return Range.fromPoints(this.start, this.end);
        };
        this.collapseRows = function() {
            if (this.end.column == 0)
            return new Range(this.start.row, 0, Math.max(this.start.row, this.end.row-1), 0)
            else
            return new Range(this.start.row, 0, this.end.row, 0)
        };
        this.toScreenRange = function(session) {
            var screenPosStart = session.documentToScreenPosition(this.start);
            var screenPosEnd = session.documentToScreenPosition(this.end);

            return new Range(
                screenPosStart.row, screenPosStart.column,
                screenPosEnd.row, screenPosEnd.column
            );
        };
        this.moveBy = function(row, column) {
            this.start.row += row;
            this.start.column += column;
            this.end.row += row;
            this.end.column += column;
        };

    }).call(Range.prototype);
    Range.fromPoints = function(start, end) {
        return new Range(start.row, start.column, end.row, end.column);
    };
    Range.comparePoints = comparePoints;

    Range.comparePoints = function(p1, p2) {
        return p1.row - p2.row || p1.column - p2.column;
    };


    exports.Range = Range;
});

define("ace/apply_delta",["require","exports","module"], function(require, exports, module) {
    "use strict";

    function throwDeltaError(delta, errorText){
        console.log("Invalid Delta:", delta);
        throw "Invalid Delta: " + errorText;
    }

    function positionInDocument(docLines, position) {
        return position.row    >= 0 && position.row    <  docLines.length &&
        position.column >= 0 && position.column <= docLines[position.row].length;
    }

    function validateDelta(docLines, delta) {
        if (delta.action != "insert" && delta.action != "remove")
        throwDeltaError(delta, "delta.action must be 'insert' or 'remove'");
        if (!(delta.lines instanceof Array))
        throwDeltaError(delta, "delta.lines must be an Array");
        if (!delta.start || !delta.end)
        throwDeltaError(delta, "delta.start/end must be an present");
        var start = delta.start;
        if (!positionInDocument(docLines, delta.start))
        throwDeltaError(delta, "delta.start must be contained in document");
        var end = delta.end;
        if (delta.action == "remove" && !positionInDocument(docLines, end))
        throwDeltaError(delta, "delta.end must contained in document for 'remove' actions");
        var numRangeRows = end.row - start.row;
        var numRangeLastLineChars = (end.column - (numRangeRows == 0 ? start.column : 0));
        if (numRangeRows != delta.lines.length - 1 || delta.lines[numRangeRows].length != numRangeLastLineChars)
        throwDeltaError(delta, "delta.range must match delta lines");
    }

    exports.applyDelta = function(docLines, delta, doNotValidate) {

        var row = delta.start.row;
        var startColumn = delta.start.column;
        var line = docLines[row] || "";
        switch (delta.action) {
            case "insert":
            var lines = delta.lines;
            if (lines.length === 1) {
                docLines[row] = line.substring(0, startColumn) + delta.lines[0] + line.substring(startColumn);
            } else {
                var args = [row, 1].concat(delta.lines);
                docLines.splice.apply(docLines, args);
                docLines[row] = line.substring(0, startColumn) + docLines[row];
                docLines[row + delta.lines.length - 1] += line.substring(startColumn);
            }
            break;
            case "remove":
            var endColumn = delta.end.column;
            var endRow = delta.end.row;
            if (row === endRow) {
                docLines[row] = line.substring(0, startColumn) + line.substring(endColumn);
            } else {
                docLines.splice(
                    row, endRow - row + 1,
                    line.substring(0, startColumn) + docLines[endRow].substring(endColumn)
                );
            }
            break;
        }
    }
});

define("ace/lib/event_emitter",["require","exports","module"], function(require, exports, module) {
    "use strict";

    var EventEmitter = {};
    var stopPropagation = function() { this.propagationStopped = true; };
    var preventDefault = function() { this.defaultPrevented = true; };

    EventEmitter._emit =
    EventEmitter._dispatchEvent = function(eventName, e) {
        this._eventRegistry || (this._eventRegistry = {});
        this._defaultHandlers || (this._defaultHandlers = {});

        var listeners = this._eventRegistry[eventName] || [];
        var defaultHandler = this._defaultHandlers[eventName];
        if (!listeners.length && !defaultHandler)
        return;

        if (typeof e != "object" || !e)
        e = {};

        if (!e.type)
        e.type = eventName;
        if (!e.stopPropagation)
        e.stopPropagation = stopPropagation;
        if (!e.preventDefault)
        e.preventDefault = preventDefault;

        listeners = listeners.slice();
        for (var i=0; i<listeners.length; i++) {
            listeners[i](e, this);
            if (e.propagationStopped)
            break;
        }

        if (defaultHandler && !e.defaultPrevented)
        return defaultHandler(e, this);
    };


    EventEmitter._signal = function(eventName, e) {
        var listeners = (this._eventRegistry || {})[eventName];
        if (!listeners)
        return;
        listeners = listeners.slice();
        for (var i=0; i<listeners.length; i++)
        listeners[i](e, this);
    };

    EventEmitter.once = function(eventName, callback) {
        var _self = this;
        callback && this.addEventListener(eventName, function newCallback() {
            _self.removeEventListener(eventName, newCallback);
            callback.apply(null, arguments);
        });
    };


    EventEmitter.setDefaultHandler = function(eventName, callback) {
        var handlers = this._defaultHandlers
        if (!handlers)
        handlers = this._defaultHandlers = {_disabled_: {}};

        if (handlers[eventName]) {
            var old = handlers[eventName];
            var disabled = handlers._disabled_[eventName];
            if (!disabled)
            handlers._disabled_[eventName] = disabled = [];
            disabled.push(old);
            var i = disabled.indexOf(callback);
            if (i != -1)
            disabled.splice(i, 1);
        }
        handlers[eventName] = callback;
    };
    EventEmitter.removeDefaultHandler = function(eventName, callback) {
        var handlers = this._defaultHandlers
        if (!handlers)
        return;
        var disabled = handlers._disabled_[eventName];

        if (handlers[eventName] == callback) {
            var old = handlers[eventName];
            if (disabled)
            this.setDefaultHandler(eventName, disabled.pop());
        } else if (disabled) {
            var i = disabled.indexOf(callback);
            if (i != -1)
            disabled.splice(i, 1);
        }
    };

    EventEmitter.on =
    EventEmitter.addEventListener = function(eventName, callback, capturing) {
        this._eventRegistry = this._eventRegistry || {};

        var listeners = this._eventRegistry[eventName];
        if (!listeners)
        listeners = this._eventRegistry[eventName] = [];

        if (listeners.indexOf(callback) == -1)
        listeners[capturing ? "unshift" : "push"](callback);
        return callback;
    };

    EventEmitter.off =
    EventEmitter.removeListener =
    EventEmitter.removeEventListener = function(eventName, callback) {
        this._eventRegistry = this._eventRegistry || {};

        var listeners = this._eventRegistry[eventName];
        if (!listeners)
        return;

        var index = listeners.indexOf(callback);
        if (index !== -1)
        listeners.splice(index, 1);
    };

    EventEmitter.removeAllListeners = function(eventName) {
        if (this._eventRegistry) this._eventRegistry[eventName] = [];
    };

    exports.EventEmitter = EventEmitter;

});

define("ace/anchor",["require","exports","module","ace/lib/oop","ace/lib/event_emitter"], function(require, exports, module) {
    "use strict";

    var oop = require("./lib/oop");
    var EventEmitter = require("./lib/event_emitter").EventEmitter;

    var Anchor = exports.Anchor = function(doc, row, column) {
        this.$onChange = this.onChange.bind(this);
        this.attach(doc);

        if (typeof column == "undefined")
        this.setPosition(row.row, row.column);
        else
        this.setPosition(row, column);
    };

    (function() {

        oop.implement(this, EventEmitter);
        this.getPosition = function() {
            return this.$clipPositionToDocument(this.row, this.column);
        };
        this.getDocument = function() {
            return this.document;
        };
        this.$insertRight = false;
        this.onChange = function(delta) {
            if (delta.start.row == delta.end.row && delta.start.row != this.row)
            return;

            if (delta.start.row > this.row)
            return;

            var point = $getTransformedPoint(delta, {row: this.row, column: this.column}, this.$insertRight);
            this.setPosition(point.row, point.column, true);
        };

        function $pointsInOrder(point1, point2, equalPointsInOrder) {
            var bColIsAfter = equalPointsInOrder ? point1.column <= point2.column : point1.column < point2.column;
            return (point1.row < point2.row) || (point1.row == point2.row && bColIsAfter);
        }

        function $getTransformedPoint(delta, point, moveIfEqual) {
            var deltaIsInsert = delta.action == "insert";
            var deltaRowShift = (deltaIsInsert ? 1 : -1) * (delta.end.row    - delta.start.row);
            var deltaColShift = (deltaIsInsert ? 1 : -1) * (delta.end.column - delta.start.column);
            var deltaStart = delta.start;
            var deltaEnd = deltaIsInsert ? deltaStart : delta.end; // Collapse insert range.
            if ($pointsInOrder(point, deltaStart, moveIfEqual)) {
                return {
                    row: point.row,
                    column: point.column
                };
            }
            if ($pointsInOrder(deltaEnd, point, !moveIfEqual)) {
                return {
                    row: point.row + deltaRowShift,
                    column: point.column + (point.row == deltaEnd.row ? deltaColShift : 0)
                };
            }

            return {
                row: deltaStart.row,
                column: deltaStart.column
            };
        }
        this.setPosition = function(row, column, noClip) {
            var pos;
            if (noClip) {
                pos = {
                    row: row,
                    column: column
                };
            } else {
                pos = this.$clipPositionToDocument(row, column);
            }

            if (this.row == pos.row && this.column == pos.column)
            return;

            var old = {
                row: this.row,
                column: this.column
            };

            this.row = pos.row;
            this.column = pos.column;
            this._signal("change", {
                old: old,
                value: pos
            });
        };
        this.detach = function() {
            this.document.removeEventListener("change", this.$onChange);
        };
        this.attach = function(doc) {
            this.document = doc || this.document;
            this.document.on("change", this.$onChange);
        };
        this.$clipPositionToDocument = function(row, column) {
            var pos = {};

            if (row >= this.document.getLength()) {
                pos.row = Math.max(0, this.document.getLength() - 1);
                pos.column = this.document.getLine(pos.row).length;
            }
            else if (row < 0) {
                pos.row = 0;
                pos.column = 0;
            }
            else {
                pos.row = row;
                pos.column = Math.min(this.document.getLine(pos.row).length, Math.max(0, column));
            }

            if (column < 0)
            pos.column = 0;

            return pos;
        };

    }).call(Anchor.prototype);

});

define("ace/document",["require","exports","module","ace/lib/oop","ace/apply_delta","ace/lib/event_emitter","ace/range","ace/anchor"], function(require, exports, module) {
    "use strict";
    var oop = require("./lib/oop");
    var applyDelta = require("./apply_delta").applyDelta;
    var EventEmitter = require("./lib/event_emitter").EventEmitter;
    var Range = require("./range").Range;
    var Anchor = require("./anchor").Anchor;

    var Document = function(textOrLines) {
        this.$lines = [""];
        if (textOrLines.length === 0) {
            this.$lines = [""];
        } else if (Array.isArray(textOrLines)) {
            this.insertMergedLines({row: 0, column: 0}, textOrLines);
        } else {
            this.insert({row: 0, column:0}, textOrLines);
        }
    };

    (function() {

        oop.implement(this, EventEmitter);
        this.setValue = function(text) {
            var len = this.getLength() - 1;
            this.remove(new Range(0, 0, len, this.getLine(len).length));
            this.insert({row: 0, column: 0}, text);
        };
        this.getValue = function() {
            return this.getAllLines().join(this.getNewLineCharacter());
        };
        this.createAnchor = function(row, column) {
            return new Anchor(this, row, column);
        };
        if ("aaa".split(/a/).length === 0) {
            this.$split = function(text) {
                return text.replace(/\r\n|\r/g, "\n").split("\n");
            };
        } else {
            this.$split = function(text) {
                return text.split(/\r\n|\r|\n/);
            };
        }


        this.$detectNewLine = function(text) {
            var match = text.match(/^.*?(\r\n|\r|\n)/m);
            this.$autoNewLine = match ? match[1] : "\n";
            this._signal("changeNewLineMode");
        };
        this.getNewLineCharacter = function() {
            switch (this.$newLineMode) {
                case "windows":
                return "\r\n";
                case "unix":
                return "\n";
                default:
                return this.$autoNewLine || "\n";
            }
        };

        this.$autoNewLine = "";
        this.$newLineMode = "auto";
        this.setNewLineMode = function(newLineMode) {
            if (this.$newLineMode === newLineMode)
            return;

            this.$newLineMode = newLineMode;
            this._signal("changeNewLineMode");
        };
        this.getNewLineMode = function() {
            return this.$newLineMode;
        };
        this.isNewLine = function(text) {
            return (text == "\r\n" || text == "\r" || text == "\n");
        };
        this.getLine = function(row) {
            return this.$lines[row] || "";
        };
        this.getLines = function(firstRow, lastRow) {
            return this.$lines.slice(firstRow, lastRow + 1);
        };
        this.getAllLines = function() {
            return this.getLines(0, this.getLength());
        };
        this.getLength = function() {
            return this.$lines.length;
        };
        this.getTextRange = function(range) {
            return this.getLinesForRange(range).join(this.getNewLineCharacter());
        };
        this.getLinesForRange = function(range) {
            var lines;
            if (range.start.row === range.end.row) {
                lines = [this.getLine(range.start.row).substring(range.start.column, range.end.column)];
            } else {
                lines = this.getLines(range.start.row, range.end.row);
                lines[0] = (lines[0] || "").substring(range.start.column);
                var l = lines.length - 1;
                if (range.end.row - range.start.row == l)
                lines[l] = lines[l].substring(0, range.end.column);
            }
            return lines;
        };
        this.insertLines = function(row, lines) {
            console.warn("Use of document.insertLines is deprecated. Use the insertFullLines method instead.");
            return this.insertFullLines(row, lines);
        };
        this.removeLines = function(firstRow, lastRow) {
            console.warn("Use of document.removeLines is deprecated. Use the removeFullLines method instead.");
            return this.removeFullLines(firstRow, lastRow);
        };
        this.insertNewLine = function(position) {
            console.warn("Use of document.insertNewLine is deprecated. Use insertMergedLines(position, ['', '']) instead.");
            return this.insertMergedLines(position, ["", ""]);
        };
        this.insert = function(position, text) {
            if (this.getLength() <= 1)
            this.$detectNewLine(text);

            return this.insertMergedLines(position, this.$split(text));
        };
        this.insertInLine = function(position, text) {
            var start = this.clippedPos(position.row, position.column);
            var end = this.pos(position.row, position.column + text.length);

            this.applyDelta({
                start: start,
                end: end,
                action: "insert",
                lines: [text]
            }, true);

            return this.clonePos(end);
        };

        this.clippedPos = function(row, column) {
            var length = this.getLength();
            if (row === undefined) {
                row = length;
            } else if (row < 0) {
                row = 0;
            } else if (row >= length) {
                row = length - 1;
                column = undefined;
            }
            var line = this.getLine(row);
            if (column == undefined)
            column = line.length;
            column = Math.min(Math.max(column, 0), line.length);
            return {row: row, column: column};
        };

        this.clonePos = function(pos) {
            return {row: pos.row, column: pos.column};
        };

        this.pos = function(row, column) {
            return {row: row, column: column};
        };

        this.$clipPosition = function(position) {
            var length = this.getLength();
            if (position.row >= length) {
                position.row = Math.max(0, length - 1);
                position.column = this.getLine(length - 1).length;
            } else {
                position.row = Math.max(0, position.row);
                position.column = Math.min(Math.max(position.column, 0), this.getLine(position.row).length);
            }
            return position;
        };
        this.insertFullLines = function(row, lines) {
            row = Math.min(Math.max(row, 0), this.getLength());
            var column = 0;
            if (row < this.getLength()) {
                lines = lines.concat([""]);
                column = 0;
            } else {
                lines = [""].concat(lines);
                row--;
                column = this.$lines[row].length;
            }
            this.insertMergedLines({row: row, column: column}, lines);
        };
        this.insertMergedLines = function(position, lines) {
            var start = this.clippedPos(position.row, position.column);
            var end = {
                row: start.row + lines.length - 1,
                column: (lines.length == 1 ? start.column : 0) + lines[lines.length - 1].length
            };

            this.applyDelta({
                start: start,
                end: end,
                action: "insert",
                lines: lines
            });

            return this.clonePos(end);
        };
        this.remove = function(range) {
            var start = this.clippedPos(range.start.row, range.start.column);
            var end = this.clippedPos(range.end.row, range.end.column);
            this.applyDelta({
                start: start,
                end: end,
                action: "remove",
                lines: this.getLinesForRange({start: start, end: end})
            });
            return this.clonePos(start);
        };
        this.removeInLine = function(row, startColumn, endColumn) {
            var start = this.clippedPos(row, startColumn);
            var end = this.clippedPos(row, endColumn);

            this.applyDelta({
                start: start,
                end: end,
                action: "remove",
                lines: this.getLinesForRange({start: start, end: end})
            }, true);

            return this.clonePos(start);
        };
        this.removeFullLines = function(firstRow, lastRow) {
            firstRow = Math.min(Math.max(0, firstRow), this.getLength() - 1);
            lastRow  = Math.min(Math.max(0, lastRow ), this.getLength() - 1);
            var deleteFirstNewLine = lastRow == this.getLength() - 1 && firstRow > 0;
            var deleteLastNewLine  = lastRow  < this.getLength() - 1;
            var startRow = ( deleteFirstNewLine ? firstRow - 1                  : firstRow                    );
            var startCol = ( deleteFirstNewLine ? this.getLine(startRow).length : 0                           );
            var endRow   = ( deleteLastNewLine  ? lastRow + 1                   : lastRow                     );
            var endCol   = ( deleteLastNewLine  ? 0                             : this.getLine(endRow).length );
            var range = new Range(startRow, startCol, endRow, endCol);
            var deletedLines = this.$lines.slice(firstRow, lastRow + 1);

            this.applyDelta({
                start: range.start,
                end: range.end,
                action: "remove",
                lines: this.getLinesForRange(range)
            });
            return deletedLines;
        };
        this.removeNewLine = function(row) {
            if (row < this.getLength() - 1 && row >= 0) {
                this.applyDelta({
                    start: this.pos(row, this.getLine(row).length),
                    end: this.pos(row + 1, 0),
                    action: "remove",
                    lines: ["", ""]
                });
            }
        };
        this.replace = function(range, text) {
            if (!(range instanceof Range))
            range = Range.fromPoints(range.start, range.end);
            if (text.length === 0 && range.isEmpty())
            return range.start;
            if (text == this.getTextRange(range))
            return range.end;

            this.remove(range);
            var end;
            if (text) {
                end = this.insert(range.start, text);
            }
            else {
                end = range.start;
            }

            return end;
        };
        this.applyDeltas = function(deltas) {
            for (var i=0; i<deltas.length; i++) {
                this.applyDelta(deltas[i]);
            }
        };
        this.revertDeltas = function(deltas) {
            for (var i=deltas.length-1; i>=0; i--) {
                this.revertDelta(deltas[i]);
            }
        };
        this.applyDelta = function(delta, doNotValidate) {
            var isInsert = delta.action == "insert";
            if (isInsert ? delta.lines.length <= 1 && !delta.lines[0]
                : !Range.comparePoints(delta.start, delta.end)) {
                    return;
                }

                if (isInsert && delta.lines.length > 20000)
                this.$splitAndapplyLargeDelta(delta, 20000);
                applyDelta(this.$lines, delta, doNotValidate);
                this._signal("change", delta);
            };

            this.$splitAndapplyLargeDelta = function(delta, MAX) {
                var lines = delta.lines;
                var l = lines.length;
                var row = delta.start.row;
                var column = delta.start.column;
                var from = 0, to = 0;
                do {
                    from = to;
                    to += MAX - 1;
                    var chunk = lines.slice(from, to);
                    if (to > l) {
                        delta.lines = chunk;
                        delta.start.row = row + from;
                        delta.start.column = column;
                        break;
                    }
                    chunk.push("");
                    this.applyDelta({
                        start: this.pos(row + from, column),
                        end: this.pos(row + to, column = 0),
                        action: delta.action,
                        lines: chunk
                    }, true);
                } while(true);
            };
            this.revertDelta = function(delta) {
                this.applyDelta({
                    start: this.clonePos(delta.start),
                    end: this.clonePos(delta.end),
                    action: (delta.action == "insert" ? "remove" : "insert"),
                    lines: delta.lines.slice()
                });
            };
            this.indexToPosition = function(index, startRow) {
                var lines = this.$lines || this.getAllLines();
                var newlineLength = this.getNewLineCharacter().length;
                for (var i = startRow || 0, l = lines.length; i < l; i++) {
                    index -= lines[i].length + newlineLength;
                    if (index < 0)
                    return {row: i, column: index + lines[i].length + newlineLength};
                }
                return {row: l-1, column: lines[l-1].length};
            };
            this.positionToIndex = function(pos, startRow) {
                var lines = this.$lines || this.getAllLines();
                var newlineLength = this.getNewLineCharacter().length;
                var index = 0;
                var row = Math.min(pos.row, lines.length);
                for (var i = startRow || 0; i < row; ++i)
                index += lines[i].length + newlineLength;

                return index + pos.column;
            };

        }).call(Document.prototype);

        exports.Document = Document;
    });

    define("ace/lib/lang",["require","exports","module"], function(require, exports, module) {
        "use strict";

        exports.last = function(a) {
            return a[a.length - 1];
        };

        exports.stringReverse = function(string) {
            return string.split("").reverse().join("");
        };

        exports.stringRepeat = function (string, count) {
            var result = '';
            while (count > 0) {
                if (count & 1)
                result += string;

                if (count >>= 1)
                string += string;
            }
            return result;
        };

        var trimBeginRegexp = /^\s\s*/;
        var trimEndRegexp = /\s\s*$/;

        exports.stringTrimLeft = function (string) {
            return string.replace(trimBeginRegexp, '');
        };

        exports.stringTrimRight = function (string) {
            return string.replace(trimEndRegexp, '');
        };

        exports.copyObject = function(obj) {
            var copy = {};
            for (var key in obj) {
                copy[key] = obj[key];
            }
            return copy;
        };

        exports.copyArray = function(array){
            var copy = [];
            for (var i=0, l=array.length; i<l; i++) {
                if (array[i] && typeof array[i] == "object")
                copy[i] = this.copyObject(array[i]);
                else
                copy[i] = array[i];
            }
            return copy;
        };

        exports.deepCopy = function deepCopy(obj) {
            if (typeof obj !== "object" || !obj)
            return obj;
            var copy;
            if (Array.isArray(obj)) {
                copy = [];
                for (var key = 0; key < obj.length; key++) {
                    copy[key] = deepCopy(obj[key]);
                }
                return copy;
            }
            if (Object.prototype.toString.call(obj) !== "[object Object]")
            return obj;

            copy = {};
            for (var key in obj)
            copy[key] = deepCopy(obj[key]);
            return copy;
        };

        exports.arrayToMap = function(arr) {
            var map = {};
            for (var i=0; i<arr.length; i++) {
                map[arr[i]] = 1;
            }
            return map;

        };

        exports.createMap = function(props) {
            var map = Object.create(null);
            for (var i in props) {
                map[i] = props[i];
            }
            return map;
        };
        exports.arrayRemove = function(array, value) {
            for (var i = 0; i <= array.length; i++) {
                if (value === array[i]) {
                    array.splice(i, 1);
                }
            }
        };

        exports.escapeRegExp = function(str) {
            return str.replace(/([.*+?^${}()|[\]\/\\])/g, '\\$1');
        };

        exports.escapeHTML = function(str) {
            return str.replace(/&/g, "&#38;").replace(/"/g, "&#34;").replace(/'/g, "&#39;").replace(/</g, "&#60;");
        };

        exports.getMatchOffsets = function(string, regExp) {
            var matches = [];

            string.replace(regExp, function(str) {
                matches.push({
                    offset: arguments[arguments.length-2],
                    length: str.length
                });
            });

            return matches;
        };
        exports.deferredCall = function(fcn) {
            var timer = null;
            var callback = function() {
                timer = null;
                fcn();
            };

            var deferred = function(timeout) {
                deferred.cancel();
                timer = setTimeout(callback, timeout || 0);
                return deferred;
            };

            deferred.schedule = deferred;

            deferred.call = function() {
                this.cancel();
                fcn();
                return deferred;
            };

            deferred.cancel = function() {
                clearTimeout(timer);
                timer = null;
                return deferred;
            };

            deferred.isPending = function() {
                return timer;
            };

            return deferred;
        };


        exports.delayedCall = function(fcn, defaultTimeout) {
            var timer = null;
            var callback = function() {
                timer = null;
                fcn();
            };

            var _self = function(timeout) {
                if (timer == null)
                timer = setTimeout(callback, timeout || defaultTimeout);
            };

            _self.delay = function(timeout) {
                timer && clearTimeout(timer);
                timer = setTimeout(callback, timeout || defaultTimeout);
            };
            _self.schedule = _self;

            _self.call = function() {
                this.cancel();
                fcn();
            };

            _self.cancel = function() {
                timer && clearTimeout(timer);
                timer = null;
            };

            _self.isPending = function() {
                return timer;
            };

            return _self;
        };
    });

    define("ace/worker/mirror",["require","exports","module","ace/range","ace/document","ace/lib/lang"], function(require, exports, module) {
        "use strict";

        var Range = require("../range").Range;
        var Document = require("../document").Document;
        var lang = require("../lib/lang");

        var Mirror = exports.Mirror = function(sender) {
            this.sender = sender;
            var doc = this.doc = new Document("");

            var deferredUpdate = this.deferredUpdate = lang.delayedCall(this.onUpdate.bind(this));

            var _self = this;
            sender.on("change", function(e) {
                var data = e.data;
                if (data[0].start) {
                    doc.applyDeltas(data);
                } else {
                    for (var i = 0; i < data.length; i += 2) {
                        if (Array.isArray(data[i+1])) {
                            var d = {action: "insert", start: data[i], lines: data[i+1]};
                        } else {
                            var d = {action: "remove", start: data[i], end: data[i+1]};
                        }
                        doc.applyDelta(d, true);
                    }
                }
                if (_self.$timeout)
                return deferredUpdate.schedule(_self.$timeout);
                _self.onUpdate();
            });
        };

        (function() {

            this.$timeout = 500;

            this.setTimeout = function(timeout) {
                this.$timeout = timeout;
            };

            this.setValue = function(value) {
                this.doc.setValue(value);
                this.deferredUpdate.schedule(this.$timeout);
            };

            this.getValue = function(callbackId) {
                this.sender.callback(this.doc.getValue(), callbackId);
            };

            this.onUpdate = function() {
            };

            this.isPending = function() {
                return this.deferredUpdate.isPending();
            };

        }).call(Mirror.prototype);

    });

    define("ace/lib/es5-shim",["require","exports","module"], function(require, exports, module) {

        function Empty() {}

        if (!Function.prototype.bind) {
            Function.prototype.bind = function bind(that) { // .length is 1
                var target = this;
                if (typeof target != "function") {
                    throw new TypeError("Function.prototype.bind called on incompatible " + target);
                }
                var args = slice.call(arguments, 1); // for normal call
                var bound = function () {

                    if (this instanceof bound) {

                        var result = target.apply(
                            this,
                            args.concat(slice.call(arguments))
                        );
                        if (Object(result) === result) {
                            return result;
                        }
                        return this;

                    } else {
                        return target.apply(
                            that,
                            args.concat(slice.call(arguments))
                        );

                    }

                };
                if(target.prototype) {
                    Empty.prototype = target.prototype;
                    bound.prototype = new Empty();
                    Empty.prototype = null;
                }
                return bound;
            };
        }
        var call = Function.prototype.call;
        var prototypeOfArray = Array.prototype;
        var prototypeOfObject = Object.prototype;
        var slice = prototypeOfArray.slice;
        var _toString = call.bind(prototypeOfObject.toString);
        var owns = call.bind(prototypeOfObject.hasOwnProperty);
        var defineGetter;
        var defineSetter;
        var lookupGetter;
        var lookupSetter;
        var supportsAccessors;
        if ((supportsAccessors = owns(prototypeOfObject, "__defineGetter__"))) {
            defineGetter = call.bind(prototypeOfObject.__defineGetter__);
            defineSetter = call.bind(prototypeOfObject.__defineSetter__);
            lookupGetter = call.bind(prototypeOfObject.__lookupGetter__);
            lookupSetter = call.bind(prototypeOfObject.__lookupSetter__);
        }
        if ([1,2].splice(0).length != 2) {
            if(function() { // test IE < 9 to splice bug - see issue #138
                function makeArray(l) {
                    var a = new Array(l+2);
                    a[0] = a[1] = 0;
                    return a;
                }
                var array = [], lengthBefore;

                array.splice.apply(array, makeArray(20));
                array.splice.apply(array, makeArray(26));

                lengthBefore = array.length; //46
                array.splice(5, 0, "XXX"); // add one element

                lengthBefore + 1 == array.length

                if (lengthBefore + 1 == array.length) {
                    return true;// has right splice implementation without bugs
                }
            }()) {//IE 6/7
                var array_splice = Array.prototype.splice;
                Array.prototype.splice = function(start, deleteCount) {
                    if (!arguments.length) {
                        return [];
                    } else {
                        return array_splice.apply(this, [
                            start === void 0 ? 0 : start,
                            deleteCount === void 0 ? (this.length - start) : deleteCount
                        ].concat(slice.call(arguments, 2)))
                    }
                };
            } else {//IE8
                Array.prototype.splice = function(pos, removeCount){
                    var length = this.length;
                    if (pos > 0) {
                        if (pos > length)
                        pos = length;
                    } else if (pos == void 0) {
                        pos = 0;
                    } else if (pos < 0) {
                        pos = Math.max(length + pos, 0);
                    }

                    if (!(pos+removeCount < length))
                    removeCount = length - pos;

                    var removed = this.slice(pos, pos+removeCount);
                    var insert = slice.call(arguments, 2);
                    var add = insert.length;
                    if (pos === length) {
                        if (add) {
                            this.push.apply(this, insert);
                        }
                    } else {
                        var remove = Math.min(removeCount, length - pos);
                        var tailOldPos = pos + remove;
                        var tailNewPos = tailOldPos + add - remove;
                        var tailCount = length - tailOldPos;
                        var lengthAfterRemove = length - remove;

                        if (tailNewPos < tailOldPos) { // case A
                            for (var i = 0; i < tailCount; ++i) {
                                this[tailNewPos+i] = this[tailOldPos+i];
                            }
                        } else if (tailNewPos > tailOldPos) { // case B
                            for (i = tailCount; i--; ) {
                                this[tailNewPos+i] = this[tailOldPos+i];
                            }
                        } // else, add == remove (nothing to do)

                        if (add && pos === lengthAfterRemove) {
                            this.length = lengthAfterRemove; // truncate array
                            this.push.apply(this, insert);
                        } else {
                            this.length = lengthAfterRemove + add; // reserves space
                            for (i = 0; i < add; ++i) {
                                this[pos+i] = insert[i];
                            }
                        }
                    }
                    return removed;
                };
            }
        }
        if (!Array.isArray) {
            Array.isArray = function isArray(obj) {
                return _toString(obj) == "[object Array]";
            };
        }
        var boxedString = Object("a"),
        splitString = boxedString[0] != "a" || !(0 in boxedString);

        if (!Array.prototype.forEach) {
            Array.prototype.forEach = function forEach(fun /*, thisp*/) {
                var object = toObject(this),
                self = splitString && _toString(this) == "[object String]" ?
                this.split("") :
                object,
                thisp = arguments[1],
                i = -1,
                length = self.length >>> 0;
                if (_toString(fun) != "[object Function]") {
                    throw new TypeError(); // TODO message
                }

                while (++i < length) {
                    if (i in self) {
                        fun.call(thisp, self[i], i, object);
                    }
                }
            };
        }
        if (!Array.prototype.map) {
            Array.prototype.map = function map(fun /*, thisp*/) {
                var object = toObject(this),
                self = splitString && _toString(this) == "[object String]" ?
                this.split("") :
                object,
                length = self.length >>> 0,
                result = Array(length),
                thisp = arguments[1];
                if (_toString(fun) != "[object Function]") {
                    throw new TypeError(fun + " is not a function");
                }

                for (var i = 0; i < length; i++) {
                    if (i in self)
                    result[i] = fun.call(thisp, self[i], i, object);
                }
                return result;
            };
        }
        if (!Array.prototype.filter) {
            Array.prototype.filter = function filter(fun /*, thisp */) {
                var object = toObject(this),
                self = splitString && _toString(this) == "[object String]" ?
                this.split("") :
                object,
                length = self.length >>> 0,
                result = [],
                value,
                thisp = arguments[1];
                if (_toString(fun) != "[object Function]") {
                    throw new TypeError(fun + " is not a function");
                }

                for (var i = 0; i < length; i++) {
                    if (i in self) {
                        value = self[i];
                        if (fun.call(thisp, value, i, object)) {
                            result.push(value);
                        }
                    }
                }
                return result;
            };
        }
        if (!Array.prototype.every) {
            Array.prototype.every = function every(fun /*, thisp */) {
                var object = toObject(this),
                self = splitString && _toString(this) == "[object String]" ?
                this.split("") :
                object,
                length = self.length >>> 0,
                thisp = arguments[1];
                if (_toString(fun) != "[object Function]") {
                    throw new TypeError(fun + " is not a function");
                }

                for (var i = 0; i < length; i++) {
                    if (i in self && !fun.call(thisp, self[i], i, object)) {
                        return false;
                    }
                }
                return true;
            };
        }
        if (!Array.prototype.some) {
            Array.prototype.some = function some(fun /*, thisp */) {
                var object = toObject(this),
                self = splitString && _toString(this) == "[object String]" ?
                this.split("") :
                object,
                length = self.length >>> 0,
                thisp = arguments[1];
                if (_toString(fun) != "[object Function]") {
                    throw new TypeError(fun + " is not a function");
                }

                for (var i = 0; i < length; i++) {
                    if (i in self && fun.call(thisp, self[i], i, object)) {
                        return true;
                    }
                }
                return false;
            };
        }
        if (!Array.prototype.reduce) {
            Array.prototype.reduce = function reduce(fun /*, initial*/) {
                var object = toObject(this),
                self = splitString && _toString(this) == "[object String]" ?
                this.split("") :
                object,
                length = self.length >>> 0;
                if (_toString(fun) != "[object Function]") {
                    throw new TypeError(fun + " is not a function");
                }
                if (!length && arguments.length == 1) {
                    throw new TypeError("reduce of empty array with no initial value");
                }

                var i = 0;
                var result;
                if (arguments.length >= 2) {
                    result = arguments[1];
                } else {
                    do {
                        if (i in self) {
                            result = self[i++];
                            break;
                        }
                        if (++i >= length) {
                            throw new TypeError("reduce of empty array with no initial value");
                        }
                    } while (true);
                }

                for (; i < length; i++) {
                    if (i in self) {
                        result = fun.call(void 0, result, self[i], i, object);
                    }
                }

                return result;
            };
        }
        if (!Array.prototype.reduceRight) {
            Array.prototype.reduceRight = function reduceRight(fun /*, initial*/) {
                var object = toObject(this),
                self = splitString && _toString(this) == "[object String]" ?
                this.split("") :
                object,
                length = self.length >>> 0;
                if (_toString(fun) != "[object Function]") {
                    throw new TypeError(fun + " is not a function");
                }
                if (!length && arguments.length == 1) {
                    throw new TypeError("reduceRight of empty array with no initial value");
                }

                var result, i = length - 1;
                if (arguments.length >= 2) {
                    result = arguments[1];
                } else {
                    do {
                        if (i in self) {
                            result = self[i--];
                            break;
                        }
                        if (--i < 0) {
                            throw new TypeError("reduceRight of empty array with no initial value");
                        }
                    } while (true);
                }

                do {
                    if (i in this) {
                        result = fun.call(void 0, result, self[i], i, object);
                    }
                } while (i--);

                return result;
            };
        }
        if (!Array.prototype.indexOf || ([0, 1].indexOf(1, 2) != -1)) {
            Array.prototype.indexOf = function indexOf(sought /*, fromIndex */ ) {
                var self = splitString && _toString(this) == "[object String]" ?
                this.split("") :
                toObject(this),
                length = self.length >>> 0;

                if (!length) {
                    return -1;
                }

                var i = 0;
                if (arguments.length > 1) {
                    i = toInteger(arguments[1]);
                }
                i = i >= 0 ? i : Math.max(0, length + i);
                for (; i < length; i++) {
                    if (i in self && self[i] === sought) {
                        return i;
                    }
                }
                return -1;
            };
        }
        if (!Array.prototype.lastIndexOf || ([0, 1].lastIndexOf(0, -3) != -1)) {
            Array.prototype.lastIndexOf = function lastIndexOf(sought /*, fromIndex */) {
                var self = splitString && _toString(this) == "[object String]" ?
                this.split("") :
                toObject(this),
                length = self.length >>> 0;

                if (!length) {
                    return -1;
                }
                var i = length - 1;
                if (arguments.length > 1) {
                    i = Math.min(i, toInteger(arguments[1]));
                }
                i = i >= 0 ? i : length - Math.abs(i);
                for (; i >= 0; i--) {
                    if (i in self && sought === self[i]) {
                        return i;
                    }
                }
                return -1;
            };
        }
        if (!Object.getPrototypeOf) {
            Object.getPrototypeOf = function getPrototypeOf(object) {
                return object.__proto__ || (
                    object.constructor ?
                    object.constructor.prototype :
                    prototypeOfObject
                );
            };
        }
        if (!Object.getOwnPropertyDescriptor) {
            var ERR_NON_OBJECT = "Object.getOwnPropertyDescriptor called on a " +
            "non-object: ";
            Object.getOwnPropertyDescriptor = function getOwnPropertyDescriptor(object, property) {
                if ((typeof object != "object" && typeof object != "function") || object === null)
                throw new TypeError(ERR_NON_OBJECT + object);
                if (!owns(object, property))
                return;

                var descriptor, getter, setter;
                descriptor =  { enumerable: true, configurable: true };
                if (supportsAccessors) {
                    var prototype = object.__proto__;
                    object.__proto__ = prototypeOfObject;

                    var getter = lookupGetter(object, property);
                    var setter = lookupSetter(object, property);
                    object.__proto__ = prototype;

                    if (getter || setter) {
                        if (getter) descriptor.get = getter;
                        if (setter) descriptor.set = setter;
                        return descriptor;
                    }
                }
                descriptor.value = object[property];
                return descriptor;
            };
        }
        if (!Object.getOwnPropertyNames) {
            Object.getOwnPropertyNames = function getOwnPropertyNames(object) {
                return Object.keys(object);
            };
        }
        if (!Object.create) {
            var createEmpty;
            if (Object.prototype.__proto__ === null) {
                createEmpty = function () {
                    return { "__proto__": null };
                };
            } else {
                createEmpty = function () {
                    var empty = {};
                    for (var i in empty)
                    empty[i] = null;
                    empty.constructor =
                    empty.hasOwnProperty =
                    empty.propertyIsEnumerable =
                    empty.isPrototypeOf =
                    empty.toLocaleString =
                    empty.toString =
                    empty.valueOf =
                    empty.__proto__ = null;
                    return empty;
                }
            }

            Object.create = function create(prototype, properties) {
                var object;
                if (prototype === null) {
                    object = createEmpty();
                } else {
                    if (typeof prototype != "object")
                    throw new TypeError("typeof prototype["+(typeof prototype)+"] != 'object'");
                    var Type = function () {};
                    Type.prototype = prototype;
                    object = new Type();
                    object.__proto__ = prototype;
                }
                if (properties !== void 0)
                Object.defineProperties(object, properties);
                return object;
            };
        }

        function doesDefinePropertyWork(object) {
            try {
                Object.defineProperty(object, "sentinel", {});
                return "sentinel" in object;
            } catch (exception) {
            }
        }
        if (Object.defineProperty) {
            var definePropertyWorksOnObject = doesDefinePropertyWork({});
            var definePropertyWorksOnDom = typeof document == "undefined" ||
            doesDefinePropertyWork(document.createElement("div"));
            if (!definePropertyWorksOnObject || !definePropertyWorksOnDom) {
                var definePropertyFallback = Object.defineProperty;
            }
        }

        if (!Object.defineProperty || definePropertyFallback) {
            var ERR_NON_OBJECT_DESCRIPTOR = "Property description must be an object: ";
            var ERR_NON_OBJECT_TARGET = "Object.defineProperty called on non-object: "
            var ERR_ACCESSORS_NOT_SUPPORTED = "getters & setters can not be defined " +
            "on this javascript engine";

            Object.defineProperty = function defineProperty(object, property, descriptor) {
                if ((typeof object != "object" && typeof object != "function") || object === null)
                throw new TypeError(ERR_NON_OBJECT_TARGET + object);
                if ((typeof descriptor != "object" && typeof descriptor != "function") || descriptor === null)
                throw new TypeError(ERR_NON_OBJECT_DESCRIPTOR + descriptor);
                if (definePropertyFallback) {
                    try {
                        return definePropertyFallback.call(Object, object, property, descriptor);
                    } catch (exception) {
                    }
                }
                if (owns(descriptor, "value")) {

                    if (supportsAccessors && (lookupGetter(object, property) ||
                    lookupSetter(object, property)))
                    {
                        var prototype = object.__proto__;
                        object.__proto__ = prototypeOfObject;
                        delete object[property];
                        object[property] = descriptor.value;
                        object.__proto__ = prototype;
                    } else {
                        object[property] = descriptor.value;
                    }
                } else {
                    if (!supportsAccessors)
                    throw new TypeError(ERR_ACCESSORS_NOT_SUPPORTED);
                    if (owns(descriptor, "get"))
                    defineGetter(object, property, descriptor.get);
                    if (owns(descriptor, "set"))
                    defineSetter(object, property, descriptor.set);
                }

                return object;
            };
        }
        if (!Object.defineProperties) {
            Object.defineProperties = function defineProperties(object, properties) {
                for (var property in properties) {
                    if (owns(properties, property))
                    Object.defineProperty(object, property, properties[property]);
                }
                return object;
            };
        }
        if (!Object.seal) {
            Object.seal = function seal(object) {
                return object;
            };
        }
        if (!Object.freeze) {
            Object.freeze = function freeze(object) {
                return object;
            };
        }
        try {
            Object.freeze(function () {});
        } catch (exception) {
            Object.freeze = (function freeze(freezeObject) {
                return function freeze(object) {
                    if (typeof object == "function") {
                        return object;
                    } else {
                        return freezeObject(object);
                    }
                };
            })(Object.freeze);
        }
        if (!Object.preventExtensions) {
            Object.preventExtensions = function preventExtensions(object) {
                return object;
            };
        }
        if (!Object.isSealed) {
            Object.isSealed = function isSealed(object) {
                return false;
            };
        }
        if (!Object.isFrozen) {
            Object.isFrozen = function isFrozen(object) {
                return false;
            };
        }
        if (!Object.isExtensible) {
            Object.isExtensible = function isExtensible(object) {
                if (Object(object) === object) {
                    throw new TypeError(); // TODO message
                }
                var name = '';
                while (owns(object, name)) {
                    name += '?';
                }
                object[name] = true;
                var returnValue = owns(object, name);
                delete object[name];
                return returnValue;
            };
        }
        if (!Object.keys) {
            var hasDontEnumBug = true,
            dontEnums = [
                "toString",
                "toLocaleString",
                "valueOf",
                "hasOwnProperty",
                "isPrototypeOf",
                "propertyIsEnumerable",
                "constructor"
            ],
            dontEnumsLength = dontEnums.length;

            for (var key in {"toString": null}) {
                hasDontEnumBug = false;
            }

            Object.keys = function keys(object) {

                if (
                    (typeof object != "object" && typeof object != "function") ||
                    object === null
                ) {
                    throw new TypeError("Object.keys called on a non-object");
                }

                var keys = [];
                for (var name in object) {
                    if (owns(object, name)) {
                        keys.push(name);
                    }
                }

                if (hasDontEnumBug) {
                    for (var i = 0, ii = dontEnumsLength; i < ii; i++) {
                        var dontEnum = dontEnums[i];
                        if (owns(object, dontEnum)) {
                            keys.push(dontEnum);
                        }
                    }
                }
                return keys;
            };

        }
        if (!Date.now) {
            Date.now = function now() {
                return new Date().getTime();
            };
        }
        var ws = "\x09\x0A\x0B\x0C\x0D\x20\xA0\u1680\u180E\u2000\u2001\u2002\u2003" +
        "\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\u2028" +
        "\u2029\uFEFF";
        if (!String.prototype.trim || ws.trim()) {
            ws = "[" + ws + "]";
            var trimBeginRegexp = new RegExp("^" + ws + ws + "*"),
            trimEndRegexp = new RegExp(ws + ws + "*$");
            String.prototype.trim = function trim() {
                return String(this).replace(trimBeginRegexp, "").replace(trimEndRegexp, "");
            };
        }

        function toInteger(n) {
            n = +n;
            if (n !== n) { // isNaN
                n = 0;
            } else if (n !== 0 && n !== (1/0) && n !== -(1/0)) {
                n = (n > 0 || -1) * Math.floor(Math.abs(n));
            }
            return n;
        }

        function isPrimitive(input) {
            var type = typeof input;
            return (
                input === null ||
                type === "undefined" ||
                type === "boolean" ||
                type === "number" ||
                type === "string"
            );
        }

        function toPrimitive(input) {
            var val, valueOf, toString;
            if (isPrimitive(input)) {
                return input;
            }
            valueOf = input.valueOf;
            if (typeof valueOf === "function") {
                val = valueOf.call(input);
                if (isPrimitive(val)) {
                    return val;
                }
            }
            toString = input.toString;
            if (typeof toString === "function") {
                val = toString.call(input);
                if (isPrimitive(val)) {
                    return val;
                }
            }
            throw new TypeError();
        }
        var toObject = function (o) {
            if (o == null) { // this matches both null and undefined
                throw new TypeError("can't convert "+o+" to object");
            }
            return Object(o);
        };

    });
    define("ace/mode/jsonata_worker",["require","exports","ace/lib/oop","ace/worker/mirror"], function(require, exports) {
        var oop = require("../lib/oop");
        var Mirror = require("../worker/mirror").Mirror;

        var JSONataWorker = exports.JSONataWorker = function(sender) {
            Mirror.call(this, sender);
            this.setTimeout(200);
        };

        oop.inherits(JSONataWorker, Mirror);

        (function() {

            this.onUpdate = function() {
                var value = this.doc.getValue();
                var errors = [];
                try {
                    if (value) {
                        jsonata(value);
                    }
                } catch (e) {
                    var pos = this.doc.indexToPosition(e.position-1);
                    var msg = e.message;
                    msg = msg.replace(/ at column \d+/,"");
                    errors.push({
                        row: pos.row,
                        column: pos.column,
                        text: msg,
                        type: "error"
                    });
                }
                this.sender.emit("annotate", errors);
            };

        }).call(JSONataWorker.prototype);

    });
